# Coronavirus phylogenetics

In this lab, we'll use Nextstrain to explore the evolution of SARS-CoV-2 and then build our own coronavirus phylogenies within R.

#### Learning objectives

After completing this chapter, you'll be able to:

1. Interpret viral phylogenies and associated data on Nextstrain.
2. Explain how incomplete sampling affects phylogenies.
3. Describe how phylogenetic trees are built through neighbor joining.


***


## Nextstrain

https://docs.nextstrain.org/en/latest/learn/interpret/how-to-read-a-tree.html

https://nextstrain.org/ncov/gisaid/global/all-time


***


## Setup

### R packages

```{r}
library(tidyverse)
library(ape)
library(ggtree)
library(harrietr)
```

## Data

#### Metadata

```{r}
accessions <- read.table("accessions.txt", header = TRUE, sep = "\t")
head(accessions)
```

```{r}
# make vectors of the GenBank IDs and full names
# these will be used as input to functions later
ids <- accessions$id
names <- accessions$name
```

#### DNA sequences

`aligned.fa`


***


## Pairwise distance matrix

Read in FASTA file of aligned sequences:
 
```{r}
dna <- read.dna("aligned.fa", format = "fasta", as.matrix = TRUE)
```

Compute the pairwise **distance matrix**:

```{r}
D <- dist.dna(dna, model = "TN93", as.matrix = TRUE)
```

Plot this matrix to visualize:

```{r}
# use the "melt_dist" function from harrietr package to convert
# the distance matrix to "long" format for ggplot
D_melted <- rbind(melt_dist(D, order = ids),
                  melt_dist(t(D), order = rev(ids)))

# plot distance matrix
ggplot(data = D_melted) +
  geom_tile(aes(x = iso1, y = iso2, fill = (dist + 1e-5))) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_viridis_c(name = "distance") +
  xlab("sample") +
  ylab("sample")
```


***


## Neighbor joining

```{r}
# build a neighbor joining tree
tree <- nj(D)
# manually "root" the tree by setting HQ166910 as an outgroup
tree <- root(tree, which(ids == "HQ166910"))
# rotate tree at nodes to make it look tidier (i.e., "ladderized")
tree <- ladderize(tree)

# plot the tree
ggtree(tree) +
  theme_tree2() +
  geom_tiplab(label = names, size = 4) +
  xlim(0, 1.2)
```


***


## Assess bootstrap support

```{r}
# set random seed
set.seed(123)
# bootstrap and build new trees to evaluate uncertainty
myBoots <- boot.phylo(tree, dna, 
                      function(x) ladderize(root(nj(dist.dna(x,
                                                             model = "TN93")),
                                                 which(ids == "HQ166910"))), 
                      rooted = TRUE)

# replace "NA" with zero in bootstrap results
myBoots[is.na(myBoots)] <- 0

# re-plot tree with bootstrap values
ggtree(tree, branch.length = "none") +
  theme_tree2() +
  geom_tiplab(label = names) +
  geom_nodelab(label = myBoots, geom = "label", fill = "#deebf7") +
  xlim(0, 30)
```