<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3.7 Homework | Human Genome Variation Lab</title>
  <meta name="description" content="Description of Course." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="3.7 Homework | Human Genome Variation Lab" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Description of Course." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3.7 Homework | Human Genome Variation Lab" />
  
  <meta name="twitter:description" content="Description of Course." />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="assets/dasl_favicon.ico" type="image/x-icon" />
<link rel="prev" href="conclusion-1.html"/>
<link rel="next" href="simulating-evolution.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="assets/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="http://jhudatascience.org/"><img src="https://jhudatascience.org/images/dasl.png" style=" width: 80%; padding-left: 40px; padding-top: 8px; vertical-align: top"</a>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Homepage</a>
<ul>
<li class="chapter" data-level="1.1" data-path="syllabus-logistics.html"><a href="syllabus-logistics.html"><i class="fa fa-check"></i><b>1.1</b> Syllabus &amp; Logistics</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="syllabus-logistics.html"><a href="syllabus-logistics.html#assessment-grading"><i class="fa fa-check"></i><b>1.1.1</b> Assessment &amp; Grading</a></li>
<li class="chapter" data-level="1.1.2" data-path="syllabus-logistics.html"><a href="syllabus-logistics.html#homework"><i class="fa fa-check"></i><b>1.1.2</b> Homework</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="course-description.html"><a href="course-description.html"><i class="fa fa-check"></i><b>1.2</b> Course Description</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="course-description.html"><a href="course-description.html#educational-objectives"><i class="fa fa-check"></i><b>1.2.1</b> Educational Objectives</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="rstudio-cloud.html"><a href="rstudio-cloud.html"><i class="fa fa-check"></i><b>1.3</b> RStudio Cloud</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="rstudio-cloud.html"><a href="rstudio-cloud.html#making-an-rstudio-cloud-account"><i class="fa fa-check"></i><b>1.3.1</b> Making an RStudio Cloud account</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="genome-browsers.html"><a href="genome-browsers.html"><i class="fa fa-check"></i><b>2</b> Genome browsers</a>
<ul>
<li class="chapter" data-level="2.1" data-path="dna-sequencing-data.html"><a href="dna-sequencing-data.html"><i class="fa fa-check"></i><b>2.1</b> DNA sequencing data</a></li>
<li class="chapter" data-level="2.2" data-path="assembling-a-genome.html"><a href="assembling-a-genome.html"><i class="fa fa-check"></i><b>2.2</b> Assembling a genome</a></li>
<li class="chapter" data-level="2.3" data-path="the-human-reference-genome.html"><a href="the-human-reference-genome.html"><i class="fa fa-check"></i><b>2.3</b> The human reference genome</a></li>
<li class="chapter" data-level="2.4" data-path="ucsc-genome-browser.html"><a href="ucsc-genome-browser.html"><i class="fa fa-check"></i><b>2.4</b> UCSC genome browser</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="ucsc-genome-browser.html"><a href="ucsc-genome-browser.html#homepage-1"><i class="fa fa-check"></i><b>2.4.1</b> Homepage</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="viewing-one-region-of-the-genome.html"><a href="viewing-one-region-of-the-genome.html"><i class="fa fa-check"></i><b>2.5</b> Viewing one region of the genome</a></li>
<li class="chapter" data-level="2.6" data-path="customizing-browser-displays.html"><a href="customizing-browser-displays.html"><i class="fa fa-check"></i><b>2.6</b> Customizing browser displays</a></li>
<li class="chapter" data-level="2.7" data-path="igv.html"><a href="igv.html"><i class="fa fa-check"></i><b>2.7</b> IGV</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="igv.html"><a href="igv.html#homepage-2"><i class="fa fa-check"></i><b>2.7.1</b> Homepage</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="navigating-igv.html"><a href="navigating-igv.html"><i class="fa fa-check"></i><b>2.8</b> Navigating IGV</a></li>
<li class="chapter" data-level="2.9" data-path="loading-sequencing-data.html"><a href="loading-sequencing-data.html"><i class="fa fa-check"></i><b>2.9</b> Loading sequencing data</a></li>
<li class="chapter" data-level="2.10" data-path="the-1000-genomes-project.html"><a href="the-1000-genomes-project.html"><i class="fa fa-check"></i><b>2.10</b> The 1000 Genomes Project</a></li>
<li class="chapter" data-level="2.11" data-path="sra.html"><a href="sra.html"><i class="fa fa-check"></i><b>2.11</b> SRA</a>
<ul>
<li class="chapter" data-level="2.11.1" data-path="sra.html"><a href="sra.html#previewing-sequencing-data"><i class="fa fa-check"></i><b>2.11.1</b> Previewing sequencing data</a></li>
</ul></li>
<li class="chapter" data-level="2.12" data-path="viewing-sequencing-reads-in-igv.html"><a href="viewing-sequencing-reads-in-igv.html"><i class="fa fa-check"></i><b>2.12</b> Viewing sequencing reads in IGV</a></li>
<li class="chapter" data-level="2.13" data-path="interpreting-igv-alignments.html"><a href="interpreting-igv-alignments.html"><i class="fa fa-check"></i><b>2.13</b> Interpreting IGV alignments</a></li>
<li class="chapter" data-level="2.14" data-path="conclusion.html"><a href="conclusion.html"><i class="fa fa-check"></i><b>2.14</b> Conclusion</a></li>
<li class="chapter" data-level="2.15" data-path="homework-1.html"><a href="homework-1.html"><i class="fa fa-check"></i><b>2.15</b> Homework</a>
<ul>
<li class="chapter" data-level="2.15.1" data-path="homework-1.html"><a href="homework-1.html#required-homework"><i class="fa fa-check"></i><b>2.15.1</b> Required homework</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="discovering-mutations.html"><a href="discovering-mutations.html"><i class="fa fa-check"></i><b>3</b> Discovering mutations</a>
<ul>
<li class="chapter" data-level="3.1" data-path="background.html"><a href="background.html"><i class="fa fa-check"></i><b>3.1</b> Background</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="background.html"><a href="background.html#de-novo-mutations"><i class="fa fa-check"></i><b>3.1.1</b> _De novo) mutations</a></li>
<li class="chapter" data-level="3.1.2" data-path="background.html"><a href="background.html#recombination"><i class="fa fa-check"></i><b>3.1.2</b> Recombination</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="setup.html"><a href="setup.html"><i class="fa fa-check"></i><b>3.2</b> Setup</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="setup.html"><a href="setup.html#using-r-markdown-files"><i class="fa fa-check"></i><b>3.2.1</b> Using R Markdown files</a></li>
<li class="chapter" data-level="3.2.2" data-path="setup.html"><a href="setup.html#r-packages"><i class="fa fa-check"></i><b>3.2.2</b> R packages</a></li>
<li class="chapter" data-level="3.2.3" data-path="setup.html"><a href="setup.html#data"><i class="fa fa-check"></i><b>3.2.3</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="visualizing-the-data.html"><a href="visualizing-the-data.html"><i class="fa fa-check"></i><b>3.3</b> Visualizing the data</a></li>
<li class="chapter" data-level="3.4" data-path="linear-models.html"><a href="linear-models.html"><i class="fa fa-check"></i><b>3.4</b> Linear models</a></li>
<li class="chapter" data-level="3.5" data-path="confidence-intervals.html"><a href="confidence-intervals.html"><i class="fa fa-check"></i><b>3.5</b> Confidence intervals</a></li>
<li class="chapter" data-level="3.6" data-path="conclusion-1.html"><a href="conclusion-1.html"><i class="fa fa-check"></i><b>3.6</b> Conclusion</a></li>
<li class="chapter" data-level="3.7" data-path="homework-2.html"><a href="homework-2.html"><i class="fa fa-check"></i><b>3.7</b> Homework</a>
<ul>
<li class="chapter" data-level="3.7.1" data-path="homework-2.html"><a href="homework-2.html#required"><i class="fa fa-check"></i><b>3.7.1</b> Required</a></li>
<li class="chapter" data-level="3.7.2" data-path="homework-2.html"><a href="homework-2.html#extra-credit-more-linear-models"><i class="fa fa-check"></i><b>3.7.2</b> Extra credit: More linear models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="simulating-evolution.html"><a href="simulating-evolution.html"><i class="fa fa-check"></i><b>4</b> Simulating evolution</a>
<ul>
<li class="chapter" data-level="4.1" data-path="background-1.html"><a href="background-1.html"><i class="fa fa-check"></i><b>4.1</b> Background</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="background-1.html"><a href="background-1.html#genetic-drift"><i class="fa fa-check"></i><b>4.1.1</b> Genetic drift</a></li>
<li class="chapter" data-level="4.1.2" data-path="background-1.html"><a href="background-1.html#the-wright-fisher-model"><i class="fa fa-check"></i><b>4.1.2</b> The Wright-Fisher model</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="modeling-allele-frequencies.html"><a href="modeling-allele-frequencies.html"><i class="fa fa-check"></i><b>4.2</b> Modeling allele frequencies</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="modeling-allele-frequencies.html"><a href="modeling-allele-frequencies.html#the-binomial-distribution"><i class="fa fa-check"></i><b>4.2.1</b> The binomial distribution</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="setup-1.html"><a href="setup-1.html"><i class="fa fa-check"></i><b>4.3</b> Setup</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="setup-1.html"><a href="setup-1.html#r-packages-1"><i class="fa fa-check"></i><b>4.3.1</b> R packages</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="a-wright-fisher-simulation.html"><a href="a-wright-fisher-simulation.html"><i class="fa fa-check"></i><b>4.4</b> A Wright-Fisher simulation</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="a-wright-fisher-simulation.html"><a href="a-wright-fisher-simulation.html#drawing-one-generation-from-a-binomial-distribution"><i class="fa fa-check"></i><b>4.4.1</b> Drawing one generation from a binomial distribution</a></li>
<li class="chapter" data-level="4.4.2" data-path="a-wright-fisher-simulation.html"><a href="a-wright-fisher-simulation.html#simulating-multiple-generations-with-a-for-loop"><i class="fa fa-check"></i><b>4.4.2</b> Simulating multiple generations with a for loop</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>4.5</b> Plotting</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="plotting.html"><a href="plotting.html#visualizing-changes-in-af-over-generations"><i class="fa fa-check"></i><b>4.5.1</b> Visualizing changes in AF over generations</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="creating-a-simulation-function.html"><a href="creating-a-simulation-function.html"><i class="fa fa-check"></i><b>4.6</b> Creating a simulation function</a>
<ul>
<li class="chapter" data-level="4.6.1" data-path="creating-a-simulation-function.html"><a href="creating-a-simulation-function.html#simulating-with-different-parameters-with-a-function"><i class="fa fa-check"></i><b>4.6.1</b> Simulating with different parameters with a function</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="conclusion-2.html"><a href="conclusion-2.html"><i class="fa fa-check"></i><b>4.7</b> Conclusion</a></li>
<li class="chapter" data-level="4.8" data-path="homework-3.html"><a href="homework-3.html"><i class="fa fa-check"></i><b>4.8</b> Homework</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="homework-3.html"><a href="homework-3.html#required-1"><i class="fa fa-check"></i><b>4.8.1</b> Required</a></li>
<li class="chapter" data-level="4.8.2" data-path="homework-3.html"><a href="homework-3.html#extra-credit-selection"><i class="fa fa-check"></i><b>4.8.2</b> Extra credit: Selection</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="about-the-authors.html"><a href="about-the-authors.html"><i class="fa fa-check"></i>About the Authors</a></li>
<li class="divider"></li>
<p style="text-align:center; font-size: 11pt"> <a href="https://github.com/jhudsl/OTTR_Template" target="blank" > Published with</a> <a href="https://bookdown.org/" target="blank"> <b>bookdown</b></a> (by <a href="http://jhudatascience.org/"> The Johns Hopkins Data Science Lab</a>) </p>
<p style="text-align:center; font-size: 12px;"> <a href="https://github.com/rstudio4edu/rstudio4edu-book/"> Style adapted from: rstudio4edu-book </a> <a href ="https://creativecommons.org/licenses/by/2.0/"> (CC-BY 2.0) </a></p>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Human Genome Variation Lab</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="homework-2" class="section level2" number="3.7">
<h2><span class="header-section-number">3.7</span> Homework</h2>
<div id="goals-learning-objectives-1" class="section level4" number="3.7.0.1">
<h4><span class="header-section-number">3.7.0.1</span> Goals &amp; Learning Objectives</h4>
<p>The goal of this homework is to explore the distribution of crossovers in the genome.</p>
<p><strong>Learning Objectives</strong></p>
<ul>
<li>Required homework: Practice visualizing data with <code>ggplot2</code></li>
<li>Extra credit: Practice interpreting linear models</li>
</ul>
</div>
<div id="required" class="section level3" number="3.7.1">
<h3><span class="header-section-number">3.7.1</span> Required</h3>
<p>We will be using a data file from the Halldorsson et al. study with information about the crossovers that they observed in probands.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="homework-2.html#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># crossovers &lt;- read_delim(&quot;https://www.science.org/doi/suppl/10.1126/science.aau1043/suppl_file/aau1043_datas4.gz&quot;, delim = &quot;\t&quot;, skip = 10)</span></span>
<span id="cb30-2"><a href="homework-2.html#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="co"># this file is on the McCoy lab onedrive because the science link is broken</span></span>
<span id="cb30-3"><a href="homework-2.html#cb30-3" aria-hidden="true" tabindex="-1"></a>crossovers <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;https://www.dropbox.com/s/imk845vw1uvnixy/aau1043_datas4.gz?dl=1&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">skip =</span> <span class="dv">10</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-4"><a href="homework-2.html#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(crossovers)</span></code></pre></div>
<pre><code>##    Chr    lbnd    ubnd   pid ptype  medsnp complex
## 1 chr1 1431813 1790040  7608     P 1583857      NA
## 2 chr1 1431813 1586215 14736     P 1583602      NA
## 3 chr1 1431813 1814287 17536     P 1583857      NA
## 4 chr1 1431813 1586215 19035     P 1583602      NA
## 5 chr1 1431813 1538787 22609     M 1487683      NA
## 6 chr1 1431813 1968793 28975     P 1584039      NA</code></pre>
<ul>
<li><code>Chr</code>: Chromosome that the crossover occured on</li>
<li><code>lbnd</code>: Lower bound of crossover region</li>
<li><code>ubnd</code>: Upper bound of crossover region</li>
<li><code>pid</code>: Proband ID</li>
<li><code>ptype</code>: Parent that the crossover originated in (M: maternal, P: paternal)</li>
<li><code>medsnp</code>: Middle of crossover region</li>
<li><code>complex</code>: Indicator for whether this was a complex crossover (0: no, 1: yes)</li>
</ul>
<p><strong>Assignment:</strong> Plot the <em>distribution</em> of crossover locations for each chromosome. Add <code>facet_wrap(~ Chr, scales = "free")</code> to your ggplot object in order to separate the data by chromosome.</p>
<hr />
<details>
<summary>
Hint
</summary>
What kind of graph should we use to show the <em>distribution</em> of data?
</details>
<hr />
<p><strong>Question:</strong> Does this look like the distribution you would expect for crossovers? Where do you expect crossovers to occur more frequently?</p>
<hr />
<details>
<summary>
Solution
</summary>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="homework-2.html#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> crossovers,</span>
<span id="cb32-2"><a href="homework-2.html#cb32-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> medsnp)) <span class="sc">+</span></span>
<span id="cb32-3"><a href="homework-2.html#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>() <span class="sc">+</span></span>
<span id="cb32-4"><a href="homework-2.html#cb32-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> Chr, <span class="at">scales =</span> <span class="st">&quot;free&quot;</span>)</span></code></pre></div>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="resources/images/02-dnm_files/figure-html/unnamed-chunk-28-1.png" width="672" /></p>
<p>It looks like a lot of crossovers are occuring near the ends of chromosomes. These regions are known to have elevated rates of recombination.</p>
</details>
<hr />
</div>
<div id="extra-credit-more-linear-models" class="section level3" number="3.7.2">
<h3><span class="header-section-number">3.7.2</span> Extra credit: More linear models</h3>
<p>Use this code to generate a new dataframe, <code>crossovers_sum</code>, which sums the number of crossovers per child and adds information about the parents’ ages:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="homework-2.html#cb34-1" aria-hidden="true" tabindex="-1"></a>crossovers_sum <span class="ot">&lt;-</span> crossovers <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="homework-2.html#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(pid) <span class="sc">%&gt;%</span></span>
<span id="cb34-3"><a href="homework-2.html#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">num_crossovers =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb34-4"><a href="homework-2.html#cb34-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">merge</span>(., age, <span class="at">by.x =</span> <span class="st">&quot;pid&quot;</span>, <span class="at">by.y =</span> <span class="st">&quot;Proband_id&quot;</span>)</span></code></pre></div>
<p><strong>Assignment:</strong> Fit two linear models to ask if there is an association between a child’s number of crossovers and maternal age or paternal age. If there is an association, how is the number of crossovers predicted to change with every year of maternal/paternal age?</p>
<hr />
<details>
<summary>
Solution
</summary>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="homework-2.html#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fit the model with maternal age</span></span>
<span id="cb35-2"><a href="homework-2.html#cb35-2" aria-hidden="true" tabindex="-1"></a>fit_mat <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">data =</span> crossovers_sum,</span>
<span id="cb35-3"><a href="homework-2.html#cb35-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">formula =</span> num_crossovers <span class="sc">~</span> Mother_age)</span>
<span id="cb35-4"><a href="homework-2.html#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit_mat)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = num_crossovers ~ Mother_age, data = crossovers_sum)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -42.460 -16.361   1.133  15.787  62.490 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept) 54.49776    0.30943 176.125  &lt; 2e-16 ***
## Mother_age  -0.04941    0.01126  -4.387 1.15e-05 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 19.13 on 84873 degrees of freedom
##   (440 observations deleted due to missingness)
## Multiple R-squared:  0.0002267,  Adjusted R-squared:  0.0002149 
## F-statistic: 19.25 on 1 and 84873 DF,  p-value: 1.15e-05</code></pre>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="homework-2.html#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the coefficient for `Mother_age`, which tells us the effect size</span></span>
<span id="cb37-2"><a href="homework-2.html#cb37-2" aria-hidden="true" tabindex="-1"></a>fit_mat<span class="sc">$</span>coefficients[<span class="st">&quot;Mother_age&quot;</span>]</span></code></pre></div>
<pre><code>##  Mother_age 
## -0.04941053</code></pre>
<p>There is a significant association between maternal age and the child’s number of crossovers. For every year of maternal age, we expect 0.05 fewer crossovers in the child.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="homework-2.html#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fit the model with paternal age</span></span>
<span id="cb39-2"><a href="homework-2.html#cb39-2" aria-hidden="true" tabindex="-1"></a>fit_pat <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">data =</span> crossovers_sum,</span>
<span id="cb39-3"><a href="homework-2.html#cb39-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">formula =</span> num_crossovers <span class="sc">~</span> Father_age)</span>
<span id="cb39-4"><a href="homework-2.html#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit_pat)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = num_crossovers ~ Father_age, data = crossovers_sum)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -42.978 -16.283   1.187  15.763  62.234 
## 
## Coefficients:
##              Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept) 56.308135   0.295117   190.8   &lt;2e-16 ***
## Father_age  -0.105913   0.009715   -10.9   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 19.12 on 84873 degrees of freedom
##   (440 observations deleted due to missingness)
## Multiple R-squared:  0.001399,   Adjusted R-squared:  0.001387 
## F-statistic: 118.9 on 1 and 84873 DF,  p-value: &lt; 2.2e-16</code></pre>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="homework-2.html#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the coefficient for `Father_age`, which tells us the effect size</span></span>
<span id="cb41-2"><a href="homework-2.html#cb41-2" aria-hidden="true" tabindex="-1"></a>fit_pat<span class="sc">$</span>coefficients[<span class="st">&quot;Father_age&quot;</span>]</span></code></pre></div>
<pre><code>## Father_age 
## -0.1059134</code></pre>
<p>There is also a significant association between paternal age and the child’s number of crossovers. For every year of paternal age, we expect 0.11 fewer crossovers in the child.</p>
</details>
<hr />

</div>
</div>
<!-- </div> -->
<hr>
<center> 
  <div class="footer">
      All illustrations <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY. </a>
      <br>
      All other materials <a href= "https://creativecommons.org/licenses/by/4.0/"> CC-BY </a> unless noted otherwise.
  </div>
</center>
            </section>

          </div>
        </div>
      </div>
<a href="conclusion-1.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="simulating-evolution.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
