<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3.2 Setup | Human Genome Variation Lab</title>
  <meta name="description" content="Description of Course." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="3.2 Setup | Human Genome Variation Lab" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Description of Course." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3.2 Setup | Human Genome Variation Lab" />
  
  <meta name="twitter:description" content="Description of Course." />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="assets/dasl_favicon.ico" type="image/x-icon" />
<link rel="prev" href="background.html"/>
<link rel="next" href="visualizing-the-data.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="assets/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="http://jhudatascience.org/"><img src="https://jhudatascience.org/images/dasl.png" style=" width: 80%; padding-left: 40px; padding-top: 8px; vertical-align: top"</a>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Homepage</a>
<ul>
<li class="chapter" data-level="1.1" data-path="syllabus-logistics.html"><a href="syllabus-logistics.html"><i class="fa fa-check"></i><b>1.1</b> Syllabus &amp; Logistics</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="syllabus-logistics.html"><a href="syllabus-logistics.html#assessment-grading"><i class="fa fa-check"></i><b>1.1.1</b> Assessment &amp; Grading</a></li>
<li class="chapter" data-level="1.1.2" data-path="syllabus-logistics.html"><a href="syllabus-logistics.html#homework"><i class="fa fa-check"></i><b>1.1.2</b> Homework</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="course-description.html"><a href="course-description.html"><i class="fa fa-check"></i><b>1.2</b> Course Description</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="course-description.html"><a href="course-description.html#educational-objectives"><i class="fa fa-check"></i><b>1.2.1</b> Educational Objectives</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="rstudio-cloud.html"><a href="rstudio-cloud.html"><i class="fa fa-check"></i><b>1.3</b> RStudio Cloud</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="rstudio-cloud.html"><a href="rstudio-cloud.html#making-an-rstudio-cloud-account"><i class="fa fa-check"></i><b>1.3.1</b> Making an RStudio Cloud account</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="genome-browsers.html"><a href="genome-browsers.html"><i class="fa fa-check"></i><b>2</b> Genome browsers</a>
<ul>
<li class="chapter" data-level="2.1" data-path="dna-sequencing-data.html"><a href="dna-sequencing-data.html"><i class="fa fa-check"></i><b>2.1</b> DNA sequencing data</a></li>
<li class="chapter" data-level="2.2" data-path="assembling-a-genome.html"><a href="assembling-a-genome.html"><i class="fa fa-check"></i><b>2.2</b> Assembling a genome</a></li>
<li class="chapter" data-level="2.3" data-path="the-human-reference-genome.html"><a href="the-human-reference-genome.html"><i class="fa fa-check"></i><b>2.3</b> The human reference genome</a></li>
<li class="chapter" data-level="2.4" data-path="ucsc-genome-browser.html"><a href="ucsc-genome-browser.html"><i class="fa fa-check"></i><b>2.4</b> UCSC genome browser</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="ucsc-genome-browser.html"><a href="ucsc-genome-browser.html#homepage-1"><i class="fa fa-check"></i><b>2.4.1</b> Homepage</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="viewing-one-region-of-the-genome.html"><a href="viewing-one-region-of-the-genome.html"><i class="fa fa-check"></i><b>2.5</b> Viewing one region of the genome</a></li>
<li class="chapter" data-level="2.6" data-path="customizing-browser-displays.html"><a href="customizing-browser-displays.html"><i class="fa fa-check"></i><b>2.6</b> Customizing browser displays</a></li>
<li class="chapter" data-level="2.7" data-path="igv.html"><a href="igv.html"><i class="fa fa-check"></i><b>2.7</b> IGV</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="igv.html"><a href="igv.html#homepage-2"><i class="fa fa-check"></i><b>2.7.1</b> Homepage</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="navigating-igv.html"><a href="navigating-igv.html"><i class="fa fa-check"></i><b>2.8</b> Navigating IGV</a></li>
<li class="chapter" data-level="2.9" data-path="loading-sequencing-data.html"><a href="loading-sequencing-data.html"><i class="fa fa-check"></i><b>2.9</b> Loading sequencing data</a></li>
<li class="chapter" data-level="2.10" data-path="the-1000-genomes-project.html"><a href="the-1000-genomes-project.html"><i class="fa fa-check"></i><b>2.10</b> The 1000 Genomes Project</a></li>
<li class="chapter" data-level="2.11" data-path="sra.html"><a href="sra.html"><i class="fa fa-check"></i><b>2.11</b> SRA</a>
<ul>
<li class="chapter" data-level="2.11.1" data-path="sra.html"><a href="sra.html#previewing-sequencing-data"><i class="fa fa-check"></i><b>2.11.1</b> Previewing sequencing data</a></li>
</ul></li>
<li class="chapter" data-level="2.12" data-path="viewing-sequencing-reads-in-igv.html"><a href="viewing-sequencing-reads-in-igv.html"><i class="fa fa-check"></i><b>2.12</b> Viewing sequencing reads in IGV</a></li>
<li class="chapter" data-level="2.13" data-path="interpreting-igv-alignments.html"><a href="interpreting-igv-alignments.html"><i class="fa fa-check"></i><b>2.13</b> Interpreting IGV alignments</a></li>
<li class="chapter" data-level="2.14" data-path="conclusion.html"><a href="conclusion.html"><i class="fa fa-check"></i><b>2.14</b> Conclusion</a></li>
<li class="chapter" data-level="2.15" data-path="homework-1.html"><a href="homework-1.html"><i class="fa fa-check"></i><b>2.15</b> Homework</a>
<ul>
<li class="chapter" data-level="2.15.1" data-path="homework-1.html"><a href="homework-1.html#required-homework"><i class="fa fa-check"></i><b>2.15.1</b> Required homework</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="discovering-mutations.html"><a href="discovering-mutations.html"><i class="fa fa-check"></i><b>3</b> Discovering mutations</a>
<ul>
<li class="chapter" data-level="3.1" data-path="background.html"><a href="background.html"><i class="fa fa-check"></i><b>3.1</b> Background</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="background.html"><a href="background.html#de-novo-mutations"><i class="fa fa-check"></i><b>3.1.1</b> _De novo) mutations</a></li>
<li class="chapter" data-level="3.1.2" data-path="background.html"><a href="background.html#recombination"><i class="fa fa-check"></i><b>3.1.2</b> Recombination</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="setup.html"><a href="setup.html"><i class="fa fa-check"></i><b>3.2</b> Setup</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="setup.html"><a href="setup.html#using-r-markdown-files"><i class="fa fa-check"></i><b>3.2.1</b> Using R Markdown files</a></li>
<li class="chapter" data-level="3.2.2" data-path="setup.html"><a href="setup.html#r-packages"><i class="fa fa-check"></i><b>3.2.2</b> R packages</a></li>
<li class="chapter" data-level="3.2.3" data-path="setup.html"><a href="setup.html#data"><i class="fa fa-check"></i><b>3.2.3</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="visualizing-the-data.html"><a href="visualizing-the-data.html"><i class="fa fa-check"></i><b>3.3</b> Visualizing the data</a></li>
<li class="chapter" data-level="3.4" data-path="linear-models.html"><a href="linear-models.html"><i class="fa fa-check"></i><b>3.4</b> Linear models</a></li>
<li class="chapter" data-level="3.5" data-path="confidence-intervals.html"><a href="confidence-intervals.html"><i class="fa fa-check"></i><b>3.5</b> Confidence intervals</a></li>
<li class="chapter" data-level="3.6" data-path="conclusion-1.html"><a href="conclusion-1.html"><i class="fa fa-check"></i><b>3.6</b> Conclusion</a></li>
<li class="chapter" data-level="3.7" data-path="homework-2.html"><a href="homework-2.html"><i class="fa fa-check"></i><b>3.7</b> Homework</a>
<ul>
<li class="chapter" data-level="3.7.1" data-path="homework-2.html"><a href="homework-2.html#required"><i class="fa fa-check"></i><b>3.7.1</b> Required</a></li>
<li class="chapter" data-level="3.7.2" data-path="homework-2.html"><a href="homework-2.html#extra-credit-more-linear-models"><i class="fa fa-check"></i><b>3.7.2</b> Extra credit: More linear models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="simulating-evolution.html"><a href="simulating-evolution.html"><i class="fa fa-check"></i><b>4</b> Simulating evolution</a>
<ul>
<li class="chapter" data-level="4.1" data-path="background-1.html"><a href="background-1.html"><i class="fa fa-check"></i><b>4.1</b> Background</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="background-1.html"><a href="background-1.html#genetic-drift"><i class="fa fa-check"></i><b>4.1.1</b> Genetic drift</a></li>
<li class="chapter" data-level="4.1.2" data-path="background-1.html"><a href="background-1.html#the-wright-fisher-model"><i class="fa fa-check"></i><b>4.1.2</b> The Wright-Fisher model</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="modeling-allele-frequencies.html"><a href="modeling-allele-frequencies.html"><i class="fa fa-check"></i><b>4.2</b> Modeling allele frequencies</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="modeling-allele-frequencies.html"><a href="modeling-allele-frequencies.html#the-binomial-distribution"><i class="fa fa-check"></i><b>4.2.1</b> The binomial distribution</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="setup-1.html"><a href="setup-1.html"><i class="fa fa-check"></i><b>4.3</b> Setup</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="setup-1.html"><a href="setup-1.html#r-packages-1"><i class="fa fa-check"></i><b>4.3.1</b> R packages</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="a-wright-fisher-simulation.html"><a href="a-wright-fisher-simulation.html"><i class="fa fa-check"></i><b>4.4</b> A Wright-Fisher simulation</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="a-wright-fisher-simulation.html"><a href="a-wright-fisher-simulation.html#drawing-one-generation-from-a-binomial-distribution"><i class="fa fa-check"></i><b>4.4.1</b> Drawing one generation from a binomial distribution</a></li>
<li class="chapter" data-level="4.4.2" data-path="a-wright-fisher-simulation.html"><a href="a-wright-fisher-simulation.html#simulating-multiple-generations-with-a-for-loop"><i class="fa fa-check"></i><b>4.4.2</b> Simulating multiple generations with a for loop</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>4.5</b> Plotting</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="plotting.html"><a href="plotting.html#visualizing-changes-in-af-over-generations"><i class="fa fa-check"></i><b>4.5.1</b> Visualizing changes in AF over generations</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="creating-a-simulation-function.html"><a href="creating-a-simulation-function.html"><i class="fa fa-check"></i><b>4.6</b> Creating a simulation function</a>
<ul>
<li class="chapter" data-level="4.6.1" data-path="creating-a-simulation-function.html"><a href="creating-a-simulation-function.html#simulating-with-different-parameters-with-a-function"><i class="fa fa-check"></i><b>4.6.1</b> Simulating with different parameters with a function</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="conclusion-2.html"><a href="conclusion-2.html"><i class="fa fa-check"></i><b>4.7</b> Conclusion</a></li>
<li class="chapter" data-level="4.8" data-path="homework-3.html"><a href="homework-3.html"><i class="fa fa-check"></i><b>4.8</b> Homework</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="homework-3.html"><a href="homework-3.html#required-1"><i class="fa fa-check"></i><b>4.8.1</b> Required</a></li>
<li class="chapter" data-level="4.8.2" data-path="homework-3.html"><a href="homework-3.html#extra-credit-selection"><i class="fa fa-check"></i><b>4.8.2</b> Extra credit: Selection</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="about-the-authors.html"><a href="about-the-authors.html"><i class="fa fa-check"></i>About the Authors</a></li>
<li class="divider"></li>
<p style="text-align:center; font-size: 11pt"> <a href="https://github.com/jhudsl/OTTR_Template" target="blank" > Published with</a> <a href="https://bookdown.org/" target="blank"> <b>bookdown</b></a> (by <a href="http://jhudatascience.org/"> The Johns Hopkins Data Science Lab</a>) </p>
<p style="text-align:center; font-size: 12px;"> <a href="https://github.com/rstudio4edu/rstudio4edu-book/"> Style adapted from: rstudio4edu-book </a> <a href ="https://creativecommons.org/licenses/by/2.0/"> (CC-BY 2.0) </a></p>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Human Genome Variation Lab</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="setup" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Setup</h2>
<div id="using-r-markdown-files" class="section level3" number="3.2.1">
<h3><span class="header-section-number">3.2.1</span> Using R Markdown files</h3>
<p>In R Markdown files like these, you can run all the code contained in a “chunk” (like the one you see below) by pressing the green <code>&gt;</code> button on the upper right-hand corner. You can also run individual lines of code by putting your cursor at the end of a line, or selecting multiple lines of code, and hitting <code>cmd + return</code> (Mac) or <code>ctrl + enter</code> (Windows).</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="setup.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="st">&quot;Welcome to the Human Genome Variation lab!&quot;</span>)</span></code></pre></div>
</div>
<div id="r-packages" class="section level3" number="3.2.2">
<h3><span class="header-section-number">3.2.2</span> R packages</h3>
<p>We’ll be using R’s <code>tidyverse</code> library to analyze our data. You can load this R package by running:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="setup.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
</div>
<div id="data" class="section level3" number="3.2.3">
<h3><span class="header-section-number">3.2.3</span> Data</h3>
<p>Our data comes from the paper <a href="https://science.sciencemag.org/content/363/6425/eaau1043">Characterizing mutagenic effects of recombination through a sequence-level genetic map</a> by Halldorsson et al., which performed whole-genome sequencing on “trios” (two parents and one child) in Iceland.</p>
<p>The word “proband”, derived from a Latin word meaning “to be proved/tested”, refers to the offspring in the study (i.e. the test subjects).</p>
<p>We’ve downloaded the <a href="https://science.sciencemag.org/content/363/6425/eaau1043/tab-figures-data">data</a> used in this study into this directory. For the primary analysis in this lab, there are two files of interest to us:</p>
<ol style="list-style-type: decimal">
<li><code>aau1043_DataS5_revision1.tsv</code>: a modified <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">variant call format file</a>, where each line in the file provides information about an observed <em>de novo</em> mutation (DNM)</li>
</ol>
<div class="figure">
<img src="resources/images/dnm/dnm_file.png" style="width:50.0%" alt="" />
<p class="caption">DNM info file</p>
</div>
<ul>
<li>The first column, <code>Chr</code>, lists the chromosome where the DNM is observed.</li>
<li>The second column, <code>Pos</code>, lists the genomic position on that chromosome where the DNM is observed.</li>
<li>The third column, <code>Ref</code>, lists the reference allele (A/C/G/T) when the DNM isn’t present.</li>
<li>The fourth column, <code>Alt</code>, lists the DNM alternate allele (A/C/G/T or a combination of these bases as seen in row 14 above).</li>
<li>The fifth column, <code>Proband_id</code>, lists the proband ID or in which proband the DNM was present.</li>
<li>The sixth column, <code>Phase_combined</code>, provides from which parent the offspring inherited this DNM.</li>
<li>The seventh column, <code>Crossover</code>, specifies whether the DNM is associated with a nearby crossover and from which parent the crossover originated (if able to be determined).</li>
<li>The final column, <code>Sanger</code>, describes whether or not the DNM was confirmed through use of PCR &amp; Sanger sequencing. For our purposes, we will largely be ignoring this information.</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li><code>aau1043_DataS7.tsv</code>: contains information about the paternal and maternal age for each proband at birth</li>
</ol>
<div class="figure">
<img src="resources/images/dnm/age_file.png" style="width:25.0%" alt="" />
<p class="caption">age info file</p>
</div>
<ul>
<li>The first column, <code>Proband_id</code>, matches the proband IDs from the first file and allows for connections between the two sets of data.</li>
<li>The second column, <code>Father_age</code>, lists the paternal age for each proband at the time of birth.</li>
<li>The third column, <code>Mother_age</code>, lists the maternal age for each proband at the time of birth.</li>
</ul>
<!-- ### Data pre-processing -->
<!-- The data from the Science paper has been aggregated and pre-processed for you. If you're interested in the details of how this was done, expand the section below. -->
<!-- *** -->
<!-- <details><summary> Pre-processing the data </summary> -->
<!-- We first read in the two tab-delimited data files from the website URLs, skipping header info. We specify that the files are tab-delimited with the `delim="\t"` arguments and skip the headers by using the `skip` argument). -->
<!-- ```{r, results = FALSE, message = FALSE, warning = FALSE} -->
<!-- library(tidyverse) -->
<!-- # science links are still broken!! -->
<!-- # dnm <- read_delim("https://www.science.org/doi/suppl/10.1126/science.aau1043/suppl_file/aau1043_datas5_revision1.tsv", delim = "\t", skip = 11) -->
<!-- # age <- read_delim("https://www.science.org/doi/suppl/10.1126/science.aau1043/suppl_file/aau1043_datas7.tsv", delim = "\t", skip = 4) -->
<!-- dnm <- read_delim("resources/data/dnm/aau1043_datas5_revision1.tsv", delim = "\t", skip = 11) -->
<!-- age <- read_delim("resources/data/dnm/aau1043_datas7.tsv", delim = "\t", skip = 4) -->
<!-- ``` -->
<!-- We then connect the information in the two files so that we keep all the information together for each proband. We use the `left_join()` function for this, specifically telling it to match the info given the `Proband_id` columns. We then take into account that one proband may have more than one DNM. For this lab, we don't care about a lot of the info for the DNMs (like the genomic location or the mutated sequences). We mostly care that for each proband we know how many DNMs were observed, and specifically, which parent the DNM originated from. We use the `summarize` function to find the number of DNMs that originated from the father for each proband -- `n_paternal_dnm`; the number of DNMs that originated from the mother for each proband -- `n_maternal_dnm`; and the number of DNMs for which it is unknown from which parent it originated -- `n_na_dnm`. Because some probands had more than DNM, when we used `left_join` to connect the two sets of data earlier, it meant that the parental ages for a given proband were repeated each time that proband was listed. So we also use the unique function so that the stored parental ages are just single values rather than a really long list repeating the same value for a given proband. Finally, we've included a column that lists the total number of DNMs each proband has. This column, `total_dnm` won't be in the pre-processed data you'll be reading in later. It's just for visualization purposes.    -->
<!-- ```{r, results = FALSE, message = FALSE, warning = FALSE} -->
<!-- dnm_by_age <- left_join(dnm, age, by="Proband_id") %>%  -->
<!--   group_by(Proband_id) %>%  -->
<!--   summarize(n_paternal_dnm = sum(Phase_combined == "father", na.rm = TRUE), -->
<!--             n_maternal_dnm = sum(Phase_combined == "mother", na.rm = TRUE), -->
<!--             n_na_dnm = sum(is.na(Phase_combined)), -->
<!--             Father_age = unique(Father_age), -->
<!--             Mother_age = unique(Mother_age)) %>%  -->
<!--   mutate(total_dnm = rowSums(.[, c("n_paternal_dnm", "n_maternal_dnm", "n_na_dnm")])) -->
<!-- ``` -->
<!-- Now you may be wondering, how many DNMs does each proband have? We'll use a histogram to visualize this.  -->
<!-- ```{r, results = FALSE, message = FALSE, warning = FALSE} -->
<!-- hist(dnm_by_age$total_dnm, breaks=50, main="Number of DNMs per proband", xlab="number of DNMs", ylab="frequency of probands") -->
<!-- ``` -->
<!-- <!-- ![Number of DNMs per proband](resources/images/dnm/dnms_per_proband.png) -->
<p>–&gt;</p>
<!-- We see that there are about 65 DNMs per proband on average. The distribution also appears to be normally distributed, but it is right-skewed and has a long tail. -->
<!-- We're also interested in how many of these DNMs are unphased. By unphased, we mean that the authors couldn't determine from which parent the DNM originated. In other words, we're interested in the `n_na_dnm` column. When we plot a histogram of this data, we see a bi-modal distribution, where, again, the right section looks like a right-skewed normal distribution. -->
<!-- ```{r, results = FALSE, message = FALSE, warning = FALSE} -->
<!-- hist(dnm_by_age$n_na_dnm, breaks=50, main="Number of unphased DNMs per proband", xlab = "number of unphased DNMs", ylab="frequency of probands") -->
<!-- abline(v = 10, col="blue", lty=2 ) -->
<!-- legend(100, 300, col="blue", lty=2, legend = "x=10") -->
<!-- ``` -->
<!-- <!-- ![Number of unphased DNMs per proband](resources/images/dnm/unphased_dnms_per_proband.png) -->
<p>–&gt;</p>
<!-- On average, 39 of a proband's DNMs are unphased. This is nearly half of the total average. If we're more precise with this, and look at the number of unphased DNMs divided the number of total DNMs for each proband, we see that on average 60% are unphased.  -->
<!-- ```{r, results = FALSE, message = FALSE, warning = FALSE} -->
<!-- mean(dnm_by_age$n_na_dnm/dnm_by_age$total_dnm) -->
<!-- ``` -->
<!-- Since we're interested in the relationship of parental age with the number of DNMs originating from that parent, we want to work with data that is mostly phased, so we decide to filter our data to concentrate on the first section of the plot above. Specifically we keep those probands who have less than 10 unphased DNMs. The below code accomplishes that filtering. -->
<!-- ```{r, results = FALSE, message = FALSE, warning = FALSE} -->
<!-- dnm_by_age <- dnm_by_age %>%  -->
<!--   filter(n_na_dnm < 10) -->
<!-- ``` -->
<!-- Now we see that on average around 3% of DNMs are unphased for each proband, and this is much more manageable for the questions we want to explore. -->
<!-- ```{r, results = FALSE, message = FALSE, warning = FALSE} -->
<!-- mean(dnm_by_age$n_na_dnm/dnm_by_age$total_dnm) -->
<!-- ``` -->
<!-- If we rerun our plot from above for the total number of DNMs per proband, now that we've filtered, we observe that the mean number of total DNMs is 67, relatively unchanged compared to the full dataset. Further, the distribution appears less skewed.  -->
<!-- ```{r, results = FALSE, message = FALSE, warning= FALSE} -->
<!-- hist(dnm_by_age$total_dnm, breaks=50, main="Number of DNMs per proband after filtering", xlab="number of DNMs", ylab="frequency of probands") -->
<!-- ``` -->
<!-- <!-- ![Number of DNMs per proband after filtering](resources/images/dnm/dnms_per_proband_after_filt.png) -->
<p>–&gt;</p>
<!-- Now that we have tidy and filtered data, we drop the `total_dnm` column since we won't need that anymore and we save the data to a `.csv` file (which is a comma delimited file). -->
<!-- ```{r, results = FALSE, message = FALSE, warning = FALSE} -->
<!-- dnm_by_age <- subset(dnm_by_age, select=-c(total_dnm)) -->
<!-- write_csv(dnm_by_age, "dnm_by_age_tidy_Halldorsson.csv") -->
<!-- ``` -->
<!-- </details> -->
<!-- *** -->
<p>We’ve pre-processed the data to make it easier to use, and you can load the pre-processed file using the code chunk below. (See the <code>Using R Markdown files</code> section above if you’re not sure how to run the code chunk.)</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="setup.html#cb3-1" aria-hidden="true" tabindex="-1"></a>dnm_by_age <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;https://www.dropbox.com/s/e6p6nalccjj6j9y/dnm_by_age_tidy_Halldorsson.tsv?dl=1&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-2"><a href="setup.html#cb3-2" aria-hidden="true" tabindex="-1"></a>dnm_by_age</span></code></pre></div>
</div>
</div>
<hr>
<center> 
  <div class="footer">
      All illustrations <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY. </a>
      <br>
      All other materials <a href= "https://creativecommons.org/licenses/by/4.0/"> CC-BY </a> unless noted otherwise.
  </div>
</center>
            </section>

          </div>
        </div>
      </div>
<a href="background.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="visualizing-the-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
