
# Evolutionary Simulations

```{r, include = FALSE}
ottrpal::set_knitr_image_path()
```

## Conclusion

In this lab, we've successfully built a Wright-Fisher simulation for one allele, allowing us to track how we expect its frequency to change over time under the principles of genetic drift.

This simple simulation forms the core of most models used in evolutionary genetics research. For example, the evolutionary simulation software [SLiM](https://messerlab.org/slim/) extends our model to perform Wright-Fisher simulations with multiple alleles simultaneously, allowing us to reconstruct complex phenomena like balancing selection (**Fig. 2**).

<center>

![**Fig. 2. Simulating balancing selection with SLiM.** Each vertical line represents a SNP, with height corresponding to the SNP's current frequency. As the simulation progresses, two distinct haplotypes of mutations begin to rise and fall in frequency together, but neither fully fixes because balancing selection favors intermediate frequencies.](balancing.gif)

</center>

## Homework

#### Goals & Learning Objectives

The goal of this homework is to explore different extensions of the Wright-Fisher model.

**Learning Objectives**

* Required homework: Practice visualizing allele frequencies with `ggplot`
* Extra credit: Practice writing functions and interpreting allele frequency trajectories

## Required

Currently, our allele frequency trajectory plot only shows the AF of one allele at a locus. (i.e., if individuals in a population have either an `A` or a `C` at some locus, we're only plotting the trajectory of the `A` allele.)

**Assignment:** Add a line to allele frequency trajectory plot that shows the frequency of the other allele at the locus. Give the two alleles different colors.

***
<details> <summary> Hint </summary>
You don't need to modify the simulation function for this.
</details>
***

***
<details> <summary> Solution </summary>
```{r}
# run the simulation again to get output data
sim <- run_sim(Ne = 1000, freq = 0.5, generations = 10000) %>%
  # add in a column with the AF of the minor allele
  # this is 1 - AF of the major allele
  dplyr::mutate(afs_minor = 1 - afs)

ggplot() +
  geom_line(data = sim, aes(x = gen, y = afs), color = "blue") +
  # add in another line with the frequency of the minor allele
  geom_line(data = sim, aes(x = gen, y = afs_minor), color = "red") +
  ylim(0, 1) +
  ylab("Allele frequency") +
  xlab("Generation")
```
</details>
***

## Extra credit: Selection

One way to extend our simple Wright-Fisher model is to add in selection as a parameter. Selection affects our model by altering the probability of sampling our allele of interest each generation (e.g., positive selection increases the probability, and negative selection decreases it).

Previously, we assumed that this probability was equivalent to the allele's frequency, or $p = \frac{i}{N_e}$, where $N_e$ is the population size and $i$ is the number of individuals who carry the allele.

For the purposes of this homework, we assume that in a model with selection, this probability is instead:

$$
p = \frac{i(1 + s)}{N_e - i + i(1+s)}
$$

where $s$ is the **selection coefficient**, and ranges from `-1` to `1`.

***
<details> <summary> **Question**: What does this probability become in the absence of selection (i.e., when $s = 0$)? </summary>
The probability becomes $\frac{i}{N_e}$, which is the same as the allele frequency.
</details>
***

**Assignment:** Modify your `run_sim` function so that it takes in a selection coefficient `s` as a parameter. Run the simulation a few times with and without (`s = 0`) selection, but keeping other parameters the same (`Ne = 10000` -- to mimic the `Ne` of humans, `freq = 0.5`, `generations = 10000`). What do you notice about the allele frequency trajectories?

***
<details> <summary> Solution </summary>
```{r}
run_sim_selection <- function(Ne, freq, generations, s) {
  
  freq_vector <- freq
  for (i in 1:generations) {
    # calculate p, the probability of sampling the allele, based on s
    i <- freq * Ne # number of individuals who currently carry the allele
    p <- i*(1+s) / (Ne - i + i*(1+s))
    
    # prob is now `p`, rather than `freq`
    new_freq <- rbinom(n = 1, size = Ne, prob = p) / Ne
    freq_vector <- c(freq_vector, new_freq)
    freq <- new_freq
  }
  
  # convert vector of AFs into a tibble for plotting
  sim_results <- tibble(afs = freq_vector,
                        gen = 1:(generations+1))
  
  # return the tibble of AFs, so that we can access the results
  return(sim_results)
}
```

Run and plot the simulation with selection:

```{r}
results <- run_sim_selection(Ne = 10000,
                             freq = 0.5,
                             generations = 10000,
                             s = -0.1)
ggplot() +
  geom_line(data = results, aes(x = gen, y = afs)) +
  ylim(0, 1) +
  ylab("Allele frequency") +
  xlab("Generation") +
  ggtitle("Simulation with selection") +
  theme(plot.title = element_text(hjust = 0.5)) # to center the title
```

Run and plot the simulation without selection:

```{r}
results <- run_sim_selection(Ne = 10000,
                             freq = 0.5,
                             generations = 10000,
                             s = 0)
ggplot() +
  geom_line(data = results, aes(x = gen, y = afs)) +
  ylim(0, 1) +
  ylab("Allele frequency") +
  xlab("Generation") +
  ggtitle("Simulation without selection") +
  theme(plot.title = element_text(hjust = 0.5)) # to center the title
```

We observe that selection (at least, strong selection where `s = 0.1`) tends to decrease the time it takes for an allele to either fix or go extinct. This is because selection directionally biases the probability of sampling that allele.

Decreasing the absolute value of the selection coefficient will make the simulation behave more like drift - most selection coefficients are thought to be [very small](https://academic.oup.com/view-large/figure/326984394/345fig4.jpeg), and the largest known selection coefficients in humans are around [0.05](https://elifesciences.org/articles/63177).
</details>
***