
```{r, include = FALSE}
ottrpal::set_knitr_image_path()
knitr::opts_knit$set(root.dir = '06-gwas1')
```

# Genome-wide association studies I

In this lab, we'll introduce and discuss the limitations of **genome-wide association studies (GWAS)**.

#### Learning objectives

After completing this chapter, you'll be able to:

1. Explain the statistical method that underlies GWAS.
2. Describe the statistical challenges of GWAS.
3. Understand how linkage disequilibrium helps and hinders GWAS.
4. Interpret common GWAS plots and summary statistics.


## Association studies

One of the central goals of human genetics is understanding the relationship between genotype and phenotype. **Genome-wide association studies (GWAS)** emerged ~20 years ago as a useful approach for discovering genetic variation that underlies variation in human traits.

To conduct GWAS, you:

* Go through every variant in the genome
* Ask if its allele frequency differs between individuals who have or don't have a phenotype of interest
  * For most variants (**Fig. 1**, SNP 1), there will be very little difference
* Identify the variants with the largest association between genotype and phenotype (**Fig. 1**, SNP 2)
<br></br>

<center>

![**Fig. 1.** In this GWAS cohort, SNP 2 is significantly associated with the phenotype.](05-gwas1/images/gwas.jpg){width=75%}

</center>

The phenotype can be any measurable trait, whether it's binary (ex: if someone has a disease) or continuous (ex: height).


## GWAS is just linear regression

At their core, GWAS involve fitting **linear models** to test for relationships between variants and phenotypes using data from large samples of individuals.

As with the linear models we covered in the [DNM module](https://mccoy-lab.github.io/hgv_modules/discovering-mutations.html), GWAS fits a line to a set of points. In this case, each point is one individual in the dataset, stratified by their genotype for a variant of interest.
<br></br>

<center>

![**Fig. 2 ([source](https://www.nature.com/articles/nrg1916)).** GWAS fits a linear model for every variant, where the x axis is genotype and the y axis is a phenotype.](05-gwas1/images/regression.jpg){width=75%}

</center>

Because there are so many variants in the genome and we perform a separate statistical test for each one, we often end up fitting millions of linear models for a GWAS.


## Multiple testing

***
<details> <summary> What are statistical challenges of performing a test multiple times? </summary>

When we perform any test multiple times, we increase the risk that a "significant" result is only significant by chance.

Under the null hypothesis, we assume that p-values follow a **uniform distribution** (i.e., a flat distribution from 0 to 1). We can plot this null distribution in R:

```{r}
# generate 1,000,000 "p-values" from a uniform distribution
pvalues <- runif(1000000)
# histogram with R's base plotting function
hist(pvalues)
```


If we use the typical **p-value threshold** of $0.05$, 5% of our tests will have $p < 0.05$, even though these p-values were simulated from a null distribution (i.e., no real association).

</details>
***

***
<details> <summary> How do we correct for multiple testing? </summary>

One common multiple testing correction method, **Bonferroni correction**, sets a stricter p-value threshold. With Bonferroni, you divide your desired p-value by the number of independent tests you conducted.

</details>
***

***
<details> <summary> Are GWAS tests (variants) statistically independent? How does this affect our p-value threshold? </summary>

As we learned in the [LD module](https://mccoy-lab.github.io/hgv_modules/linkage-disequilibrium.html), the genotypes of nearby variants are correlated.

This non-independence means that we can be _less_ strict with multiple testing correction, because we aren't performing as many _independent_ tests as we think we are.

Researchers have calculated that $\mathbf{5*10^{-8}}$ is an appropriate p-value threshold for GWAS in humans, given the amount of LD in human genomes.

</details>
***


## LD and GWAS

LD is both a blessing and a curse for GWAS. On one hand, LD means that we need not genotype every SNP to discover associations. We merely need to genotype **"tag SNPs"**, which are in LD with variants that causally influence the phenotype.

On the other hand, this also means that even when we find a signficant association, it is often challenging to disentangle the **causal** gene and/or variant that drives the association from all the other variants on the same haplotype.

<center>

![**Fig. 3 ([source](https://www.nature.com/articles/nrg1916)).** Even without finding the causal variant, we can still discover the causal haplotype through genotyping variants in LD .](05-gwas1/images/ld_gwas.jpg){width=75%}

</center>


## Imputation

LD also means that we can perform **imputation** to improve our GWAS discovery power. If we know which variants tend to occur together on the same haplotype, we can infer the presence of variants in an individual even if those SNPs were never sequenced.

In imputation, you:

* Genotype individuals using the sequencing data you have (**Fig. 4A**)
* Use a reference panel of haplotypes to fill in variants you didn't genotype (**Fig. 4B**)
* Perform GWAS

Imputation can discover GWAS signals that were hidden in the original variant set (**Fig. 4B** vs. **Fig. 4F**), or further support signals you already found.
<br></br>

<center>

![**Fig. 4 ([source](https://www.nature.com/articles/nrg1916)).** Imputing variants that were not actually sequenced.](05-gwas1/images/imputation.jpg)

</center>


## QQ plots

One common visualization for GWAS results is a **QQ plot**, which compares the distribution of p-values in our results to a null distribution (i.e., the uniform distribution that we plotted earlier).

***
<details> <summary> How do you make a QQ plot? </summary>

* Generate simulated p-values from a uniform distribution -- the number of simulated p-values should equal the number of actual p-values
* Sort both your real and simulated p-values in descending order
* Plot the first, second, third, etc. p-values, where
  * x-axis is the simulated value
  * y-axis is the actual value

</details>
***

<center>

![**Fig. 5 ([source](https://www.nature.com/articles/nrg1916)).** QQ plots visualize the distriution of p-values compared to a null distribution.](05-gwas1/images/qqplot.jpg){width=60%}

</center>

There are three areas of this plot where points can fall:

1. **On the $\mathbf{x = y}$ line**: No association signal
2. **Above the $\mathbf{x = y}$ line**: Some association signal
3. **Below the $\mathbf{x = y}$ line**: Issue with our statistical test (ex: not appropriately adjusting for covariates)


## Manhattan plots

**Manhattan plots** show the distribution of GWAS hits across the genome, where the y-axis is p-value.
<br></br.

<center>

![**Fig. 6.** Manhattan plot of GWAS results. The red line is the $5*10^{-8}$ genome-wide significance threshold.](05-gwas1/images/manhattanplot.jpg){width=75%}

</center>

***
<details> <summary> Why are there peaks in the Manhattan plot? </summary>

Each peak is composed of variants from the same haplotype, which all have a strong association with the phenotype because of LD.

</details>
***


## Sample size

As GWAS sample size has increased over the years, each larger study is able to discover more variants.

<center>

![**Fig. 7 ([source](https://www.sciencedirect.com/science/article/pii/S0002929711005337)).** Increase in GWAS discovery power with sample size.](05-gwas1/images/sample_size.jpg){width=75%}

</center>

***
<details> <summary> Why do we find more associations with larger studies? </summary>

1. A larger dataset captures more **rare variation**
2. A larger dataset provides more **statistical power** -- we can more confidently say whether allele frequencies are different between individuals with and without a phenotype

</details>
***


## Interpreting GWAS results

Let's look at a [browser of GWAS results](https://pheweb.org/UKB-Neale/), generated by a study that used genotype and phenotype data from the **UK Biobank**.

***
<details> <summary> UK Biobank </summary>

[**UK Biobank**](https://www.nature.com/collections/bpthhnywqk) is one of the largest cohorts of genotype and phenotype data available.

This study has data from ~500,000 individuals across the UK, including health records and responses to surveys, all of which are publicly available for research.

</details>
***

***
<details> <summary> As an example, let's look up associations with standing height. </summary>

<center>

![**Fig. 8.** GWAS results for standing height.](05-gwas1/images/height_gwas.png)

</center>

* For each phenotype, the browser shows a Manhattan plot of associated variants
* The top p-values for this trait look extremely significant
* Bottom of the page also has a QQ plot

Although these p-values are huge, keep in mind that for the vast majority of extremely strong associations, the total proportion of variance explained is very small (< 10%). (i.e., most of the natural variation in the phenotype is not explained by that variant.)

</details>
***


## Conclusion

In this lab, we gave a conceptual overview of genome-wide association studies.

* GWAS are an attempt to answer one of the oldest questions in genetics: How does genotype impact **phenotype**?
<br></br>
* In GWAS, you go through every variant in the genome and fit a **linear model** to ask if genotype at that variant is associated with a phenotype of interest. **Multiple testing correction** accounts for the statistical burden of these tests.
<br></br>
* **Linkage disequilibrium** helps us discover more GWAS hits, but also limits our ability to identify causal variants/genes.
<br></br>
* **QQ plots** and **Manhattan plots** are common visualizations of GWAS results.


## Homework

#### Learning Objectives

* Interpret the summary statistics typically reported in GWAS studies
* Understand the limitations and biases of GWAS

#### Assignment

Find any GWAS paper. Read it and report:

* Phenotype being studied
* Sample size
* Population being studied (homogeneous? Multi-ethnic? If it's multi-ethnic, how do they correct for the effect of ancestry?)
* For the top asssociation:
  * p-value (would you call it significant genome wide?)
  * Effect size
  * Did the authors replicate the result in an independent cohort?
  * Haplotype structure, nearby genes, causal variant

***
<details> <summary> Example solution </summary>

Example GWAS: [Genome-wide analysis identifies genetic effects on reproductive success and ongoing natural selection at the FADS locus](https://www.nature.com/articles/s41562-023-01528-6).

* **Phenotype:** Number of children ever born
* **Sample size:**: 785,604
* **Population**: European; no population structure correction (but did control for family structure)
* **Top asssociation:** rs201815280, `chr3:85546181`, `A`->`ACACC` (from Supp. Table 7)
  * **p-value:** `5.25e-26`, seems genome-wide significant
  * **Effect size:** `0.0249` (with every copy of the `ACACC` allele, individuals have one more child born)
  * **Replication:** _Did not_ replicate in a cohort of 34,367 women from the FinnGen study (p = `0.177`)
  * **Haplotype structure, nearby genes, causal variant:** In an intron of CADM2; no causal variant or LD analysis

</details>
***