# Discovering mutations

In this module, we'll use DNA sequencing data from human families to explore the relationship between parental age and *de novo* mutations in their children.

#### Learning objectives

After completing this chapter, you'll be able to:

1. Load and explore tabular data.
2. Create plots to visualize relationships between variables.
3. Compare the impact of maternal vs. paternal age on _de novo_ mutation counts.
3. Estimate how many mutations accumulate each year in maternal and paternal gametes.


***


## Setup

### Using R Markdown files

In R Markdown files like these, you can run all the code contained in a "chunk" (like the one you see below) by pressing the green `>` button on the upper right-hand corner. You can also run individual lines of code by putting your cursor at the end of a line, or selecting multiple lines of code, and hitting `cmd + return` (Mac) or `ctrl + enter` (Windows).

```{r}
message("Welcome to the Human Genome Variation lab!")
```


### R packages

We're using R's `tidyverse` library to analyze our data. You can load this R package by running:

```{r}
library(tidyverse)
```


### Data

Load the pre-processed data by running the code chunk below.

```{r}
dnm_by_age <- read.table("dnm_by_age_tidy_Halldorsson.tsv", sep = "\t", header = TRUE)
dnm_by_age
```


***


## Visualizing the data

How does parental age affect the number of DNMs for males and females? Use the `dnm_by_age` data to plot this relationship for _males_:

```{r}
ggplot(data = dnm_by_age,
       # specify where ggplot should be getting the x location for each data point
       aes(x = Father_age,
           # specify where ggplot should be getting the y location for each data point
           y = n_paternal_dnm)) +
  # specify that the data should be plotted as points
  geom_point()
```


In the code block below, modify your code to visualize the relationship between age and number of DNMs for _females_.

```{r}
ggplot(data = __________,
       aes(x = __________,
           y = __________)) +
  geom_point()
```


***


## Linear models

We'll fit a linear model using R's `lm` function. Run the following code block to open a manual describing the function.

```{r}
?lm
```

Run the following code block to fit the model and print a summary of its results:

```{r}
# fit linear model for paternal DNMs
fit_pat <- lm(formula = n_paternal_dnm ~ Father_age,
              data = dnm_by_age)

# print results of model
summary(fit_pat)
```

In the code block below, modify the model fitting code to assess the relationship between _maternal_ age and number of _maternal_ DNMs.

```{r}
fit_mat <- lm(formula = __________ ~ __________,
              data = __________)
summary()
```


***


## Confidence intervals

Do sperm and oocytes accumulate DNMs at different rates? The maternal and paternal coefficients look different, but we need statistical evidence that they actually are. To do this, we compare the **confidence intervals** of our coefficient estimates.

```{r}
?confint
```

Run the code block below to get the 95% CI for the paternal DNM coefficient.

```{r}
confint(fit_pat, 'Father_age', level = 0.95)
```

In the code block below, modify your code to get the 95% CI for the _`Mother_age`_ slope.

```{r}
confint(__________, __________, level = __________)
```


***


## Homework

So far, we've only looked at the _de novo_ mutation data from [the Halldorsson et al. paper](https://science.sciencemag.org/content/363/6425/eaau1043). Now we'll use their data on the number of maternal and paternal origin crossovers (i.e., how many crossovers occurred across all chromosomes in the maternal and paternal gametes).

#### Learning Objectives

* Practice visualizing data with `ggplot2`
* Interpret p-values and effect sizes from linear models


## Required homework

The data from the paper has been pre-filtered for you. Run this code block to read it in:

```{r}
# read data
crossovers <- read.table("crossovers.tsv", header = TRUE)

# preview data
head(crossovers)
```

The columns in this table are:

1. `Proband_id`: ID of the child
2. `n_pat_xover`: Number of crossovers (carried by the child) that occurred in the paternal gametes
3. `n_mat_xover`: Number of crossovers that occurred in the maternal gametes
4. `Father_age`: Father's age at proband's birth
5. `Mother_age`: Mother's age at proband's birth

**Assignment:** Using the `ggplot` code from this module, plot the relationship between parental age and number of crossovers. As with the DNM data, make one plot for the maternal crossovers and one plot for the paternal. Do you think parental age impacts crossover number?

```{r}
# paternal crossovers plot
```


```{r}
# maternal crossovers plot
```


***


## Optional homework

**Assignment:** Fit _two_ linear models (one paternal, one maternal) to ask if there is an association between the number of parental crossovers and parental age. If there is an association, how is the number of crossovers predicted to change with every year of maternal/paternal age?

```{r}
# paternal model
```


```{r}
# maternal model
```