
```{r, include = FALSE}
ottrpal::set_knitr_image_path()
knitr::opts_knit$set(root.dir = '06-gwas2')
# install packages for the lab
# install.packages('qqman')
```

# Genome-wide association studies II

In this lab, we'll perform a GWAS with data adapted from a [workshop created by Heather Wheeler]((https://github.com/hwheeler01/GWAS_workshop) from Loyola University Chicago.

#### Learning objectives

After completing this chapter, you'll be able to:

1. Perform GWAS "by hand" using linear regression.
2. Use PLINK to perform GWAS on all SNPs in a VCF.
3. Create and interpret common GWAS visualization plots.


## Setup

The premise for this exercise is that you're part of a company developing a cancer drug called **GS451**. Some individuals in early phase trials have experienced a side effect of the drug called lymphocytopenia (low lymphocyte counts).

You are now tasked with performing a GWAS on data from lymphblastoid cell lines to search for risk factors for lymphocytopenia. The phenotype that was measured is the $\mathbf{IC_{50}}$, defined as the concentration of the drug at which the cells experience 50% viability.

### R packages

We'll use the `tidyverse` and `vcfR` packages we've seen in previous weeks, as well as `qqman`, which includes functions for creating Manhattan and QQ plots.

```{r, results = FALSE, message = FALSE, warning = FALSE}
library(tidyverse)
library(vcfR)
library(qqman)
```

### Data

GWAS requires information on both the **genotype** and a **phenotype** of interest in the same individuals.

The genotype data we're using are real data from the Yoruba population in the [1000 Genomes Project](https://mccoy-lab.github.io/hgv_modules/the-1000-genomes-project.html), but the phenotype data is **simulated**.

***
<details> <summary> Why can't we use real phenotype data? </summary>

The combination of genotype and phenotype data poses a privacy risk, so real genotype and phenotype data are often stored in controlled-access databases such as [**dbGaP**](https://www.ncbi.nlm.nih.gov/gap/).

Although these data are still available to researchers who want to work with it, access usually requires submitting an application to explain what your intend to do with it.

</details>
***


## Genotype data

As in the population structure module, our genotype data is stored in a [**Variant Call Format (VCF)**](https://mccoy-lab.github.io/hgv_modules/variant-call-format-vcf.html) file.

```{r, results = FALSE, eval = FALSE}
# load the VCF
vcf <- read.vcfR("genotypes.vcf.gz")
```

```{r, echo = FALSE, results = FALSE, warning = FALSE, message = FALSE}
vcf <- read.vcfR("https://www.dropbox.com/s/gfjf4klkekqglb3/genotypes.vcf.gz?dl=1")
```

We'll first extract just the first SNP in the dataset, using the `vcfR2tidy` function to convert the `vcf` object into tidy dataframes.

```{r}
# extract first SNP and convert to tidy df
test_snp <- vcfR2tidy(vcf[1, ])
```

The `gt` dataframe within `test_snp` contains the genotypes of our SNP of interest. In this dataframe, every row is a different individual in the dataset.

```{r}
# extract genotype dataframe
test_snp_gt <- test_snp$gt

head(test_snp_gt)
```


## Counting allele dosage

We're often interested in encoding genotypes as a 0, 1, or 2, which you can think of as the "**dosage**" of the minor allele. This is an _additive_ model, and assumes that the phenotype of the heterozygote is intermediate between those of the two homozygotes.

We can use the `table` function on the `gt_GT_alleles` column to quickly check how many individuals have each genotype.

```{r}
# tabulate genotype counts
table(test_snp_gt$gt_GT_alleles)
```

Now we'll use the `mutate` function to create a new column of the dataframe that counts the dosage of the minor allele (i.e., how many G's each person has at that SNP):

```{r}
# convert genotypes to counts (i.e., dosage) of minor allele
test_snp_gt <- test_snp_gt %>%
  # count number of Gs
  mutate(dosage = str_count(gt_GT_alleles, "G")) %>%
  drop_na()

head(test_snp_gt)
```

***
<details><summary> Checking our work with `table` </summary>

If we run `table` on the `dosage` column, we should get the same breakdown of genotypes as we got from the `gt_GT_alleles` columns.

```{r}
# make sure we get the same genotype counts
table(test_snp_gt$dosage)
```

</details>
***


## Phenotype data

Our phenotype for this GWAS is the $\mathbf{IC_{50}}$ -- the concentration of the GS451 drug that at which we observe 50% viability in cell culture.

```{r}
# read in phenotypes
phenotypes <- read.table("GS451_IC50.txt", header = TRUE)

head(phenotypes)
```

The columns of this table are:

* **`FID`** & **`IID`**: Family and individual IDs of the individual
* **`GS451_IC50`**: Measured $\mathrm{IC_{50}}$ for the drug of interest

***
<details><summary> Plot the distribution of the phenotype. </summary>

```{r}
ggplot(data = phenotypes,
       aes(x = GS451_IC50)) +
  geom_histogram()
```

This data looks approximately normally distributed. This is important to check because this is one of the assumptions of linear regression, which we'll be using to perform the GWAS.

</details>
***


## Merging genotype and phenotype data

To perform a GWAS, we need to combine genotype and phenotype information for the same individuals. This means **merging** these two data for our SNP of interest.

***
<details><summary> Merging requires the sample ID to be identical between the two dataframes. </summary>

Note that in `test_snp_gt`, the sample ID is in the format `FID_IID` (i.e., `1001_1001`), whereas in `phenotypes` the FID and IID are separate. In order to merge, we create a new column in `phenotypes` that combines them:

```{r}
phenotypes <- phenotypes %>%
  # create column called `Indiv` that combines FID and IID
  mutate(Indiv = paste(FID, IID, sep = "_"))

head(phenotypes)
```

</details>
***

Now we can use `merge` to join the data:

```{r}
# merge genotype and phenotype info for test SNP
gwas_data <- merge(test_snp_gt, phenotypes,
                   by = "Indiv")

head(gwas_data)
```


## GWAS for one variant

Under the hood, GWAS is just linear regression -- simple statistical models to assess evidence of a relationship between two variables. We can perform this linear regression by hand, using data from the first SNP in the VCF.

In our model, we'll be asking whether there's a relationship between an individual's **genotype** (their dosage of the SNP) and **phenotype** (their $\mathrm{IC_{50}}$ for GS451).

***
<details><summary> Why did we merge our genotype and phenotype data? </summary>

When we fit linear models in the [DNM module](https://mccoy-lab.github.io/hgv_modules/fitting-a-linear-model-for-dnms.html), we needed our variables (age and # of DNMs) to be separate columns of the same table.

Similarly, now that our variables are genotype and phenotype, they need to be in the same dataframe.

</details>
***


## Genotype-phenotype boxplots

First, let's plot the relationship between genotype and phenotype to see if it looks interesting.

***
<details><summary> Create boxplots of the phenotype, stratified by genotype of the test SNP. </summary>

```{r}
ggplot(data = gwas_data) +
  geom_boxplot(aes(x = gt_GT_alleles,
                   y = GS451_IC50))
```

It's unclear whether there's a relationship here, because the phenotype distributions for these two genotypes are mostly overlapping. To be certain, we'll now test this with linear regression.

</details>
***


## Linear regression

The function to perform linear regression in R is `lm()`. It takes as arguments a data frame (`gwas_data`) and a model formula of the form `outcome ~ predictors`.

In the case of GWAS, our outcome is the phenotype, and our predictor is the SNP genotype. We may also include covariates such as sex, age, or ancestry as additional predictors (called **covariates**) to control for their potential confounding effects. No such data are available here, so we just run the simple genotype vs. phenotype test.

```{r}
# test for association between genotype and phenotype
lm(data = gwas_data,
   formula = GS451_IC50 ~ dosage) %>%
  # directly pipe (%>%) model results to the `summary()` function
  summary()
```

***
<details><summary> How do we interpret the results of the linear model? </summary>

The **coefficient** for `dosage` indicates that on average, each copy of the "G" allele increases $\mathrm{IC_{50}}$ by $1.38$.

The **p-value** indicates that this slope of $1.38$ is significantly greater than 0 ($p = 0.0449$).

</details>
***

***
<details><summary> Do you think this SNP would reach genome-wide significance? </summary>

This p-value is borderline, sitting very close to the arbitrary cutoff of $0.05$ which is generally used to determine statistical significance.

If this was the only SNP that we were investigating, we might find this result promising. However, this SNP is just one of hundreds of thousands of SNPs that we will test for association, so the burden of proof will need to be much higher. Recall that the genome-wide significance threshold for GWAS in humans is $5 * 10^{-8}$.

</details>
***


## GWAS for multiple SNPs

A GWAS performs the linear regression we just did for every SNP in the dataset. We _could_ write a for loop to do this in R ourselves, but it would be a bit slow because there are 256,896 variants in the VCF.

Because GWAS is such a common approach, researchers have developed software to standardize this process and make it extremely efficient. The most popular software package for GWAS is called **`PLINK`**, which is preloaded into your Cloud session.

PLINK is a "command-line" tool, so we could either use it by working from the "Terminal" tab in Posit Cloud, or using the `system()` command within R. For this class we'll use the latter approach.

***
<details><summary> What's the `system()` command? </summary>

The command line is a text interface that takes in commands for your computer's operating system to run. RStudio and Posit Cloud are a more interactive interface for writing code that you'd normally have to run on the command line.

The `system()` command tells RStudio to run a snippet of command line code for you, without you having to leave the R environment.

</details>
***


## GWAS of one SNP with PLINK

First, we'll replicate the GWAS that we did in R with just the first SNP of the VCF, `rs9699599`.

```{r, eval = FALSE}
# replicate first SNP association with PLINK
system(command = "./plink --file genotypes --linear --snp rs9699599 --pheno GS451_IC50.txt --pheno-name GS451_IC50")
```

***
<details><summary> Breakdown of the PLINK command </summary>

* `./plink`: Use the PLINK software
* `--file genotypes`: Genotype data files begins with the string "genotypes"
* `--linear`:Run a linear additive association test for each SNP
* `--snp rs9699599`: Only run the analysis for a single SNP (`rs9699599`)
* `--pheno GS451_IC50.txt`: Phenotype data is located in a file called `GS451_IC50.txt`
* `--pheno-name GS451_IC50`: The column heading of the phenotype to use in the phenotype file is `GS451_IC50`

</details>
***

After running PLINK, we get an output file called `plink.assoc.linear`.

Now look at the output of the `plink.assoc.linear` output file that PLINK produced.

```{r, eval = FALSE}
snp1 <- read.table("plink.assoc.linear", header = TRUE)

head(snp1)
```

```{r, echo = FALSE, warning = FALSE, message = FALSE}
snp1 <- read.table("https://www.dropbox.com/s/y2s2vc48ppe5jls/plink.assoc.linear.snp1?dl=1", header = TRUE)
head(snp1)
```

***
<details><summary> How do these results compare to performing the GWAS by hand? </summary>

Notice that the **beta** (i.e., the "slope" or coefficient) and **p-value** perfectly matches the results we obtained previously with R.

</details>
***


## GWAS of all SNPs with PLINK

Now let's allow PLINK to run the statistical tests for all SNPs by removing the `--snp` flag.

```{r, eval = FALSE}
system(command = "./plink --file genotypes --linear --pheno GS451_IC50.txt --pheno-name GS451_IC50")
```

The `plink.assoc.linear` file should now have ~260,000 lines. Load the file into R to look at the results:

```{r, eval = FALSE}
results <- read.table(file = "plink.assoc.linear", header = TRUE) %>%
  # create `index` column to store the SNP number before we reorder
  mutate(index = row_number()) %>%
  # order table by lowest pvalue
  arrange(P)

head(results)
```

```{r, echo = FALSE, warning = FALSE, message = FALSE}
results <- read.table(file = "https://www.dropbox.com/s/ukmczgmxhr29aa7/plink.assoc.linear?dl=1", header = TRUE) %>%
  # create `index` column to store the SNP number before we reorder
  mutate(index = row_number()) %>%
  # order table by lowest pvalue
  arrange(P)
head(results)
```


## Plotting GWAS results

The `qq()` and `manhattan()` functions in the `qqman` package let us easily create QQ and Manhattan plots to visualize our GWAS results.

```{r}
# qq plot using the P (pvalues) column
qq(results$P)
```

```{r}
# manhattan plot
manhattan(results)
```

SNPs with low p-values occur in peaks of multiple variants. These are not independent associations, but rather groups of variants in LD.


## Top GWAS SNP

One common future direction for GWAS studies is following up on the top SNP(s). We can plot boxplots of the top SNP genotype stratified by phenotype:

```{r}
# extract genotypes of the top SNP
top_snp <- vcfR2tidy(vcf[240916, ])
top_snp_gt <- top_snp$gt %>%
  drop_na()

# merge with phenotype data
gwas_data <- merge(top_snp_gt, phenotypes,
                   by = "Indiv")

# plot boxplots
ggplot(data = gwas_data) +
  geom_boxplot(aes(x = gt_GT_alleles,
                   y = GS451_IC50))
```

Other potential follow-up directions include:

* Investigating the genomic environment in the [UCSC Genome Browser](https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38)
* Looking at nearby haplotype structure with [LDproxy](https://ldlink.nci.nih.gov/?tab=ldproxy)
  * Note that the genotype data we're using come from the Yoruba population
* Using the [Geography of Genetic Variants](https://popgen.uchicago.edu/ggv/) browser to find the global allele frequencies of the variant
* Search for SNP in a phenotype database to see if there are other associations with it


## Conclusion

We used genotype and simulated phenotype data from the 1000 Genomes Project to perform a **genome-wide association study** for variants associated with drug $\mathrm{IC_{50}}$.

* Using **linear regression**, we first did GWAS "by hand" on just one variant in the VCF. We fit a linear model to ask whether there's a significant relationship between genotype and phenotype.
<br></br>
* We then used **PLINK** to perform this test on every SNP in the genome.
<br></br>
* We followed up on the top SNP from our GWAS by plotting boxplots of phenotype stratified by genotype.


## Homework

#### Learning Objectives

* Interpret results of a GWAS
* Practice manipulating `vcfR` and tabular data

#### Assignment

Run a GWAS of $\mathrm{IC_{50}}$ for the drug CB1908, using the same genotype data as before. The phenotypes are located in `CB1908_IC50.txt`.

Make a QQ plot and a Manhattan plot of your results. Do you have any genome-wide significant hits? Are they located in or near a gene? For the top GWAS hit, plot the phenotype stratified by genotype.

***
<details> <summary> Solution </summary>

```{r, eval = FALSE}
# perform association test with PLINK
system(command = "./plink --file genotypes --linear --pheno CB1908_IC50.txt --pheno-name CB1908_IC50")
```

```{r, eval = FALSE}
# read in gwas results
results <- read.table(file = "plink.assoc.linear", header = TRUE) %>%
  mutate(index = row_number()) %>%
  arrange(P)

# qq plot
qq(results$P)

# manhattan plot
manhattan(results)
```

```{r, echo = FALSE, warning = FALSE, message = FALSE}
# read in gwas results
results <- read.table(file = "https://www.dropbox.com/s/g9ldm9kuknbae3l/plink.assoc.linear.hw?dl=1", header = TRUE) %>%
  mutate(index = row_number()) %>%
  arrange(P)

# qq plot
qq(results$P)

# manhattan plot
manhattan(results)
```

On the Manhattan plot, there's one hit that reaches genome-wide significance.

```{r}
# view top GWAS hit
results[1, ]
```

From looking it up in the UCSC Genome Browser, `rs10876043` lies within an intron of the _DIP2B_ gene. Finally, we plot this SNP's genotype stratified by phenotype. (We can extract its genotype information from the `vcf` object because we know its index number from the `results` table.)

```{r}
# extract top SNP and convert to tidy df
top_snp <- vcfR2tidy(vcf[184405, ])
# get genotype dataframe
top_snp_gt <- top_snp$gt %>%
  drop_na()

# read in phenotype dataframe
phenotypes <- read.table("CB1908_IC50.txt", header = TRUE) %>%
  # create column called `Indiv` that combines FID and IID
  mutate(Indiv = paste(FID, IID, sep = "_"))

# merge genotype and phenotype info
gwas_data <- merge(top_snp_gt, phenotypes,
                   by = "Indiv")

# plot genotype by phenotype boxplots
ggplot(data = gwas_data) +
  geom_boxplot(aes(x = gt_GT_alleles,
                   y = CB1908_IC50))
```
