# Genome-wide association studies II

In this lab, we'll perform a GWAS with data adapted from a [workshop created by Heather Wheeler]((https://github.com/hwheeler01/GWAS_workshop) from Loyola University Chicago.

#### Learning objectives

After completing this chapter, you'll be able to:

1. Perform GWAS "by hand" using linear regression.
2. Use PLINK to perform GWAS on all SNPs in a VCF.
3. Create and interpret common GWAS visualization plots.


***


## Setup

```{r}
library(tidyverse)
library(vcfR)
library(qqman)
```


***


## Genotype data

```{r}
# load the VCF
vcf <- read.vcfR("genotypes_subset.vcf")

# extract first SNP and convert to tidy df
test_snp <- vcfR2tidy(vcf[1, ])

# extract genotype dataframe
test_snp_gt <- test_snp$gt
head(test_snp_gt)
```


***


## Counting allele dosage

```{r}
# tabulate genotype counts
table(test_snp_gt$gt_GT_alleles)

# convert genotypes to counts (i.e., dosage) of minor allele
test_snp_gt <- test_snp_gt %>%
  # count number of Gs
  mutate(dosage = str_count(gt_GT_alleles, "G")) %>%
  drop_na()
head(test_snp_gt)
```

```{r}
# check genotype counts with `table`
table(test_snp_gt$dosage)
```


***


## Phenotype data

```{r}
# read in phenotypes
phenotypes <- read.table("GS451_IC50.txt", header = TRUE)
head(phenotypes)
```

Plot the distribution of phenotypes:

```{r}

```


***


## Merging genotype and phenotype data

```{r}
phenotypes <- phenotypes %>%
  # create column called `Indiv` that combines FID and IID
  mutate(Indiv = paste(FID, IID, sep = "_"))
head(phenotypes)
```

```{r}
# merge genotype and phenotype info for test SNP
gwas_data <- merge(test_snp_gt, phenotypes,
                   by = "Indiv")
head(gwas_data)
```


***


## Genotype-phenotype boxplots

Create boxplots of the phenotype, stratified by genotype of the test SNP:

```{r}

```


***


## Linear regression

```{r}
# test for association between genotype and phenotype
lm(data = gwas_data,
   formula = GS451_IC50 ~ dosage) %>%
  # directly pipe (%>%) model results to the `summary()` function
  summary()
```


***


## GWAS with PLINK

Re-run GWAS on just the first SNP:

```{r}
system(command = "./plink --file genotypes --linear --snp rs9699599 --pheno GS451_IC50.txt --pheno-name GS451_IC50")
```

Read in `plink.assoc.linear` file:

```{r}
snp1 <- read.table("plink.assoc.linear", header = TRUE)
head(snp1)
```


GWAS on all SNPs in VCF:

```{r}
system(command = "./plink --file genotypes --linear --pheno GS451_IC50.txt --pheno-name GS451_IC50")

# read in `plink.assoc.linear` file
results <- read.table(file = "plink.assoc.linear", header = TRUE) %>%
  # order table by lowest pvalue
  arrange(P)
head(results)
```


***


## Plotting GWAS results

```{r}
# qq plot
qq(results$P)

# manhattan plot
manhattan(results)
```


***


## Top GWAS association

Plot top SNP phenotypes stratified by genotype:

```{r}
# extract genotypes of the top SNP
top_snp <- vcfR2tidy(read.vcfR("top_snp.vcf"))
top_snp_gt <- top_snp$gt %>%
  drop_na()

# merge with phenotype data
gwas_data <- merge(top_snp_gt, phenotypes,
                   by = "Indiv")

# plot boxplots
ggplot(data = gwas_data) +
  geom_boxplot(aes(x = gt_GT_alleles,
                   y = GS451_IC50))
```


***


## Homework

#### Learning Objectives

* Interpret results of a GWAS
* Practice manipulating `vcfR` and tabular data

#### Assignment

Run a GWAS of $\mathrm{IC_{50}}$ for the drug CB1908, using the same genotype data as before. The phenotypes are located in `CB1908_IC50.txt`.

Make a QQ plot and a Manhattan plot of your results. Do you have any genome-wide significant hits? Are they located in or near a gene? For the top GWAS hit, plot the phenotype stratified by genotype.