## Data for class

To avoid VCF wrangling in R, I created tables of the variant and genotype info from `random_snippet.vcf.gz`.

#### Load VCF

```{r}
library(tidyverse)
library(vcfR)

# load vcf
vcf <- read.vcfR(file = "random_snippet.vcf.gz")

# convert vcf into three tidy dataframes
tidied <- vcfR2tidy(vcf,
                    # tell vcfR to turn these INFO fields into integers
                    info_types = c(AF = "n", EAS_AF = "n",
                                   EUR_AF = "n", AFR_AF = "n",
                                   AMR_AF = "n", SAS_AF = "n"))
```

### All variants

```{r}
# make variants dataframe
variants <- tidied$fix

# extract genotypes and recode into counts
gt <- extract.gt(vcf) %>%
  # convert to dataframe so we can use tidyverse functions
  as.data.frame() %>%
  # apply `recode` function to all values in dataframe
  mutate_all(recode,
             "0|0" = 0, "1|0" = 1, "0|1" = 1, "1|1" = 2)

# add columns of variants data
vcf_table <- cbind(variants[, c("AF", "AFR_AF", "AMR_AF", "EAS_AF", "EUR_AF", "SAS_AF")],
                   gt)
vcf_table$ID <- row.names(vcf_table) 
# write to table
write.table(vcf_table, "all_variants.txt",
            sep = "\t", quote = FALSE)
# bgzip to reduce size
```

### Common variants

Create VCF subsetted to common variation:

```{r}
# choose rows of `variants` that have AFs between 0.05 and 0.95
common_rows <- (variants$AF > 0.05) & (variants$AF < 0.95)

# subset rows of the vcf object
vcf_common <- vcf[common_rows, ]
# subset variants dataframe
variants_common <- variants[common_rows, ]

# extract genotypes and recode into counts
gt_common <- extract.gt(vcf_common) %>%
  # convert to dataframe so we can use tidyverse functions
  as.data.frame() %>%
  # apply `recode` function to all values in dataframe
  mutate_all(recode,
             "0|0" = 0, "1|0" = 1, "0|1" = 1, "1|1" = 2)

# add columns of variants data
common_final <- cbind(variants_common[, c("AF", "AFR_AF", "AMR_AF", "EAS_AF", "EUR_AF", "SAS_AF")],
                      gt_common)
# write to table
write.table(common_final, "common_variants.txt",
            sep = "\t", quote = FALSE)
```


## Homework

Create unknown VCF:

```
bcftools query -l 2504_snippet.vcf.gz > samples.txt
# chose NA21121
bcftools view -s NA21121 -O z -o unknown.vcf.gz 2504_snippet.vcf.gz
```

Read in the unknown VCF file, extract the genotype matrix and recode the genotypes as numbers.

***
<details><summary> Solution </summary>

```{r}
# read VCF
unknown <- read.vcfR(file = "unknown.vcf.gz")

# extract and recode genotypes
unknown_matrix <- extract.gt(unknown) %>%
  as.data.frame() %>%
  mutate_all(recode,
             "0|0" = 0, "1|0" = 1, "0|1" = 1, "1|1" = 2) %>%
  as.matrix()

# write to file
write.table(unknown_matrix, "unknown.txt",
            sep = "\t", quote = FALSE)
```

</details>
***