### Data

#### Genotype data

Our genotype data is in a VCF file called `random_snippet.vcf.gz`, which contains a random subset of variants on chromosome 21 that were genotyped in 1000 Genomes. Use `vcfR` to read in the data:

```{r}
vcf <- read.vcfR(file = "random_snippet.vcf.gz")
```


## Tidying VCF data

If we try to look at `vcf`, we can see that it's an "Object of Class vcfR":

```{r}
vcf
```

`vcfR` reformats VCFs into its own data class, which is easy to manipulate using `vcfR` functions, but hard to access directly from the `vcf` variable.

We'll use the `vcfR2tidy` function in order to convert `vcf` into **tidy dataframes** suitable for analysis with R:

```{r, message = FALSE}
# convert vcf into three tidy dataframes
tidied <- vcfR2tidy(vcf,
                    # tell vcfR to turn these INFO fields into integers
                    info_types = c(AF = "n", EAS_AF = "n",
                                   EUR_AF = "n", AFR_AF = "n",
                                   AMR_AF = "n", SAS_AF = "n"))
tidied
```

`tidied` is a _list_ (a type of R object) of three dataframes:

* `fix`: Location, identity, allele frequency, etc. of **variants** in the VCF
* `gt`: **Genotypes** of the VCF samples
* `meta`: Information from the VCF **header**

We'll be using the variants information from `tidied`, so let's store it in a separate table:

```{r}
# make variants dataframe
variants <- tidied$fix
variants
```

 ## Subsetting to common variants

Subset the original VCF object to just common variation, using the `variants` dataframe (which contains the AF information):

```{r}
# choose rows of `variants` that have AFs within range
common_rows <- (variants$AF > 0.05) & (variants$AF < 0.95)

# subset rows of the vcf object
vcf_common <- vcf[common_rows, ]
vcf_common
```

As expected from the cutoff lines on the AFS plot, there are only 960 variants left after subsetting -- substantially less than our original count of 10,000.

We'll also subset the `variants` dataframe itself to only keep information for the common variants. 

```{r}
variants_common <- variants[common_rows, ]
```

## Reformatting data for PCA

We're using R's `prcomp` function to perform PCA on our genotype data. This function takes a matrix where the rows are the data objects (i.e., individuals) and the columns are the associated measurements (i.e., variants). The values within the matrix also have to be numeric.

To reformat our data for `prcomp`, we need to:

1. Extract genotypes from the VCF with vcfR's `extract.gt` function
2. Convert the genotypes into numeric values, where
    * **`0`** is homozygous reference (`0|0`)
    * **`1`** is heterozygous (`0|1`, `1|0`)
    * **`2`** is homozygous alternate (`1|1`)
    * These values also represent the number of alternative alleles that an individual carries.
3. **Transpose** (i.e., rotate) the matrix so the rows are samples and the columns are variants

We'll use a tidyverse function to convert, or "recode", the genotype values in every column of the matrix.

```{r}
# extract genotypes and recode into counts
gt_matrix <- extract.gt(vcf_common) %>%
  # convert to dataframe so we can use tidyverse functions
  as.data.frame() %>%
  # apply `recode` function to all values in dataframe
  mutate_all(recode,
             "0|0" = 0, "1|0" = 1, "0|1" = 1, "1|1" = 2) %>%
  as.matrix()
```

We transpose the matrix with `prcomp`'s `t` function:

```{r, results = FALSE}
gt_matrix_T <- t(gt_matrix)
head(gt_matrix_T)
```