
```{r, include = FALSE}
ottrpal::set_knitr_image_path()
knitr::opts_knit$set(root.dir = '04-pop_structure')
```

# Population structure

In this lab, we'll perform principal component analysis (PCA) to visualize how genetic variation segregates between populations.

#### Learning objectives

After completing this chapter, you'll be able to:

1. Describe the data stored in a Variant Call Format (VCF) file.
2. Plot and interpret an allele frequency spectrum.
3. Perform and visualize the results of a principal component analysis.


## What is a population?

In population genetics, the term **population** refers to a group of interbreeding individuals.

Determining whether a group of individuals is a "population" is subjective. Groups exchange migrants at different rates (a process called **gene flow**), and there is no definitive boundary for when they constitute separate populations.


## Population structure

**Population structure** is a consequence of the fact that when two groups of individuals do not freely interbreed, the two populations typically develop different patterns of genetic variation.

Individuals within a population tend to be share greater genetic similarity with each other than with individuals in other populations. These differences manifest through differences in **allele frequencies** among populations, and result from genetic drift, natural selection, and other evolutionary forces.

We can measure such allele frequency differences to reveal evolutionary relationships among populations, as well as evidence of historical natural selection.

<center>

![**Fig. 1.** Two populations polymorphic for alleles `A` and `a`. The frequency of `A` in Population 1 is $0.83$, while its frequency in Population 2 is $0.15$.](04-pop_structure/images/pop_structure.jpg){width=75%}

</center>


## Geography of Genetic Variants

One quick way to visualize population structure in humans is to look at the allele frequencies of genetic variants in different regions of the world. The [**Geography of Genetic Variants (GGV)**](https://popgen.uchicago.edu/ggv/?data=%221000genomes) browser is a website that plots allele frequencies from the [**1000 Genomes**](https://mccoy-lab.github.io/hgv_modules/the-1000-genomes-project.html) dataset.

For a given variant, GGV plots piecharts of its frequency in the 26 populations in 1000 Genomes, superimposed over the population's geographical location.

While some variants have very similar frequencies across populations, others are much more common in specific populations -- usually suggesting that these populations are more closely related to each other.
<br></br>

<center>

![**Fig. 2.** GGV visualization of the allele frequencies for a SNP at `chr1:222087833`.](04-pop_structure/images/ggv.png)

</center>


## Setup

We'll use the 1000 Genomes data to visualize genetic differentiation between its populations.

### R packages

```{r, results = FALSE, message = FALSE, warning = FALSE}
library(tidyverse)
```


## Genotype data

We've summarized genotype data from the [1000 Genomes Project](https://mccoy-lab.github.io/hgv_modules/the-1000-genomes-project.html) into two files:

* `all_variants.txt.gz`, which contains a random selection of variants on chr21
* `common_variants.txt`, which contains only those variants in `all_variants` that are common across populations

```{r}
# all variants
all <- read.table("all_variants.txt.gz")

# only common variants
common <- read.table("common_variants.txt.gz")

# preview first 10 columns of `all` dataframe
head(all[, 1:10])
```

The **index** of the dataframe is the variant ID.
* The first column (`AF`) contains the variant's **allele frequency (AF)** dataset-wide.
* The next five columns contain the variant's AF in each of the five 1000 Genomes superpopulations.
* The rest of the columns provide variant **genotypes** for each individual in 1000 Genomes, where:
    * `0` is homozygous reference
    * `1` is heterozygous
    * `2` is homozygous for variant


## Metadata

We'll also read in a metadata file, which tells us which population each individual is from.

```{r}
metadata <- read.table("integrated_call_samples.txt",
                       header = TRUE)

head(metadata)
```


## The allele frequency spectrum

One common visualization of genotype data is the **allele frequency spectrum (AFS)**, which is the _distribution_ of the allele frequencies of the variants.

***
<details><summary> Plot the AFS of all variants, using the data in the `AF` column. </summary>

```{r}
ggplot(data = all,
       aes(x = AF)) + 
  geom_histogram(bins = 100)
```

</details>
***

***
<details><summary> Interpreting the AFS </summary>

The allele frequencies in this dataset range from 0 to 1, with an exponential decay from zero, indicating that the majority of variants are rare.

This is the expected shape of an AFS, since all variants arise in one individual and are unlikely to spread widely through the population just by chance. This distribution is exacerbated in human populations, where recent population expansions have resulted in an excess of rare variation.

</details>
***


## Theoretical AFS

Population geneticists have estimated that under neutral demographic expectations, each bin of the AFS should have a height that is equal to 1 over its bin number.

We can use this to plot the expected AFS:

```{r}
# make dataframe with theoretical AFS bins
# create `af_bin` column with the bin number
ideal_pop <- tibble(af_bin = 1:100) %>%
  # create `prop` column with the expected proportion of variants
  mutate(., prop = 1 / af_bin)

head(ideal_pop)
```

```{r}
# plot expected AFS
ggplot(ideal_pop,
       aes(x = af_bin, y = prop)) +
  geom_bar(stat = "identity")
```


***
<details><summary> How does this compare to the AFS we see from human data? </summary>

The human AFS has many more rare variants, which manifests as a higher peak on the left side of the AFS. This is due to recent population expansion in humans, which results in more human individuals and an accumulation of excess new rare variation.

</details>
***

***
<details><summary> How would you expect the AFS to look for a contracting population (ex: endangered species)? </summary>

A contracting population would result in the extinction of many alleles, resulting in more variants that drift to high frequency or go extinct. The AFS for this type of population would look more flat than the neutral expectation (fewer rare alleles, more common ones).

</details>
***


## AF correlations between populations

What if we compare AFs between populations? Do we expect the same variant to have the same AFs in, for example, Africa and Europe?

***
<details><summary> Plot African vs. European AF on a scatterplot. </summary>

```{r}
ggplot(all,
       aes(x = AFR_AF, y = EUR_AF)) +
  geom_point()
```

Most of the variants lie near the `x = y` line, showing that there's a lot of correlated AFs between African and European populations. This is due to these populations' recent common ancestry.

Outlier variants, with very different frequencies in different populations, may have reached these different AFs due to the effects of **selection** -- which we'll discuss in [a later module](https://mccoy-lab.github.io/hgv_modules/scans-for-selection.html).

</details>
***

***
<details><summary> Plot AF correlations for some other population pairs. Do you notice any differences in the distributions? </summary>

```{r}
# east asian vs. european
ggplot(all,
       aes(x = EAS_AF,  y = EUR_AF)) +
  geom_point()

# east asian vs. south asian
ggplot(all,
       aes(x = EAS_AF, y = SAS_AF)) +
  geom_point()
```

There's less spread away from the `y = x` line for the EAS-SAS comparison. Because these populations share a common ancestor more recently than EAS-AFR, there has been less time for drift to change AFs between the populations.

</details>
***


## Common variation

For the rest of this lab, we'll use the `common` dataframe, which includes only variants where $0.05 < \textrm{AF} < 0.95$.  We can look at where this set of **common variants** lies on the full AFS by adding vertical lines at the cutoff allele frequencies:

```{r}
ggplot(data = all, 
       aes(x = AF)) + 
  geom_histogram(bins = 100) + 
  geom_vline(xintercept = 0.05, linetype = "dashed", color = "blue") +
  geom_vline(xintercept = 0.95, linetype = "dashed", color = "brown")
```

***
<details><summary> Plot the AFS of the `common` dataframe, including the dashed lines we used above. </summary>

```{r}
ggplot(data = common,
       aes(x = AF)) +
  geom_histogram() + 
  geom_vline(xintercept = 0.05, linetype = "dashed", color = "blue") +
  geom_vline(xintercept = 0.95, linetype = "dashed", color = "brown")
```

All of these variants lie within the dashed lines. Even with just common variation, we still observe an **exponential decay** of the allele frequencies.

Also note that there are only 960 variants in the `common` dataframe -- substantially less than the 10,000 in the `all` dataframe.

</details>
***

***
<details><summary> Why only work with common variants? </summary>

Rare variants are more likely to show fine-grained population structure -- for example, a variant may be carried by just one individual, or just one family. Because there are so many rare variants, including them causes [differences between individuals to be more pronounced than differences between populations](https://bmcgenomdata.biomedcentral.com/articles/10.1186/s12863-020-0833-x).

While this is a biologically true statement, it makes it harder to visualize population structure, which is why we subset to common variation for PCA.

</details>
***


## Principal component analysis

**Principal component analysis (PCA)** is a method for representing high-dimensional data (i.e., data with many variables) within a smaller number of dimensions. In our case, every individual in the VCF has genotype measurements for hundreds of variants.

You can think of PCA as a projection all the individuals in our dataset into a cloud, where their position is determined by their combination of genotypes.

* The first **principal component (PC)** is the vector through the cloud of data points that captures the greatest possible variance.
* The second PC is the vector that captures the second greatest possible variance, and must also be **perpendicular** to the first vector.
* The same idea applies to the third, fourth, fifth, etc. PCs.
<br></br>

<center>

![**Fig. 3.** A PCA plot that simplifies three-dimensional data into two dimensions.](04-pop_structure/images/pca.jpg)

</center>

For an in-depth visual walkthrough of PCA, you can go to [this website](http://setosa.io/ev/principal-component-analysis/).


## Reformatting data for PCA

We're using R's `prcomp` function to perform PCA on our genotype data. This function takes a matrix where the **rows** are the data objects (i.e., individuals) and the **columns** are the associated measurements (i.e., variants).

First, run the code below to subset our data to just the genotypes:

```{r}
# subset to just genotype columns
gt_matrix <- common[, 7:2510] %>%
  as.matrix()

# view first 10 columns of matrix
head(gt_matrix[, 1:10])
```

We then transpose the matrix with `prcomp`'s `t` function so that the rows are individuals and the columns are variants:

```{r, results = FALSE}
# transpose (i.e., rotate)
gt_matrix_T <- t(gt_matrix)

# view first 10 columns of transposed matrix
head(gt_matrix_T[, 1:10])
```


## Performing PCA

We perform PCA on our genotype matrix with `prcomp`.

```{r}
pca <- prcomp(gt_matrix_T)
```

Our output is a `prcomp` object. This object comprises several tables, which you can preview by typing `pca$` into the console and seeing what R suggests:

* `sdev`: standard deviations of the principle components
* `rotation`, `center`, `scale`: tables we won't use in this lab
* `x`: coordinates of the data objects (the 1000 Genomes individuals) on each PC


## Reformatting PCA output

We can plot our PCA output using the information in `pca$x`.

```{r, results = FALSE}
# extract `x` table
x <- pca$x

# preview first 10 columns
head(x[, 1:10])
```

Every row of `x` is an individual, every column is a PC (going up to 960 PCs!), and the value in each cell represents the sample's coordinate on each PC axis.

Run the code below to create a dataframe of the first three PCs to plot:

```{r}
# create column of sample names
pca_results <- data.frame(sample = rownames(x),
                          PC1 = x[, 1], # PC1 values
                          PC2 = x[, 2], # PC2 values
                          PC3 = x[, 3]) # PC3 values
head(pca_results)
```


## Annotate with population labels

Our last step is adding a column to our PCA dataframe with information about each individual's population.

To do this, we **merge** `pca_results` with our `metadata` table. The `merge` function combines two tables, merging them by matching a column of your choice (specified with `by = `).

```{r}
# merge pca_results and metadata
pca_results <- merge(pca_results, metadata,
                     # specify columns to merge on
                     by.x = "sample", by.y = "sample")

head(pca_results)
```


## PCA plot

***
<details><summary> Create a scatterplot of PC1 vs. PC2, coloring by the `superpop` column. </summary>

```{r}
ggplot(data = pca_results, 
       aes(x = PC1, y = PC2, color = superpop)) +
  geom_point()
```

PC1 separates out the African populations from other populations. This is in line with our knowledge that all non-African populations descend from historical migrations out of Africa -- African populations contain significant genetic diversity that is not represented outside of Africa.

PC2 seems to separate the East Asian and European populations from the other three superpopulations.

</details>
***

***
<details><summary> Repeat the plot with PC2 vs. PC3. Which superpopulations do you observe separating on PC3? </summary>

```{r}
ggplot(data = pca_results, 
       aes(x = PC2, y = PC3, color = superpop)) +
  geom_point()
```

PC3 separates out the Admixed American and South Asian populations, which were collapsed into one group in the first PCA plot.

</details>
***


## Proportion of variance explained

It's hard to tell from the PCA plot whether the separation of populations we see is meaningful, or if the plot is just exaggerating extremely minor differences between groups.

We quantify this by calculating the **proportion of variance explained** for each PC. This tells us how much of the variation in our data is being captured by PC1, PC2, etc.

Variance is the square of the standard deviation, so we can calculate proportion of variance explained from the `sdev` item in our `pca` object. Each value corresponds to the standard deviation for one PC.

```{r}
sd <- pca$sdev
head(sd)
```

The proportion of variance explained by a PC is its variance, divided by the sum of the variances across all PCs. Conveniently, you can calculate this for every PC at once in R:

```{r}
# divide variance of each PC by sum of all variances
var_explained <- sd^2 / sum(sd^2)

# proportion of variance explained for:
var_explained[1] # PC1
var_explained[2] # PC2
var_explained[3] # PC3
```

So, PC1 explains only 9.65% of the variance in our data, PC2 explains 4.34%, and PC3 explains 1.34%.

***
<details><summary> Add x and y axis labels to your plots with the proportion of variance explained by each PC. This is common practice for PCA. </summary>

```{r}
ggplot(data = pca_results, 
       aes(x = PC1, y = PC2, color = superpop)) + 
  geom_point() + 
  xlab("PC1 (9.65%)") +
  ylab("PC2 (4.34%)")
```

```{r}
ggplot(data = pca_results, 
       aes(x = PC2, y = PC3, color = superpop)) + 
  geom_point() + 
  xlab("PC2 (4.34%)") +
  ylab("PC3 (1.34%)")
```

</details>
***


## Conclusion

In this lab, we used genotype data from the 1000 Genomes Project to calculate two measures of **population structure**.

* We explored the [**Geography of Genetic Variants**](https://popgen.uchicago.edu/ggv/?data=%221000genomes) browser, a useful resource for visualizing allele frequency differences between human populations.
<br></br>
* Using genotype data from the 1000 Genomes Project, we plotted the **allele frequency spectrum** of variants in human populations.
  * We saw that humans carry an excess of **rare variation** due to recent population expansion.
<br></br>
* Finally, we used **principal component analysis** to cluster individuals in our dataset by their genotype information. Plotting individuals in PCA space allowed us to distinguish the five superpopulations of 1000 Genomes.


## Homework

We'll now perform PCA using _all_ SNPs in the initial VCF -- not just those that were common in 1000 Genomes individuals. In the optional homework, you'll also use your newly generated PCA plot to predict the ancestry of an unknown sample.

#### Learning Objectives

* Perform and interpret the results of a PCA
* Become familiar with understanding and reusing code


## Required homework

**Assignment:** Re-run the steps we used to generate our PCA plot, this time using the `all` dataframe. Do these plots look any different from our plots with just common variants?

***
<details><summary> Solution </summary>

```{r}
# extract genotypes and convert to matrix
gt_matrix_all <- all[, 7:2510] %>%
  as.matrix()

# transpose
gt_matrix_T_all <- t(gt_matrix_all)
# perform PCA
pca_all <- prcomp(gt_matrix_T_all)

# extract coordinates from PCA object
x_all <- pca_all$x

# create dataframe for plotting
pca_results_all <- data.frame(sample = rownames(x_all),
                              PC1 = x_all[, 1],
                              PC2 = x_all[, 2],
                              PC3 = x_all[, 3])
# merge with metadata
pca_results_all <- merge(pca_results_all, metadata,
                         # specify columns to merge on
                         by.x = "sample", by.y = "sample")

# calculate variance explained by each PC
var_explained_all <- pca_all$sdev^2 / sum(pca_all$sdev^2)
# print for PC1-PC3
var_explained_all[1:3]

# PC1 vs. PC2 plot
ggplot(data = pca_results_all, 
       aes(x = PC1, y = PC2, color = superpop)) + 
  geom_point() + 
  xlab("PC1 (9.15%)") +
  ylab("PC2 (3.82%)")

# PC2 vs. PC3 plot
ggplot(data = pca_results_all, 
       aes(x = PC2, y = PC3, color = superpop)) + 
  geom_point() + 
  xlab("PC2 (3.82%)") +
  ylab("PC3 (1.21%)")
```

The PCA plots actually look pretty similar to the plots with just common variants!

</details>
***


## Optional homework

We can think of our PCA as a _model_ of human individuals. If we have a mystery individual but we know their genotypes for the variants in our PCA, we can _predict_ where they should lie in PCA space and thus guess their ancestry.

We've prepared a file, `unknown.txt`, which contains genotypes for one mystery sample (`NA21121`). We'll compare it to the PCA model that you created for the required homework.

Follow the instructions to predict NA21121's placement on your PCA plot.

#### Prepare unknown sample for PCA

**Assignment:** Read in `unknown.txt`, convert it to a matrix, and transpose.

***
<details><summary> Solution </summary>

```{r}
# read VCF
unknown <- read.table("unknown.txt") %>%
  as.matrix()

# transpose matrix
unknown_T <- t(unknown)
```

</details>
***

#### Predict PCA placement of unknown sample

**Assignment:** Run the code block below to predict and plot NA21121 on top of your PCA plot from the required homework. If necessary, plot PC2 vs. PC3 as well. What superpopulation do you think NA21121 is from?

***
<details><summary> Solution </summary>

```{r}
# predict pca placement of unknown data
unknown_pca <- predict(pca_all,
                       unknown_T)

# create dataframe from predicted PCA
unknown_results <- data.frame("PC1" = unknown_pca[, "PC1"],
                              "PC2" = unknown_pca[, "PC2"],
                              "PC3" = unknown_pca[, "PC3"],
                              "sample" = "NA21121")

# plot PC1 vs. PC2 and then predicted sample
ggplot() +
  # PCA plot from required homework
  geom_point(data = pca_results_all, 
             aes(x = PC1, y = PC2, color = superpop)) +
  # plots the unknown sample's location on the PCs
  geom_label(data = unknown_results,
             aes(x = PC1, y = PC2, label = sample)) + 
  xlab("PC1 (9.15%)") +
  ylab("PC2 (3.82%)")

# plot PC2 vs. PC3
ggplot() +
  geom_point(data = pca_results_all, 
             aes(x = PC2, y = PC3, color = superpop)) +
  geom_label(data = unknown_results,
             aes(x = PC2, y = PC3, label = sample)) + 
  xlab("PC2 (3.82%)") +
  ylab("PC3 (1.21%)")
```

`NA21121` seems to be part of the **SAS (South Asian)** superpopulation. If we look up the sample ID in the [1000 Genomes database](https://www.internationalgenome.org/), we can confirm that it's part of the Gujarati Indians in Houston, TX.

</details>
***