# Population structure

In this lab, we'll implement two common approaches for measuring and visualizing population structure: F~ST~ and principal component analysis (PCA).

## Learning objectives

After completing this chapter, you'll be able to:

1. Describe the data stored in a Variant Call Format (VCF) file.
2. Measure population differentiation from genotype data.
3. Perform and visualize the results of a principal component analysis.


***

## Geography of Genetic Variants

https://popgen.uchicago.edu/ggv/?data=%221000genomes


***


## Setup

### R packages

```{r}
library(tidyverse)
```

### Genotype data

```{r}
# all variants
all <- read.table("all_variants.txt.gz")

# only common variants
common <- read.table("common_variants.txt.gz")
```

```{r}
# preview first 10 columns of `all` dataframe
head(all[, 1:10])
```

### Metadata

```{r}
metadata <- read.table("integrated_call_samples.txt",
                       header = TRUE)

head(metadata)
```


***


## The allele frequency spectrum

Plot the _distribution_ of variant allele frequencies, using the `AF` column of `variants`.

```{r}

```


***


## Theoretical AFS

```{r}
# make dataframe with theoretical AFS bins
# create `af_bin` column with the bin number
ideal_pop <- tibble(af_bin = 1:100) %>%
  # create `prop` column with the expected proportion of variants
  mutate(., prop = 1 / af_bin)

head(ideal_pop)
```

```{r}
# plot expected AFS
ggplot(ideal_pop,
       aes(x = af_bin, y = prop)) +
  geom_bar(stat = "identity")
```


***


## AF correlations between populations

Plot African vs. European AF on a scatterplot.

```{r}

```

Plot AF correlations for some other population pairs.

```{r}

```


***


## Common variation

We can look at which variants are "common" by adding vertical lines to the AFS plot.

```{r}
ggplot(data = all,
       aes(x = AF)) + 
  geom_histogram(bins = 100) + 
  geom_vline(xintercept = 0.05, linetype = "dashed", color = "blue") +
  geom_vline(xintercept = 0.95, linetype = "dashed", color = "brown")
```

Plot the AFS of the `common` dataframe, including the dashed lines we used above.

```{r}

```


***


## Principal component analysis

http://setosa.io/ev/principal-component-analysis/

### Reformatting data for PCA

We're using R's `prcomp` function to perform PCA on our genotype data. To reformat our data for `prcomp`, we need to:

1. Subset the `common` dataframe to just the genotype columnss
2. Convert into a matrix
3. **Transpose** (i.e., rotate) the matrix so the rows are samples and the columns are variants

```{r}
# subset to just genotype columns
gt_matrix <- common[, 7:2510] %>%
  as.matrix()

head(gt_matrix[, 1:10])
```

Transpose the matrix with `prcomp`'s `t` function:

```{r}
gt_matrix_T <- t(gt_matrix)

head(gt_matrix_T[, 1:10])
```

### Performing PCA

```{r}
pca <- prcomp(gt_matrix_T)
```


***


## PCA output

We can plot our PCA output using the information in `pca$x`.

```{r, results = FALSE}
x <- pca$x
head(x)
```

Create a dataframe of the first three PCs to plot:

```{r}
# create column of sample names
pca_results <- data.frame(sample = rownames(x),
                          PC1 = x[, 1], # PC1 values
                          PC2 = x[, 2], # PC2 values
                          PC3 = x[, 3]) # PC3 values
head(pca_results)
```

To annotate individuals on our plot, we'll **merge** `pca_results` with our `metadata` table.

```{r}
# merge pca_results and metadata
pca_results <- merge(pca_results, metadata,
                     # specify columns to merge on
                     by.x = "sample", by.y = "sample")
head(pca_results)
```

### Plot PCA output

Create _scatterplots_ of PC1 vs. PC2 and PC2 vs. PC3, coloring by the `superpop` column:

```{r}
# PC1 vs. PC2

```

```{r}
# PC2 vs. PC3

```


***


## Proportion of variance explained

We can calculate proportion of variance explained from the `sdev` item in our `pca` object.

```{r}
sd <- pca$sdev
head(sd)
```

The proportion of variance explained by a PC is the variance, divided by the sum of the variances across all PCs. Run this code block to calculate it for all PCs:

```{r}
# divide variance of each PC by sum of all variances
var_explained <- sd^2 / sum(sd^2)

# proportion of variance explained for:
var_explained[1] # PC1
var_explained[2] # PC2
var_explained[3] # PC3
```

Add x and y axis labels to your PCA plots with the proportion of variance explained by each PC.

```{r}
# PC1 vs. PC2

```

```{r}
# PC2 vs. PC3

```


***


## Homework

We'll now perform PCA using _all_ SNPs in the initial VCF -- not just those that were common in 1000 Genomes individuals. In the optional homework, you'll also use your newly generated PCA plot to predict the ancestry of an unknown sample.

#### Learning Objectives

* Required homework: Practice performing and interpreting PCA
* Optional homework: Practice performing PCA and reading code


## Required homework

**Assignment:** Re-run the steps we used to generate our PCA plot, this time using the `all` dataframe. Do these plots look any different from our plots with just common variants?

```{r}

```


***


## Optional homework

We can think of our PCA as a _model_ of human individuals. If we have a mystery individual but we know their genotypes for the variants in our PCA, we can _predict_ where they should lie in PCA space and thus guess their ancestry.

We've prepared a file, `unknown.txt`, which contains genotypes for one mystery sample (`NA21121`). We'll compare it to the PCA model that you created for the required homework.

Follow the instructions to predict NA21121's placement on your PCA plot.

#### Prepare unknown sample for PCA

**Assignment:** Read in `unknown.txt`, convert it to a matrix, and transpose.

```{r}
# read file and convert to matrix
unknown <- _________

# transpose matrix
unknown_T <- _________
```

### Predict PCA placement of unknown sample

**Assignment:** Run the code below to predict and plot NA21121 on top of your PCA plot from the required homework. If necessary, plot PC2 vs. PC3 as well. What superpopulation do you think NA21121 is from?

```{r}
# predict pca placement of unknown data
unknown_pca <- predict(_________, # your PCA object with `all` dataframe
                       unknown_T)

# create dataframe from predicted PCA
unknown_results <- data.frame("PC1" = unknown_pca[, "PC1"],
                              "PC2" = unknown_pca[, "PC2"],
                              "PC3" = unknown_pca[, "PC3"],
                              "sample" = "NA21121")

# plot PC1 vs. PC2 and then predicted sample
ggplot() +
  # PCA plot from required homework
  _________ +
  # plots the unknown sample's location on the PCs
  geom_label(data = unknown_results,
             aes(x = PC1, y = PC2, label = sample)) + 
  xlab("PC1 (9.15%)") +
  ylab("PC2 (3.82%)")
```