<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3.5 Confidence intervals | Human Genome Variation Lab</title>
  <meta name="description" content="Description of Course." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="3.5 Confidence intervals | Human Genome Variation Lab" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Description of Course." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3.5 Confidence intervals | Human Genome Variation Lab" />
  
  <meta name="twitter:description" content="Description of Course." />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="assets/dasl_favicon.ico" type="image/x-icon" />
<link rel="prev" href="linear-models.html"/>
<link rel="next" href="conclusion-1.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="assets/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="http://jhudatascience.org/"><img src="https://jhudatascience.org/images/dasl.png" style=" width: 80%; padding-left: 40px; padding-top: 8px; vertical-align: top"</a>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Homepage</a>
<ul>
<li class="chapter" data-level="1.1" data-path="syllabus-logistics.html"><a href="syllabus-logistics.html"><i class="fa fa-check"></i><b>1.1</b> Syllabus &amp; Logistics</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="syllabus-logistics.html"><a href="syllabus-logistics.html#assessment-grading"><i class="fa fa-check"></i><b>1.1.1</b> Assessment &amp; Grading</a></li>
<li class="chapter" data-level="1.1.2" data-path="syllabus-logistics.html"><a href="syllabus-logistics.html#homework"><i class="fa fa-check"></i><b>1.1.2</b> Homework</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="course-description.html"><a href="course-description.html"><i class="fa fa-check"></i><b>1.2</b> Course Description</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="course-description.html"><a href="course-description.html#educational-objectives"><i class="fa fa-check"></i><b>1.2.1</b> Educational Objectives</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="rstudio-cloud.html"><a href="rstudio-cloud.html"><i class="fa fa-check"></i><b>1.3</b> RStudio Cloud</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="rstudio-cloud.html"><a href="rstudio-cloud.html#making-an-rstudio-cloud-account"><i class="fa fa-check"></i><b>1.3.1</b> Making an RStudio Cloud account</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="genome-browsers.html"><a href="genome-browsers.html"><i class="fa fa-check"></i><b>2</b> Genome browsers</a>
<ul>
<li class="chapter" data-level="2.1" data-path="dna-sequencing-data.html"><a href="dna-sequencing-data.html"><i class="fa fa-check"></i><b>2.1</b> DNA sequencing data</a></li>
<li class="chapter" data-level="2.2" data-path="assembling-a-genome.html"><a href="assembling-a-genome.html"><i class="fa fa-check"></i><b>2.2</b> Assembling a genome</a></li>
<li class="chapter" data-level="2.3" data-path="the-human-reference-genome.html"><a href="the-human-reference-genome.html"><i class="fa fa-check"></i><b>2.3</b> The human reference genome</a></li>
<li class="chapter" data-level="2.4" data-path="ucsc-genome-browser.html"><a href="ucsc-genome-browser.html"><i class="fa fa-check"></i><b>2.4</b> UCSC genome browser</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="ucsc-genome-browser.html"><a href="ucsc-genome-browser.html#homepage-1"><i class="fa fa-check"></i><b>2.4.1</b> Homepage</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="viewing-one-region-of-the-genome.html"><a href="viewing-one-region-of-the-genome.html"><i class="fa fa-check"></i><b>2.5</b> Viewing one region of the genome</a></li>
<li class="chapter" data-level="2.6" data-path="customizing-browser-displays.html"><a href="customizing-browser-displays.html"><i class="fa fa-check"></i><b>2.6</b> Customizing browser displays</a></li>
<li class="chapter" data-level="2.7" data-path="igv.html"><a href="igv.html"><i class="fa fa-check"></i><b>2.7</b> IGV</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="igv.html"><a href="igv.html#homepage-2"><i class="fa fa-check"></i><b>2.7.1</b> Homepage</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="navigating-igv.html"><a href="navigating-igv.html"><i class="fa fa-check"></i><b>2.8</b> Navigating IGV</a></li>
<li class="chapter" data-level="2.9" data-path="loading-sequencing-data.html"><a href="loading-sequencing-data.html"><i class="fa fa-check"></i><b>2.9</b> Loading sequencing data</a></li>
<li class="chapter" data-level="2.10" data-path="the-1000-genomes-project.html"><a href="the-1000-genomes-project.html"><i class="fa fa-check"></i><b>2.10</b> The 1000 Genomes Project</a></li>
<li class="chapter" data-level="2.11" data-path="sra.html"><a href="sra.html"><i class="fa fa-check"></i><b>2.11</b> SRA</a>
<ul>
<li class="chapter" data-level="2.11.1" data-path="sra.html"><a href="sra.html#previewing-sequencing-data"><i class="fa fa-check"></i><b>2.11.1</b> Previewing sequencing data</a></li>
</ul></li>
<li class="chapter" data-level="2.12" data-path="viewing-sequencing-reads-in-igv.html"><a href="viewing-sequencing-reads-in-igv.html"><i class="fa fa-check"></i><b>2.12</b> Viewing sequencing reads in IGV</a></li>
<li class="chapter" data-level="2.13" data-path="interpreting-igv-alignments.html"><a href="interpreting-igv-alignments.html"><i class="fa fa-check"></i><b>2.13</b> Interpreting IGV alignments</a></li>
<li class="chapter" data-level="2.14" data-path="conclusion.html"><a href="conclusion.html"><i class="fa fa-check"></i><b>2.14</b> Conclusion</a></li>
<li class="chapter" data-level="2.15" data-path="homework-1.html"><a href="homework-1.html"><i class="fa fa-check"></i><b>2.15</b> Homework</a>
<ul>
<li class="chapter" data-level="2.15.1" data-path="homework-1.html"><a href="homework-1.html#required-homework"><i class="fa fa-check"></i><b>2.15.1</b> Required homework</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="discovering-mutations.html"><a href="discovering-mutations.html"><i class="fa fa-check"></i><b>3</b> Discovering mutations</a>
<ul>
<li class="chapter" data-level="3.1" data-path="background.html"><a href="background.html"><i class="fa fa-check"></i><b>3.1</b> Background</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="background.html"><a href="background.html#de-novo-mutations"><i class="fa fa-check"></i><b>3.1.1</b> _De novo) mutations</a></li>
<li class="chapter" data-level="3.1.2" data-path="background.html"><a href="background.html#recombination"><i class="fa fa-check"></i><b>3.1.2</b> Recombination</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="setup.html"><a href="setup.html"><i class="fa fa-check"></i><b>3.2</b> Setup</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="setup.html"><a href="setup.html#using-r-markdown-files"><i class="fa fa-check"></i><b>3.2.1</b> Using R Markdown files</a></li>
<li class="chapter" data-level="3.2.2" data-path="setup.html"><a href="setup.html#r-packages"><i class="fa fa-check"></i><b>3.2.2</b> R packages</a></li>
<li class="chapter" data-level="3.2.3" data-path="setup.html"><a href="setup.html#data"><i class="fa fa-check"></i><b>3.2.3</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="visualizing-the-data.html"><a href="visualizing-the-data.html"><i class="fa fa-check"></i><b>3.3</b> Visualizing the data</a></li>
<li class="chapter" data-level="3.4" data-path="linear-models.html"><a href="linear-models.html"><i class="fa fa-check"></i><b>3.4</b> Linear models</a></li>
<li class="chapter" data-level="3.5" data-path="confidence-intervals.html"><a href="confidence-intervals.html"><i class="fa fa-check"></i><b>3.5</b> Confidence intervals</a></li>
<li class="chapter" data-level="3.6" data-path="conclusion-1.html"><a href="conclusion-1.html"><i class="fa fa-check"></i><b>3.6</b> Conclusion</a></li>
<li class="chapter" data-level="3.7" data-path="homework-2.html"><a href="homework-2.html"><i class="fa fa-check"></i><b>3.7</b> Homework</a>
<ul>
<li class="chapter" data-level="3.7.1" data-path="homework-2.html"><a href="homework-2.html#required"><i class="fa fa-check"></i><b>3.7.1</b> Required</a></li>
<li class="chapter" data-level="3.7.2" data-path="homework-2.html"><a href="homework-2.html#extra-credit-more-linear-models"><i class="fa fa-check"></i><b>3.7.2</b> Extra credit: More linear models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="simulating-evolution.html"><a href="simulating-evolution.html"><i class="fa fa-check"></i><b>4</b> Simulating evolution</a>
<ul>
<li class="chapter" data-level="4.1" data-path="background-1.html"><a href="background-1.html"><i class="fa fa-check"></i><b>4.1</b> Background</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="background-1.html"><a href="background-1.html#genetic-drift"><i class="fa fa-check"></i><b>4.1.1</b> Genetic drift</a></li>
<li class="chapter" data-level="4.1.2" data-path="background-1.html"><a href="background-1.html#the-wright-fisher-model"><i class="fa fa-check"></i><b>4.1.2</b> The Wright-Fisher model</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="modeling-allele-frequencies.html"><a href="modeling-allele-frequencies.html"><i class="fa fa-check"></i><b>4.2</b> Modeling allele frequencies</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="modeling-allele-frequencies.html"><a href="modeling-allele-frequencies.html#the-binomial-distribution"><i class="fa fa-check"></i><b>4.2.1</b> The binomial distribution</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="setup-1.html"><a href="setup-1.html"><i class="fa fa-check"></i><b>4.3</b> Setup</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="setup-1.html"><a href="setup-1.html#r-packages-1"><i class="fa fa-check"></i><b>4.3.1</b> R packages</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="a-wright-fisher-simulation.html"><a href="a-wright-fisher-simulation.html"><i class="fa fa-check"></i><b>4.4</b> A Wright-Fisher simulation</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="a-wright-fisher-simulation.html"><a href="a-wright-fisher-simulation.html#drawing-one-generation-from-a-binomial-distribution"><i class="fa fa-check"></i><b>4.4.1</b> Drawing one generation from a binomial distribution</a></li>
<li class="chapter" data-level="4.4.2" data-path="a-wright-fisher-simulation.html"><a href="a-wright-fisher-simulation.html#simulating-multiple-generations-with-a-for-loop"><i class="fa fa-check"></i><b>4.4.2</b> Simulating multiple generations with a for loop</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>4.5</b> Plotting</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="plotting.html"><a href="plotting.html#visualizing-changes-in-af-over-generations"><i class="fa fa-check"></i><b>4.5.1</b> Visualizing changes in AF over generations</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="creating-a-simulation-function.html"><a href="creating-a-simulation-function.html"><i class="fa fa-check"></i><b>4.6</b> Creating a simulation function</a>
<ul>
<li class="chapter" data-level="4.6.1" data-path="creating-a-simulation-function.html"><a href="creating-a-simulation-function.html#simulating-with-different-parameters-with-a-function"><i class="fa fa-check"></i><b>4.6.1</b> Simulating with different parameters with a function</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="conclusion-2.html"><a href="conclusion-2.html"><i class="fa fa-check"></i><b>4.7</b> Conclusion</a></li>
<li class="chapter" data-level="4.8" data-path="homework-3.html"><a href="homework-3.html"><i class="fa fa-check"></i><b>4.8</b> Homework</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="homework-3.html"><a href="homework-3.html#required-1"><i class="fa fa-check"></i><b>4.8.1</b> Required</a></li>
<li class="chapter" data-level="4.8.2" data-path="homework-3.html"><a href="homework-3.html#extra-credit-selection"><i class="fa fa-check"></i><b>4.8.2</b> Extra credit: Selection</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="about-the-authors.html"><a href="about-the-authors.html"><i class="fa fa-check"></i>About the Authors</a></li>
<li class="divider"></li>
<p style="text-align:center; font-size: 11pt"> <a href="https://github.com/jhudsl/OTTR_Template" target="blank" > Published with</a> <a href="https://bookdown.org/" target="blank"> <b>bookdown</b></a> (by <a href="http://jhudatascience.org/"> The Johns Hopkins Data Science Lab</a>) </p>
<p style="text-align:center; font-size: 12px;"> <a href="https://github.com/rstudio4edu/rstudio4edu-book/"> Style adapted from: rstudio4edu-book </a> <a href ="https://creativecommons.org/licenses/by/2.0/"> (CC-BY 2.0) </a></p>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Human Genome Variation Lab</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="confidence-intervals" class="section level2" number="3.5">
<h2><span class="header-section-number">3.5</span> Confidence intervals</h2>
<p>Let’s formalize the comparison between the two associations by considering the confidence intervals. A confidence interval is a range of values in which, for some probability, the estimated interval will contain the true value of the unknown estimate (in our case the slope parameter). For a 95% confidence interval, the probability is 0.95. The interval itself has two boundaries, a lower boundary and an upper boundary. For a 95% confidence interval, these boundaries are labelled 2.5 % and 97.5%. Confidence intervals for two parameters may be non-overlapping or overlapping. If 95% confidence intervals for parameters from independent populations do not overlap, then this strongly supports that the parameters are significantly different from one another (for the 0.05 level of significance). The inverse statement, however, is not necessarily true. If 95% confidence intervals for parameters from independent populations do overlap, the parameters may still be significantly different; more analysis, like a hypothesis test, is needed to conclude one way or the other.</p>
<p>We can visually inspect whether the confidence intervals overlap using <code>ggplot</code>. But before that, we’ll first see how to tangibly extract the slope values as well as the confidence intervals. We’ll then construct a dataframe that has all 3 values (slope coefficient, the lower or 2.5 % confidence boundary, and the upper or 97.5 % confidence boundary) for both the paternal model and the maternal model.</p>
<p>To extract a single slope coefficient from the linear model summarized results, we must first decide which coefficient we want. There is one for the intercept and one for the predictor variable. From the following code, we see both of these slope parameters.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="confidence-intervals.html#cb16-1" aria-hidden="true" tabindex="-1"></a>fit_pat<span class="sc">$</span>coefficients</span></code></pre></div>
<p>To extract just the <code>Father_age</code> slope, we add the name of the coefficient, as is done below.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="confidence-intervals.html#cb17-1" aria-hidden="true" tabindex="-1"></a>fit_pat<span class="sc">$</span>coefficients[<span class="st">&quot;Father_age&quot;</span>]</span></code></pre></div>
<p><strong>Coding Practice:</strong> Use the example above to extract just the <code>Mother_age</code> slope. Using the gray boxes from the <em>Pattern code</em> as a guide of where code needs to be added, fill in the following <em>Code block</em> and run it when you are ready.</p>
<p><em>Pattern code</em></p>
<p><span style="background: dimgray; color: dimgray"> __ </span><code>$coefficients[</code><span style="background: dimgray; color: dimgray"> __ </span><code>]</code></p>
<p><br /><br /><br />
<em>Code block</em></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="confidence-intervals.html#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>coefficients[]</span></code></pre></div>
<hr />
<details>
<summary>
Answer
</summary>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="confidence-intervals.html#cb19-1" aria-hidden="true" tabindex="-1"></a>fit_mat<span class="sc">$</span>coefficients[<span class="st">&quot;Mother_age&quot;</span>]</span></code></pre></div>
</details>
<hr />
<p>The confindence intervals can be obtained with a function called <code>confint</code>.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="confidence-intervals.html#cb20-1" aria-hidden="true" tabindex="-1"></a>?confint</span></code></pre></div>
<p>From the manual, we see that we will pass the function a fitted model, which parameter(s) we would like a confidence interval for from that model, as well as the <code>level</code> or the probability for which the confidence interval will contain the true value of the parameter. As an example below, we can extract the <code>Father_age</code> slope parameter’s 95% confidence interval from the <code>fit_pat</code> fitted model.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="confidence-intervals.html#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">confint</span>(fit_pat, <span class="st">&#39;Father_age&#39;</span>, <span class="at">level=</span><span class="fl">0.95</span>)</span></code></pre></div>
<p><strong>Coding Practice:</strong> Use the example above to obtain the 95% confidence interval associated with the <code>Mother_age</code> slope parameter from the maternal fitted model. Using the gray boxes from the <em>Pattern code</em> as a guide of where code needs to be added, fill in the following <em>Code block</em> and run it when you are ready.</p>
<p><em>Pattern code</em></p>
<p><code>confint(</code><span style="background: dimgray; color: dimgray"> __ </span> <code>,</code> <span style="background: dimgray; color: dimgray"> __ </span> <code>, level=</code> <span style="background: dimgray; color: dimgray"> __ </span> <code>)</code></p>
<p><br /><br /><br />
<em>Code block</em></p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="confidence-intervals.html#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">confint</span>( , , <span class="at">level =</span> )</span></code></pre></div>
<hr />
<details>
<summary>
Answer
</summary>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="confidence-intervals.html#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">confint</span>(fit_mat, <span class="st">&#39;Mother_age&#39;</span>, <span class="at">level=</span><span class="fl">0.95</span>)</span></code></pre></div>
</details>
<hr />
<p>We combine all of this data into a single dataframe using the code below. In this dataframe, the associations we were testing (<code>Father_age</code> and <code>Mother_age</code>) are separated into rows, and then the slope estimates as well as the confidence intervals are found in the columns.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="confidence-intervals.html#cb24-1" aria-hidden="true" tabindex="-1"></a>pat_confint <span class="ot">&lt;-</span> <span class="fu">c</span>(fit_pat<span class="sc">$</span>coefficients[<span class="st">&quot;Father_age&quot;</span>], <span class="fu">confint</span>(fit_pat, <span class="st">&#39;Father_age&#39;</span>, <span class="at">level=</span><span class="fl">0.95</span>))</span>
<span id="cb24-2"><a href="confidence-intervals.html#cb24-2" aria-hidden="true" tabindex="-1"></a>mat_confint <span class="ot">&lt;-</span> <span class="fu">c</span>(fit_mat<span class="sc">$</span>coefficients[<span class="st">&quot;Mother_age&quot;</span>], <span class="fu">confint</span>(fit_mat, <span class="st">&#39;Mother_age&#39;</span>, <span class="at">level=</span><span class="fl">0.95</span>))</span>
<span id="cb24-3"><a href="confidence-intervals.html#cb24-3" aria-hidden="true" tabindex="-1"></a>confint_data <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(<span class="fu">data.frame</span>(<span class="at">Mother_age =</span> mat_confint, <span class="at">Father_age =</span> pat_confint)))</span>
<span id="cb24-4"><a href="confidence-intervals.html#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(confint_data) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Slopes&quot;</span>, <span class="st">&quot;Lower&quot;</span>, <span class="st">&quot;Upper&quot;</span>)</span>
<span id="cb24-5"><a href="confidence-intervals.html#cb24-5" aria-hidden="true" tabindex="-1"></a>confint_data</span></code></pre></div>
<p>We can then plot these intervals next to each other either horizontally or vertically. Horizontally follows this pattern.</p>
<pre><code>ggplot(dataframe_name) + 
  geom_point(aes(column_name_with_estimates, rownames(dataframe_name))) + 
  geom_errorbar(aes(xmin = column_name_with_lower_boundaries , xmax = column_name_with_upper_boundaries, y = rownames(dataframe_name), width = 0.1)) + 
  theme_minimal() + 
  ylab(&quot;Model&quot;) + 
  xlab(&quot;Coefficient&quot;)</code></pre>
<p>Vertically follows this pattern. Notice that they differ in whether the estimates, range boundaries, labels, etc. are passed as <code>x</code> values or <code>y</code> values.</p>
<pre><code>ggplot(dataframe_name) + 
  geom_point(aes(rownames(dataframe_name), column_name_with_estimates)) + 
  geom_errorbar(aes(ymin = column_name_with_lower_boundaries , ymax = column_name_with_upper_boundaries, x = rownames(dataframe_name), width = 0.1)) + 
  theme_minimal() + 
  ylab(&quot;Coefficient&quot;) + 
  xlab(&quot;Model&quot;)</code></pre>
<p><strong>Coding Practice:</strong> Use the above code outline to fill in the code block and plot the estimates and confidence intervals (replacing things like <code>dataframe_name</code>, <code>column_name_with_estimates</code>, etc.) and determine whether the intervals are overlapping or non-overlapping. Plotting vertically or horizontally is your choice.</p>
<p><em>Pattern code</em></p>
<div class="line-block"><code>ggplot(</code> <span style="background: dimgray; color: dimgray"> __ </span> <code>) +</code><br />
    <code>geom_point(aes(rownames(</code> <span style="background: dimgray; color: dimgray"> __ </span> <code>),</code> <span style="background: dimgray; color: dimgray"> __ </span> <code>)) +</code><br />
    <code>geom_errorbar(aes(</code> <span style="background: dimgray; color: dimgray"> __ </span><code>min=</code> <span style="background: dimgray; color: dimgray"> __ </span> <code>,</code> <span style="background: dimgray; color: dimgray"> __ </span><code>max =</code> <span style="background: dimgray; color: dimgray"> __ </span> <code>,</code> <span style="background: dimgray; color: dimgray"> __ </span> <code>=rownames(</code> <span style="background: dimgray; color: dimgray"> __ </span> <code>), width = 0.1)) +</code><br />
    <code>theme_minimal() +</code><br />
    <span style="background: dimgray; color: dimgray"> __ </span><code>lab("Coefficient") +</code><br />
    <span style="background: dimgray; color: dimgray"> __ </span><code>lab("Model")</code></div>
<p><br /><br /><br /></p>
<p><em>Code block</em></p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="confidence-intervals.html#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb27-2"><a href="confidence-intervals.html#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="fu">rownames</span>(), )) <span class="sc">+</span> </span>
<span id="cb27-3"><a href="confidence-intervals.html#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>( <span class="at">min=</span> , <span class="at">max =</span> , <span class="ot">=</span><span class="fu">rownames</span>(), <span class="at">width =</span> <span class="fl">0.1</span>)) <span class="sc">+</span> </span>
<span id="cb27-4"><a href="confidence-intervals.html#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minnimal</span>() <span class="sc">+</span> </span>
<span id="cb27-5"><a href="confidence-intervals.html#cb27-5" aria-hidden="true" tabindex="-1"></a>   <span class="fu">lab</span>(<span class="st">&quot;Coefficient&quot;</span>) <span class="sc">+</span> </span>
<span id="cb27-6"><a href="confidence-intervals.html#cb27-6" aria-hidden="true" tabindex="-1"></a>   <span class="fu">lab</span>(<span class="st">&quot;Model&quot;</span>)</span></code></pre></div>
<hr />
<details>
<summary>
Vertical Answer
</summary>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="confidence-intervals.html#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(confint_data) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="fu">rownames</span>(confint_data), Slopes)) <span class="sc">+</span> <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> Lower, <span class="at">ymax =</span> Upper, <span class="at">x=</span> <span class="fu">rownames</span>(confint_data), <span class="at">width =</span> <span class="fl">0.1</span>)) <span class="sc">+</span> <span class="fu">theme_minimal</span>() <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Coefficient&quot;</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;Model&quot;</span>)</span></code></pre></div>
<p><img src="resources/images/02-dnm_files/figure-html/unnamed-chunk-25-1.png" width="672" /></p>
<p>Note that the intervals are non-overlapping.
***</p>
<hr />
<details>
<summary>
Horizontal Answer
</summary>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="confidence-intervals.html#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(confint_data) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="fu">aes</span>(Slopes, <span class="fu">rownames</span>(confint_data))) <span class="sc">+</span> <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">xmin =</span> Lower, <span class="at">xmax =</span> Upper, <span class="at">y =</span> <span class="fu">rownames</span>(confint_data), <span class="at">width =</span> <span class="fl">0.1</span>)) <span class="sc">+</span> <span class="fu">theme_minimal</span>() <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Model&quot;</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;Coefficient&quot;</span>)</span></code></pre></div>
<p><img src="resources/images/02-dnm_files/figure-html/unnamed-chunk-26-1.png" width="672" /></p>
<p>Note that the intervals are non-overlapping.
***</p>
</div>
<hr>
<center> 
  <div class="footer">
      All illustrations <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY. </a>
      <br>
      All other materials <a href= "https://creativecommons.org/licenses/by/4.0/"> CC-BY </a> unless noted otherwise.
  </div>
</center>
            </section>

          </div>
        </div>
      </div>
<a href="linear-models.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="conclusion-1.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
