<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4.8 Homework | Human Genome Variation Lab</title>
  <meta name="description" content="Description of Course." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="4.8 Homework | Human Genome Variation Lab" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Description of Course." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4.8 Homework | Human Genome Variation Lab" />
  
  <meta name="twitter:description" content="Description of Course." />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="assets/dasl_favicon.ico" type="image/x-icon" />
<link rel="prev" href="conclusion-2.html"/>
<link rel="next" href="about-the-authors.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="assets/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="http://jhudatascience.org/"><img src="https://jhudatascience.org/images/dasl.png" style=" width: 80%; padding-left: 40px; padding-top: 8px; vertical-align: top"</a>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Homepage</a>
<ul>
<li class="chapter" data-level="1.1" data-path="syllabus-logistics.html"><a href="syllabus-logistics.html"><i class="fa fa-check"></i><b>1.1</b> Syllabus &amp; Logistics</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="syllabus-logistics.html"><a href="syllabus-logistics.html#assessment-grading"><i class="fa fa-check"></i><b>1.1.1</b> Assessment &amp; Grading</a></li>
<li class="chapter" data-level="1.1.2" data-path="syllabus-logistics.html"><a href="syllabus-logistics.html#homework"><i class="fa fa-check"></i><b>1.1.2</b> Homework</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="course-description.html"><a href="course-description.html"><i class="fa fa-check"></i><b>1.2</b> Course Description</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="course-description.html"><a href="course-description.html#educational-objectives"><i class="fa fa-check"></i><b>1.2.1</b> Educational Objectives</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="rstudio-cloud.html"><a href="rstudio-cloud.html"><i class="fa fa-check"></i><b>1.3</b> RStudio Cloud</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="rstudio-cloud.html"><a href="rstudio-cloud.html#making-an-rstudio-cloud-account"><i class="fa fa-check"></i><b>1.3.1</b> Making an RStudio Cloud account</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="genome-browsers.html"><a href="genome-browsers.html"><i class="fa fa-check"></i><b>2</b> Genome browsers</a>
<ul>
<li class="chapter" data-level="2.1" data-path="dna-sequencing-data.html"><a href="dna-sequencing-data.html"><i class="fa fa-check"></i><b>2.1</b> DNA sequencing data</a></li>
<li class="chapter" data-level="2.2" data-path="assembling-a-genome.html"><a href="assembling-a-genome.html"><i class="fa fa-check"></i><b>2.2</b> Assembling a genome</a></li>
<li class="chapter" data-level="2.3" data-path="the-human-reference-genome.html"><a href="the-human-reference-genome.html"><i class="fa fa-check"></i><b>2.3</b> The human reference genome</a></li>
<li class="chapter" data-level="2.4" data-path="ucsc-genome-browser.html"><a href="ucsc-genome-browser.html"><i class="fa fa-check"></i><b>2.4</b> UCSC genome browser</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="ucsc-genome-browser.html"><a href="ucsc-genome-browser.html#homepage-1"><i class="fa fa-check"></i><b>2.4.1</b> Homepage</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="viewing-one-region-of-the-genome.html"><a href="viewing-one-region-of-the-genome.html"><i class="fa fa-check"></i><b>2.5</b> Viewing one region of the genome</a></li>
<li class="chapter" data-level="2.6" data-path="customizing-browser-displays.html"><a href="customizing-browser-displays.html"><i class="fa fa-check"></i><b>2.6</b> Customizing browser displays</a></li>
<li class="chapter" data-level="2.7" data-path="igv.html"><a href="igv.html"><i class="fa fa-check"></i><b>2.7</b> IGV</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="igv.html"><a href="igv.html#homepage-2"><i class="fa fa-check"></i><b>2.7.1</b> Homepage</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="navigating-igv.html"><a href="navigating-igv.html"><i class="fa fa-check"></i><b>2.8</b> Navigating IGV</a></li>
<li class="chapter" data-level="2.9" data-path="loading-sequencing-data.html"><a href="loading-sequencing-data.html"><i class="fa fa-check"></i><b>2.9</b> Loading sequencing data</a></li>
<li class="chapter" data-level="2.10" data-path="the-1000-genomes-project.html"><a href="the-1000-genomes-project.html"><i class="fa fa-check"></i><b>2.10</b> The 1000 Genomes Project</a></li>
<li class="chapter" data-level="2.11" data-path="sra.html"><a href="sra.html"><i class="fa fa-check"></i><b>2.11</b> SRA</a>
<ul>
<li class="chapter" data-level="2.11.1" data-path="sra.html"><a href="sra.html#previewing-sequencing-data"><i class="fa fa-check"></i><b>2.11.1</b> Previewing sequencing data</a></li>
</ul></li>
<li class="chapter" data-level="2.12" data-path="viewing-sequencing-reads-in-igv.html"><a href="viewing-sequencing-reads-in-igv.html"><i class="fa fa-check"></i><b>2.12</b> Viewing sequencing reads in IGV</a></li>
<li class="chapter" data-level="2.13" data-path="interpreting-igv-alignments.html"><a href="interpreting-igv-alignments.html"><i class="fa fa-check"></i><b>2.13</b> Interpreting IGV alignments</a></li>
<li class="chapter" data-level="2.14" data-path="conclusion.html"><a href="conclusion.html"><i class="fa fa-check"></i><b>2.14</b> Conclusion</a></li>
<li class="chapter" data-level="2.15" data-path="homework-1.html"><a href="homework-1.html"><i class="fa fa-check"></i><b>2.15</b> Homework</a>
<ul>
<li class="chapter" data-level="2.15.1" data-path="homework-1.html"><a href="homework-1.html#required-homework"><i class="fa fa-check"></i><b>2.15.1</b> Required homework</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="discovering-mutations.html"><a href="discovering-mutations.html"><i class="fa fa-check"></i><b>3</b> Discovering mutations</a>
<ul>
<li class="chapter" data-level="3.1" data-path="de-novo-mutations.html"><a href="de-novo-mutations.html"><i class="fa fa-check"></i><b>3.1</b> <em>De novo</em> mutations</a></li>
<li class="chapter" data-level="3.2" data-path="recombination.html"><a href="recombination.html"><i class="fa fa-check"></i><b>3.2</b> Recombination</a></li>
<li class="chapter" data-level="3.3" data-path="setup.html"><a href="setup.html"><i class="fa fa-check"></i><b>3.3</b> Setup</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="setup.html"><a href="setup.html#r-packages"><i class="fa fa-check"></i><b>3.3.1</b> R packages</a></li>
<li class="chapter" data-level="3.3.2" data-path="setup.html"><a href="setup.html#data"><i class="fa fa-check"></i><b>3.3.2</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="visualizing-the-data.html"><a href="visualizing-the-data.html"><i class="fa fa-check"></i><b>3.4</b> Visualizing the data</a></li>
<li class="chapter" data-level="3.5" data-path="linear-models.html"><a href="linear-models.html"><i class="fa fa-check"></i><b>3.5</b> Linear models</a></li>
<li class="chapter" data-level="3.6" data-path="fitting-a-linear-model-for-dnms.html"><a href="fitting-a-linear-model-for-dnms.html"><i class="fa fa-check"></i><b>3.6</b> Fitting a linear model for DNMs</a></li>
<li class="chapter" data-level="3.7" data-path="confidence-intervals.html"><a href="confidence-intervals.html"><i class="fa fa-check"></i><b>3.7</b> Confidence intervals</a></li>
<li class="chapter" data-level="3.8" data-path="calculate-95-cis.html"><a href="calculate-95-cis.html"><i class="fa fa-check"></i><b>3.8</b> Calculate 95% CIs</a></li>
<li class="chapter" data-level="3.9" data-path="conclusion-1.html"><a href="conclusion-1.html"><i class="fa fa-check"></i><b>3.9</b> Conclusion</a></li>
<li class="chapter" data-level="3.10" data-path="homework-2.html"><a href="homework-2.html"><i class="fa fa-check"></i><b>3.10</b> Homework</a></li>
<li class="chapter" data-level="3.11" data-path="required-homework-1.html"><a href="required-homework-1.html"><i class="fa fa-check"></i><b>3.11</b> Required homework</a></li>
<li class="chapter" data-level="3.12" data-path="optional-homework.html"><a href="optional-homework.html"><i class="fa fa-check"></i><b>3.12</b> Optional homework</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="simulating-evolution.html"><a href="simulating-evolution.html"><i class="fa fa-check"></i><b>4</b> Simulating evolution</a>
<ul>
<li class="chapter" data-level="4.1" data-path="background.html"><a href="background.html"><i class="fa fa-check"></i><b>4.1</b> Background</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="background.html"><a href="background.html#genetic-drift"><i class="fa fa-check"></i><b>4.1.1</b> Genetic drift</a></li>
<li class="chapter" data-level="4.1.2" data-path="background.html"><a href="background.html#the-wright-fisher-model"><i class="fa fa-check"></i><b>4.1.2</b> The Wright-Fisher model</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="modeling-allele-frequencies.html"><a href="modeling-allele-frequencies.html"><i class="fa fa-check"></i><b>4.2</b> Modeling allele frequencies</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="modeling-allele-frequencies.html"><a href="modeling-allele-frequencies.html#the-binomial-distribution"><i class="fa fa-check"></i><b>4.2.1</b> The binomial distribution</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="setup-1.html"><a href="setup-1.html"><i class="fa fa-check"></i><b>4.3</b> Setup</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="setup-1.html"><a href="setup-1.html#r-packages-1"><i class="fa fa-check"></i><b>4.3.1</b> R packages</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="a-wright-fisher-simulation.html"><a href="a-wright-fisher-simulation.html"><i class="fa fa-check"></i><b>4.4</b> A Wright-Fisher simulation</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="a-wright-fisher-simulation.html"><a href="a-wright-fisher-simulation.html#drawing-one-generation-from-a-binomial-distribution"><i class="fa fa-check"></i><b>4.4.1</b> Drawing one generation from a binomial distribution</a></li>
<li class="chapter" data-level="4.4.2" data-path="a-wright-fisher-simulation.html"><a href="a-wright-fisher-simulation.html#simulating-multiple-generations-with-a-for-loop"><i class="fa fa-check"></i><b>4.4.2</b> Simulating multiple generations with a for loop</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>4.5</b> Plotting</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="plotting.html"><a href="plotting.html#visualizing-changes-in-af-over-generations"><i class="fa fa-check"></i><b>4.5.1</b> Visualizing changes in AF over generations</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="creating-a-simulation-function.html"><a href="creating-a-simulation-function.html"><i class="fa fa-check"></i><b>4.6</b> Creating a simulation function</a>
<ul>
<li class="chapter" data-level="4.6.1" data-path="creating-a-simulation-function.html"><a href="creating-a-simulation-function.html#simulating-with-different-parameters-with-a-function"><i class="fa fa-check"></i><b>4.6.1</b> Simulating with different parameters with a function</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="conclusion-2.html"><a href="conclusion-2.html"><i class="fa fa-check"></i><b>4.7</b> Conclusion</a></li>
<li class="chapter" data-level="4.8" data-path="homework-3.html"><a href="homework-3.html"><i class="fa fa-check"></i><b>4.8</b> Homework</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="homework-3.html"><a href="homework-3.html#required"><i class="fa fa-check"></i><b>4.8.1</b> Required</a></li>
<li class="chapter" data-level="4.8.2" data-path="homework-3.html"><a href="homework-3.html#extra-credit-selection"><i class="fa fa-check"></i><b>4.8.2</b> Extra credit: Selection</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="about-the-authors.html"><a href="about-the-authors.html"><i class="fa fa-check"></i>About the Authors</a></li>
<li class="divider"></li>
<p style="text-align:center; font-size: 11pt"> <a href="https://github.com/jhudsl/OTTR_Template" target="blank" > Published with</a> <a href="https://bookdown.org/" target="blank"> <b>bookdown</b></a> (by <a href="http://jhudatascience.org/"> The Johns Hopkins Data Science Lab</a>) </p>
<p style="text-align:center; font-size: 12px;"> <a href="https://github.com/rstudio4edu/rstudio4edu-book/"> Style adapted from: rstudio4edu-book </a> <a href ="https://creativecommons.org/licenses/by/2.0/"> (CC-BY 2.0) </a></p>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Human Genome Variation Lab</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="homework-3" class="section level2" number="4.8">
<h2><span class="header-section-number">4.8</span> Homework</h2>
<div id="goals-learning-objectives-2" class="section level4" number="4.8.0.1">
<h4><span class="header-section-number">4.8.0.1</span> Goals &amp; Learning Objectives</h4>
<p>The goal of this homework is to explore different extensions of the Wright-Fisher model.</p>
<p><strong>Learning Objectives</strong></p>
<ul>
<li>Required homework: Practice visualizing allele frequencies with <code>ggplot</code></li>
<li>Extra credit: Practice writing functions and interpreting allele frequency trajectories</li>
</ul>
</div>
<div id="required" class="section level3" number="4.8.1">
<h3><span class="header-section-number">4.8.1</span> Required</h3>
<p>Currently, our allele frequency trajectory plot only shows the AF of one allele at a locus. (i.e., if individuals in a population have either an <code>A</code> or a <code>C</code> at some locus, we’re only plotting the trajectory of the <code>A</code> allele.)</p>
<p><strong>Assignment:</strong> Add a line to allele frequency trajectory plot that shows the frequency of the other allele at the locus. Give the two alleles different colors.</p>
<hr />
<details>
<summary>
Hint
</summary>
You don’t need to modify the simulation function for this.
</details>
<hr />
<hr />
<details>
<summary>
Solution
</summary>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="homework-3.html#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run the simulation again to get output data</span></span>
<span id="cb55-2"><a href="homework-3.html#cb55-2" aria-hidden="true" tabindex="-1"></a>sim <span class="ot">&lt;-</span> <span class="fu">run_sim</span>(<span class="at">Ne =</span> <span class="dv">1000</span>, <span class="at">freq =</span> <span class="fl">0.5</span>, <span class="at">generations =</span> <span class="dv">10000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb55-3"><a href="homework-3.html#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add in a column with the AF of the minor allele</span></span>
<span id="cb55-4"><a href="homework-3.html#cb55-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># this is 1 - AF of the major allele</span></span>
<span id="cb55-5"><a href="homework-3.html#cb55-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">afs_minor =</span> <span class="dv">1</span> <span class="sc">-</span> afs)</span>
<span id="cb55-6"><a href="homework-3.html#cb55-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-7"><a href="homework-3.html#cb55-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb55-8"><a href="homework-3.html#cb55-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">data =</span> sim, <span class="fu">aes</span>(<span class="at">x =</span> gen, <span class="at">y =</span> afs), <span class="at">color =</span> <span class="st">&quot;blue&quot;</span>) <span class="sc">+</span></span>
<span id="cb55-9"><a href="homework-3.html#cb55-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add in another line with the frequency of the minor allele</span></span>
<span id="cb55-10"><a href="homework-3.html#cb55-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">data =</span> sim, <span class="fu">aes</span>(<span class="at">x =</span> gen, <span class="at">y =</span> afs_minor), <span class="at">color =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb55-11"><a href="homework-3.html#cb55-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>, <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb55-12"><a href="homework-3.html#cb55-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Allele frequency&quot;</span>) <span class="sc">+</span></span>
<span id="cb55-13"><a href="homework-3.html#cb55-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Generation&quot;</span>)</span></code></pre></div>
<img src="resources/images/resources/images/resources/images/hgv_lab_files/figure-html/unnamed-chunk-36-1.png" width="672" />
</details>
<hr />
</div>
<div id="extra-credit-selection" class="section level3" number="4.8.2">
<h3><span class="header-section-number">4.8.2</span> Extra credit: Selection</h3>
<p>One way to extend our simple Wright-Fisher model is to add in selection as a parameter. Selection affects our model by altering the probability of sampling our allele of interest each generation (e.g., positive selection increases the probability, and negative selection decreases it).</p>
<p>Previously, we assumed that this probability was equivalent to the allele’s frequency, or <span class="math inline">\(p = \frac{i}{N_e}\)</span>, where <span class="math inline">\(N_e\)</span> is the population size and <span class="math inline">\(i\)</span> is the number of individuals who carry the allele.</p>
<p>For the purposes of this homework, we assume that in a model with selection, this probability is instead:</p>
<p><span class="math display">\[
p = \frac{i(1 + s)}{N_e - i + i(1+s)}
\]</span></p>
<p>where <span class="math inline">\(s\)</span> is the <strong>selection coefficient</strong>, and ranges from <code>-1</code> to <code>1</code>.</p>
<hr />
<details>
<summary>
<strong>Question</strong>: What does this probability become in the absence of selection (i.e., when <span class="math inline">\(s = 0\)</span>)?
</summary>
The probability becomes <span class="math inline">\(\frac{i}{N_e}\)</span>, which is the same as the allele frequency.
</details>
<hr />
<p><strong>Assignment:</strong> Modify your <code>run_sim</code> function so that it takes in a selection coefficient <code>s</code> as a parameter. Run the simulation a few times with and without (<code>s = 0</code>) selection, but keeping other parameters the same (<code>Ne = 10000</code> – to mimic the <code>Ne</code> of humans, <code>freq = 0.5</code>, <code>generations = 10000</code>). What do you notice about the allele frequency trajectories?</p>
<hr />
<details>
<summary>
Solution
</summary>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="homework-3.html#cb56-1" aria-hidden="true" tabindex="-1"></a>run_sim_selection <span class="ot">&lt;-</span> <span class="cf">function</span>(Ne, freq, generations, s) {</span>
<span id="cb56-2"><a href="homework-3.html#cb56-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb56-3"><a href="homework-3.html#cb56-3" aria-hidden="true" tabindex="-1"></a>  freq_vector <span class="ot">&lt;-</span> freq</span>
<span id="cb56-4"><a href="homework-3.html#cb56-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>generations) {</span>
<span id="cb56-5"><a href="homework-3.html#cb56-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># calculate p, the probability of sampling the allele, based on s</span></span>
<span id="cb56-6"><a href="homework-3.html#cb56-6" aria-hidden="true" tabindex="-1"></a>    i <span class="ot">&lt;-</span> freq <span class="sc">*</span> Ne <span class="co"># number of individuals who currently carry the allele</span></span>
<span id="cb56-7"><a href="homework-3.html#cb56-7" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> i<span class="sc">*</span>(<span class="dv">1</span><span class="sc">+</span>s) <span class="sc">/</span> (Ne <span class="sc">-</span> i <span class="sc">+</span> i<span class="sc">*</span>(<span class="dv">1</span><span class="sc">+</span>s))</span>
<span id="cb56-8"><a href="homework-3.html#cb56-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb56-9"><a href="homework-3.html#cb56-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># prob is now `p`, rather than `freq`</span></span>
<span id="cb56-10"><a href="homework-3.html#cb56-10" aria-hidden="true" tabindex="-1"></a>    new_freq <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="at">n =</span> <span class="dv">1</span>, <span class="at">size =</span> Ne, <span class="at">prob =</span> p) <span class="sc">/</span> Ne</span>
<span id="cb56-11"><a href="homework-3.html#cb56-11" aria-hidden="true" tabindex="-1"></a>    freq_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(freq_vector, new_freq)</span>
<span id="cb56-12"><a href="homework-3.html#cb56-12" aria-hidden="true" tabindex="-1"></a>    freq <span class="ot">&lt;-</span> new_freq</span>
<span id="cb56-13"><a href="homework-3.html#cb56-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb56-14"><a href="homework-3.html#cb56-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb56-15"><a href="homework-3.html#cb56-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># convert vector of AFs into a tibble for plotting</span></span>
<span id="cb56-16"><a href="homework-3.html#cb56-16" aria-hidden="true" tabindex="-1"></a>  sim_results <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">afs =</span> freq_vector,</span>
<span id="cb56-17"><a href="homework-3.html#cb56-17" aria-hidden="true" tabindex="-1"></a>                        <span class="at">gen =</span> <span class="dv">1</span><span class="sc">:</span>(generations<span class="sc">+</span><span class="dv">1</span>))</span>
<span id="cb56-18"><a href="homework-3.html#cb56-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb56-19"><a href="homework-3.html#cb56-19" aria-hidden="true" tabindex="-1"></a>  <span class="co"># return the tibble of AFs, so that we can access the results</span></span>
<span id="cb56-20"><a href="homework-3.html#cb56-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(sim_results)</span>
<span id="cb56-21"><a href="homework-3.html#cb56-21" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Run and plot the simulation with selection:</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="homework-3.html#cb57-1" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">run_sim_selection</span>(<span class="at">Ne =</span> <span class="dv">10000</span>,</span>
<span id="cb57-2"><a href="homework-3.html#cb57-2" aria-hidden="true" tabindex="-1"></a>                             <span class="at">freq =</span> <span class="fl">0.5</span>,</span>
<span id="cb57-3"><a href="homework-3.html#cb57-3" aria-hidden="true" tabindex="-1"></a>                             <span class="at">generations =</span> <span class="dv">10000</span>,</span>
<span id="cb57-4"><a href="homework-3.html#cb57-4" aria-hidden="true" tabindex="-1"></a>                             <span class="at">s =</span> <span class="sc">-</span><span class="fl">0.1</span>)</span>
<span id="cb57-5"><a href="homework-3.html#cb57-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb57-6"><a href="homework-3.html#cb57-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">data =</span> results, <span class="fu">aes</span>(<span class="at">x =</span> gen, <span class="at">y =</span> afs)) <span class="sc">+</span></span>
<span id="cb57-7"><a href="homework-3.html#cb57-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>, <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb57-8"><a href="homework-3.html#cb57-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Allele frequency&quot;</span>) <span class="sc">+</span></span>
<span id="cb57-9"><a href="homework-3.html#cb57-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Generation&quot;</span>) <span class="sc">+</span></span>
<span id="cb57-10"><a href="homework-3.html#cb57-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Simulation with selection&quot;</span>) <span class="sc">+</span></span>
<span id="cb57-11"><a href="homework-3.html#cb57-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>)) <span class="co"># to center the title</span></span></code></pre></div>
<p><img src="resources/images/resources/images/resources/images/hgv_lab_files/figure-html/unnamed-chunk-38-1.png" width="672" /></p>
<p>Run and plot the simulation without selection:</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="homework-3.html#cb58-1" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">run_sim_selection</span>(<span class="at">Ne =</span> <span class="dv">10000</span>,</span>
<span id="cb58-2"><a href="homework-3.html#cb58-2" aria-hidden="true" tabindex="-1"></a>                             <span class="at">freq =</span> <span class="fl">0.5</span>,</span>
<span id="cb58-3"><a href="homework-3.html#cb58-3" aria-hidden="true" tabindex="-1"></a>                             <span class="at">generations =</span> <span class="dv">10000</span>,</span>
<span id="cb58-4"><a href="homework-3.html#cb58-4" aria-hidden="true" tabindex="-1"></a>                             <span class="at">s =</span> <span class="dv">0</span>)</span>
<span id="cb58-5"><a href="homework-3.html#cb58-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb58-6"><a href="homework-3.html#cb58-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">data =</span> results, <span class="fu">aes</span>(<span class="at">x =</span> gen, <span class="at">y =</span> afs)) <span class="sc">+</span></span>
<span id="cb58-7"><a href="homework-3.html#cb58-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>, <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb58-8"><a href="homework-3.html#cb58-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Allele frequency&quot;</span>) <span class="sc">+</span></span>
<span id="cb58-9"><a href="homework-3.html#cb58-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Generation&quot;</span>) <span class="sc">+</span></span>
<span id="cb58-10"><a href="homework-3.html#cb58-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Simulation without selection&quot;</span>) <span class="sc">+</span></span>
<span id="cb58-11"><a href="homework-3.html#cb58-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>)) <span class="co"># to center the title</span></span></code></pre></div>
<p><img src="resources/images/resources/images/resources/images/hgv_lab_files/figure-html/unnamed-chunk-39-1.png" width="672" /></p>
<p>We observe that selection (at least, strong selection where <code>s = 0.1</code>) tends to decrease the time it takes for an allele to either fix or go extinct. This is because selection directionally biases the probability of sampling that allele.</p>
Decreasing the absolute value of the selection coefficient will make the simulation behave more like drift - most selection coefficients are thought to be <a href="https://academic.oup.com/view-large/figure/326984394/345fig4.jpeg">very small</a>, and the largest known selection coefficients in humans are around <a href="https://elifesciences.org/articles/63177">0.05</a>.
</details>
<hr />

</div>
</div>
<!-- </div> -->
<hr>
<center> 
  <div class="footer">
      All illustrations <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY. </a>
      <br>
      All other materials <a href= "https://creativecommons.org/licenses/by/4.0/"> CC-BY </a> unless noted otherwise.
  </div>
</center>
            </section>

          </div>
        </div>
      </div>
<a href="conclusion-2.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="about-the-authors.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
