<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5.5 Tidying VCF data | Human Genome Variation Lab</title>
  <meta name="description" content="Description of Course" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="5.5 Tidying VCF data | Human Genome Variation Lab" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Description of Course" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5.5 Tidying VCF data | Human Genome Variation Lab" />
  
  <meta name="twitter:description" content="Description of Course" />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="assets/favicon.ico" type="image/x-icon" />
<link rel="prev" href="setup-2.html"/>
<link rel="next" href="the-allele-frequency-spectrum.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<link rel="shortcut icon" href="assets/favicon.ico?" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="assets/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://mccoy-lab.github.io/hgv_modules/"><img src="assets/logo.png" style=" width: 80%; padding-left: 40px; padding-top: 8px; vertical-align: top"</a>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Homepage</a>
<ul>
<li class="chapter" data-level="1.1" data-path="course-description.html"><a href="course-description.html"><i class="fa fa-check"></i><b>1.1</b> Course Description</a></li>
<li class="chapter" data-level="1.2" data-path="assessment-grading.html"><a href="assessment-grading.html"><i class="fa fa-check"></i><b>1.2</b> Assessment &amp; Grading</a></li>
<li class="chapter" data-level="1.3" data-path="posit-cloud.html"><a href="posit-cloud.html"><i class="fa fa-check"></i><b>1.3</b> Posit Cloud</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="genome-browsers.html"><a href="genome-browsers.html"><i class="fa fa-check"></i><b>2</b> Genome browsers</a>
<ul>
<li class="chapter" data-level="2.1" data-path="dna-sequencing-data.html"><a href="dna-sequencing-data.html"><i class="fa fa-check"></i><b>2.1</b> DNA sequencing data</a></li>
<li class="chapter" data-level="2.2" data-path="assembling-a-genome.html"><a href="assembling-a-genome.html"><i class="fa fa-check"></i><b>2.2</b> Assembling a genome</a></li>
<li class="chapter" data-level="2.3" data-path="the-human-reference-genome.html"><a href="the-human-reference-genome.html"><i class="fa fa-check"></i><b>2.3</b> The human reference genome</a></li>
<li class="chapter" data-level="2.4" data-path="ucsc-genome-browser.html"><a href="ucsc-genome-browser.html"><i class="fa fa-check"></i><b>2.4</b> UCSC genome browser</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="ucsc-genome-browser.html"><a href="ucsc-genome-browser.html#homepage-1"><i class="fa fa-check"></i><b>2.4.1</b> Homepage</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="viewing-one-region-of-the-genome.html"><a href="viewing-one-region-of-the-genome.html"><i class="fa fa-check"></i><b>2.5</b> Viewing one region of the genome</a></li>
<li class="chapter" data-level="2.6" data-path="igv.html"><a href="igv.html"><i class="fa fa-check"></i><b>2.6</b> IGV</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="igv.html"><a href="igv.html#homepage-2"><i class="fa fa-check"></i><b>2.6.1</b> Homepage</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="navigating-igv.html"><a href="navigating-igv.html"><i class="fa fa-check"></i><b>2.7</b> Navigating IGV</a></li>
<li class="chapter" data-level="2.8" data-path="loading-sequencing-data.html"><a href="loading-sequencing-data.html"><i class="fa fa-check"></i><b>2.8</b> Loading sequencing data</a></li>
<li class="chapter" data-level="2.9" data-path="the-1000-genomes-project.html"><a href="the-1000-genomes-project.html"><i class="fa fa-check"></i><b>2.9</b> The 1000 Genomes Project</a></li>
<li class="chapter" data-level="2.10" data-path="sra.html"><a href="sra.html"><i class="fa fa-check"></i><b>2.10</b> SRA</a>
<ul>
<li class="chapter" data-level="2.10.1" data-path="sra.html"><a href="sra.html#previewing-sequencing-data"><i class="fa fa-check"></i><b>2.10.1</b> Previewing sequencing data</a></li>
</ul></li>
<li class="chapter" data-level="2.11" data-path="viewing-sequencing-reads-in-igv.html"><a href="viewing-sequencing-reads-in-igv.html"><i class="fa fa-check"></i><b>2.11</b> Viewing sequencing reads in IGV</a></li>
<li class="chapter" data-level="2.12" data-path="interpreting-igv-alignments.html"><a href="interpreting-igv-alignments.html"><i class="fa fa-check"></i><b>2.12</b> Interpreting IGV alignments</a></li>
<li class="chapter" data-level="2.13" data-path="conclusion.html"><a href="conclusion.html"><i class="fa fa-check"></i><b>2.13</b> Conclusion</a></li>
<li class="chapter" data-level="2.14" data-path="homework-1.html"><a href="homework-1.html"><i class="fa fa-check"></i><b>2.14</b> Homework</a>
<ul>
<li class="chapter" data-level="2.14.1" data-path="homework-1.html"><a href="homework-1.html#required-homework"><i class="fa fa-check"></i><b>2.14.1</b> Required homework</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="discovering-mutations.html"><a href="discovering-mutations.html"><i class="fa fa-check"></i><b>3</b> Discovering mutations</a>
<ul>
<li class="chapter" data-level="3.1" data-path="de-novo-mutations.html"><a href="de-novo-mutations.html"><i class="fa fa-check"></i><b>3.1</b> <em>De novo</em> mutations</a></li>
<li class="chapter" data-level="3.2" data-path="recombination.html"><a href="recombination.html"><i class="fa fa-check"></i><b>3.2</b> Recombination</a></li>
<li class="chapter" data-level="3.3" data-path="setup.html"><a href="setup.html"><i class="fa fa-check"></i><b>3.3</b> Setup</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="setup.html"><a href="setup.html#r-packages"><i class="fa fa-check"></i><b>3.3.1</b> R packages</a></li>
<li class="chapter" data-level="3.3.2" data-path="setup.html"><a href="setup.html#data"><i class="fa fa-check"></i><b>3.3.2</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="visualizing-the-data.html"><a href="visualizing-the-data.html"><i class="fa fa-check"></i><b>3.4</b> Visualizing the data</a></li>
<li class="chapter" data-level="3.5" data-path="linear-models.html"><a href="linear-models.html"><i class="fa fa-check"></i><b>3.5</b> Linear models</a></li>
<li class="chapter" data-level="3.6" data-path="fitting-a-linear-model-for-dnms.html"><a href="fitting-a-linear-model-for-dnms.html"><i class="fa fa-check"></i><b>3.6</b> Fitting a linear model for DNMs</a></li>
<li class="chapter" data-level="3.7" data-path="confidence-intervals.html"><a href="confidence-intervals.html"><i class="fa fa-check"></i><b>3.7</b> Confidence intervals</a></li>
<li class="chapter" data-level="3.8" data-path="calculate-95-cis.html"><a href="calculate-95-cis.html"><i class="fa fa-check"></i><b>3.8</b> Calculate 95% CIs</a></li>
<li class="chapter" data-level="3.9" data-path="conclusion-1.html"><a href="conclusion-1.html"><i class="fa fa-check"></i><b>3.9</b> Conclusion</a></li>
<li class="chapter" data-level="3.10" data-path="homework-2.html"><a href="homework-2.html"><i class="fa fa-check"></i><b>3.10</b> Homework</a></li>
<li class="chapter" data-level="3.11" data-path="required-homework-1.html"><a href="required-homework-1.html"><i class="fa fa-check"></i><b>3.11</b> Required homework</a></li>
<li class="chapter" data-level="3.12" data-path="optional-homework.html"><a href="optional-homework.html"><i class="fa fa-check"></i><b>3.12</b> Optional homework</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="linkage-disequilibrium.html"><a href="linkage-disequilibrium.html"><i class="fa fa-check"></i><b>4</b> Linkage disequilibrium</a>
<ul>
<li class="chapter" data-level="4.1" data-path="what-is-linkage-disequilibrium.html"><a href="what-is-linkage-disequilibrium.html"><i class="fa fa-check"></i><b>4.1</b> What is linkage disequilibrium?</a></li>
<li class="chapter" data-level="4.2" data-path="why-do-we-care-about-ld.html"><a href="why-do-we-care-about-ld.html"><i class="fa fa-check"></i><b>4.2</b> Why do we care about LD?</a></li>
<li class="chapter" data-level="4.3" data-path="setup-1.html"><a href="setup-1.html"><i class="fa fa-check"></i><b>4.3</b> Setup</a></li>
<li class="chapter" data-level="4.4" data-path="are-these-snps-in-ld.html"><a href="are-these-snps-in-ld.html"><i class="fa fa-check"></i><b>4.4</b> Are these SNPs in LD?</a></li>
<li class="chapter" data-level="4.5" data-path="counting-haplotypes-with-table.html"><a href="counting-haplotypes-with-table.html"><i class="fa fa-check"></i><b>4.5</b> Counting haplotypes with <code>table</code></a></li>
<li class="chapter" data-level="4.6" data-path="fishers-exact-test.html"><a href="fishers-exact-test.html"><i class="fa fa-check"></i><b>4.6</b> Fisher’s exact test</a></li>
<li class="chapter" data-level="4.7" data-path="measuring-ld-with-d.html"><a href="measuring-ld-with-d.html"><i class="fa fa-check"></i><b>4.7</b> Measuring LD with <span class="math inline">\(D\)</span></a></li>
<li class="chapter" data-level="4.8" data-path="calculating-d.html"><a href="calculating-d.html"><i class="fa fa-check"></i><b>4.8</b> Calculating <span class="math inline">\(D\)</span></a></li>
<li class="chapter" data-level="4.9" data-path="measuring-ld-with-d-1.html"><a href="measuring-ld-with-d-1.html"><i class="fa fa-check"></i><b>4.9</b> Measuring LD with <span class="math inline">\(D&#39;\)</span></a></li>
<li class="chapter" data-level="4.10" data-path="measuring-ld-with-r2.html"><a href="measuring-ld-with-r2.html"><i class="fa fa-check"></i><b>4.10</b> Measuring LD with <span class="math inline">\(r^2\)</span></a></li>
<li class="chapter" data-level="4.11" data-path="ldlink.html"><a href="ldlink.html"><i class="fa fa-check"></i><b>4.11</b> LDlink</a></li>
<li class="chapter" data-level="4.12" data-path="visualizing-ld-blocks.html"><a href="visualizing-ld-blocks.html"><i class="fa fa-check"></i><b>4.12</b> Visualizing LD blocks</a></li>
<li class="chapter" data-level="4.13" data-path="ld-in-association-studies.html"><a href="ld-in-association-studies.html"><i class="fa fa-check"></i><b>4.13</b> LD in association studies</a></li>
<li class="chapter" data-level="4.14" data-path="conclusion-2.html"><a href="conclusion-2.html"><i class="fa fa-check"></i><b>4.14</b> Conclusion</a></li>
<li class="chapter" data-level="4.15" data-path="homework-3.html"><a href="homework-3.html"><i class="fa fa-check"></i><b>4.15</b> Homework</a></li>
<li class="chapter" data-level="4.16" data-path="required-homework-2.html"><a href="required-homework-2.html"><i class="fa fa-check"></i><b>4.16</b> Required homework</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>5</b> Population structure</a>
<ul>
<li class="chapter" data-level="5.1" data-path="what-is-a-population.html"><a href="what-is-a-population.html"><i class="fa fa-check"></i><b>5.1</b> What is a population?</a></li>
<li class="chapter" data-level="5.2" data-path="geography-of-genetic-variants.html"><a href="geography-of-genetic-variants.html"><i class="fa fa-check"></i><b>5.2</b> Geography of Genetic Variants</a></li>
<li class="chapter" data-level="5.3" data-path="variant-call-format-vcf.html"><a href="variant-call-format-vcf.html"><i class="fa fa-check"></i><b>5.3</b> Variant Call Format (VCF)</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="variant-call-format-vcf.html"><a href="variant-call-format-vcf.html#what-is-a-vcf"><i class="fa fa-check"></i><b>5.3.1</b> What is a VCF?</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="setup-2.html"><a href="setup-2.html"><i class="fa fa-check"></i><b>5.4</b> Setup</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="setup-2.html"><a href="setup-2.html#r-packages-1"><i class="fa fa-check"></i><b>5.4.1</b> R packages</a></li>
<li class="chapter" data-level="5.4.2" data-path="setup-2.html"><a href="setup-2.html#data-1"><i class="fa fa-check"></i><b>5.4.2</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="tidying-vcf-data.html"><a href="tidying-vcf-data.html"><i class="fa fa-check"></i><b>5.5</b> Tidying VCF data</a></li>
<li class="chapter" data-level="5.6" data-path="the-allele-frequency-spectrum.html"><a href="the-allele-frequency-spectrum.html"><i class="fa fa-check"></i><b>5.6</b> The allele frequency spectrum</a></li>
<li class="chapter" data-level="5.7" data-path="common-variation.html"><a href="common-variation.html"><i class="fa fa-check"></i><b>5.7</b> Common variation</a></li>
<li class="chapter" data-level="5.8" data-path="subsetting-to-common-variants.html"><a href="subsetting-to-common-variants.html"><i class="fa fa-check"></i><b>5.8</b> Subsetting to common variants</a></li>
<li class="chapter" data-level="5.9" data-path="the-fst-statistic.html"><a href="the-fst-statistic.html"><i class="fa fa-check"></i><b>5.9</b> The F<sub>ST</sub> statistic</a></li>
<li class="chapter" data-level="5.10" data-path="the-genetic_diff-function.html"><a href="the-genetic_diff-function.html"><i class="fa fa-check"></i><b>5.10</b> The <code>genetic_diff</code> function</a></li>
<li class="chapter" data-level="5.11" data-path="calculating-fst.html"><a href="calculating-fst.html"><i class="fa fa-check"></i><b>5.11</b> Calculating F<sub>ST</sub></a></li>
<li class="chapter" data-level="5.12" data-path="distribution-of-gst-across-the-genome.html"><a href="distribution-of-gst-across-the-genome.html"><i class="fa fa-check"></i><b>5.12</b> Distribution of G<sub>ST</sub> across the genome</a></li>
<li class="chapter" data-level="5.13" data-path="top-gst-hits.html"><a href="top-gst-hits.html"><i class="fa fa-check"></i><b>5.13</b> Top G<sub>ST</sub> hits</a></li>
<li class="chapter" data-level="5.14" data-path="viewing-gst-hits-in-ggv.html"><a href="viewing-gst-hits-in-ggv.html"><i class="fa fa-check"></i><b>5.14</b> Viewing G<sub>ST</sub> hits in GGV</a></li>
<li class="chapter" data-level="5.15" data-path="principal-component-analysis.html"><a href="principal-component-analysis.html"><i class="fa fa-check"></i><b>5.15</b> Principal component analysis</a></li>
<li class="chapter" data-level="5.16" data-path="reformatting-data-for-pca.html"><a href="reformatting-data-for-pca.html"><i class="fa fa-check"></i><b>5.16</b> Reformatting data for PCA</a></li>
<li class="chapter" data-level="5.17" data-path="performing-pca.html"><a href="performing-pca.html"><i class="fa fa-check"></i><b>5.17</b> Performing PCA</a></li>
<li class="chapter" data-level="5.18" data-path="reformatting-pca-output.html"><a href="reformatting-pca-output.html"><i class="fa fa-check"></i><b>5.18</b> Reformatting PCA output</a></li>
<li class="chapter" data-level="5.19" data-path="pca-plot.html"><a href="pca-plot.html"><i class="fa fa-check"></i><b>5.19</b> PCA plot</a></li>
<li class="chapter" data-level="5.20" data-path="proportion-of-variance-explained.html"><a href="proportion-of-variance-explained.html"><i class="fa fa-check"></i><b>5.20</b> Proportion of variance explained</a></li>
<li class="chapter" data-level="5.21" data-path="conclusion-3.html"><a href="conclusion-3.html"><i class="fa fa-check"></i><b>5.21</b> Conclusion</a></li>
<li class="chapter" data-level="5.22" data-path="homework-4.html"><a href="homework-4.html"><i class="fa fa-check"></i><b>5.22</b> Homework</a></li>
<li class="chapter" data-level="5.23" data-path="required-homework-3.html"><a href="required-homework-3.html"><i class="fa fa-check"></i><b>5.23</b> Required homework</a></li>
<li class="chapter" data-level="5.24" data-path="optional-homework-1.html"><a href="optional-homework-1.html"><i class="fa fa-check"></i><b>5.24</b> Optional homework</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="authors.html"><a href="authors.html"><i class="fa fa-check"></i>Authors</a></li>
<li class="divider"></li>
<p style="text-align:center; font-size: 11pt"> <a href="https://github.com/jhudsl/OTTR_Template" target="blank" > Published with</a> <a href="https://bookdown.org/" target="blank"> <b>bookdown</b></a> (by <a href="http://jhudatascience.org/"> The Johns Hopkins Data Science Lab</a>) </p>
<p style="text-align:center; font-size: 12px;"> <a href="https://github.com/rstudio4edu/rstudio4edu-book/"> Style adapted from: rstudio4edu-book </a> <a href ="https://creativecommons.org/licenses/by/2.0/"> (CC-BY 2.0) </a></p>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Human Genome Variation Lab</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="tidying-vcf-data" class="section level2" number="5.5">
<h2><span class="header-section-number">5.5</span> Tidying VCF data</h2>
<p>If we try to look at <code>vcf</code>, we can see that it’s an “Object of Class vcfR”:</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="tidying-vcf-data.html#cb77-1" aria-hidden="true" tabindex="-1"></a>vcf</span></code></pre></div>
<pre><code>## ***** Object of Class vcfR *****
## 2504 samples
## 1 CHROMs
## 10,000 variants
## Object size: 194.1 Mb
## 0 percent missing data
## *****        *****         *****</code></pre>
<p><code>vcfR</code> reformats VCFs into its own data class, which is easy to manipulate using <code>vcfR</code> functions, but hard to access directly from the <code>vcf</code> variable.</p>
<p>We’ll use the <code>vcfR2tidy</code> function in order to convert <code>vcf</code> into <strong>tidy dataframes</strong> suitable for analysis with R:</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="tidying-vcf-data.html#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert vcf into three tidy dataframes</span></span>
<span id="cb79-2"><a href="tidying-vcf-data.html#cb79-2" aria-hidden="true" tabindex="-1"></a>tidied <span class="ot">&lt;-</span> <span class="fu">vcfR2tidy</span>(vcf,</span>
<span id="cb79-3"><a href="tidying-vcf-data.html#cb79-3" aria-hidden="true" tabindex="-1"></a>                    <span class="co"># tell vcfR to turn these INFO fields into integers</span></span>
<span id="cb79-4"><a href="tidying-vcf-data.html#cb79-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">info_types =</span> <span class="fu">c</span>(<span class="at">AF =</span> <span class="st">&quot;n&quot;</span>, <span class="at">EAS_AF =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb79-5"><a href="tidying-vcf-data.html#cb79-5" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">EUR_AF =</span> <span class="st">&quot;n&quot;</span>, <span class="at">AFR_AF =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb79-6"><a href="tidying-vcf-data.html#cb79-6" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">AMR_AF =</span> <span class="st">&quot;n&quot;</span>, <span class="at">SAS_AF =</span> <span class="st">&quot;n&quot;</span>))</span>
<span id="cb79-7"><a href="tidying-vcf-data.html#cb79-7" aria-hidden="true" tabindex="-1"></a>tidied</span></code></pre></div>
<pre><code>## $fix
## # A tibble: 10,000 x 20
##    ChromKey CHROM    POS ID    REF   ALT    QUAL FILTER    AF AC    NS    AN   
##       &lt;int&gt; &lt;chr&gt;  &lt;int&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
##  1        1 chr21 1.00e7 &lt;NA&gt;  C     A        NA PASS    0.02 89    2548  5008 
##  2        1 chr21 1.03e7 &lt;NA&gt;  A     G        NA PASS    0.02 84    2548  5008 
##  3        1 chr21 1.03e7 &lt;NA&gt;  G     A        NA PASS    0    3     2548  5008 
##  4        1 chr21 1.03e7 &lt;NA&gt;  A     C        NA PASS    0    1     2548  5008 
##  5        1 chr21 1.03e7 &lt;NA&gt;  C     A        NA PASS    0    3     2548  5008 
##  6        1 chr21 1.03e7 &lt;NA&gt;  A     T        NA PASS    0    4     2548  5008 
##  7        1 chr21 1.03e7 &lt;NA&gt;  G     A        NA PASS    0    4     2548  5008 
##  8        1 chr21 1.03e7 &lt;NA&gt;  A     G        NA PASS    0    7     2548  5008 
##  9        1 chr21 1.03e7 &lt;NA&gt;  A     C        NA PASS    0    1     2548  5008 
## 10        1 chr21 1.03e7 &lt;NA&gt;  G     C        NA PASS    0    1     2548  5008 
## # … with 9,990 more rows, and 8 more variables: EAS_AF &lt;dbl&gt;, EUR_AF &lt;dbl&gt;,
## #   AFR_AF &lt;dbl&gt;, AMR_AF &lt;dbl&gt;, SAS_AF &lt;dbl&gt;, VT &lt;chr&gt;, EX_TARGET &lt;chr&gt;,
## #   DP &lt;chr&gt;
## 
## $gt
## # A tibble: 25,040,000 x 5
##    ChromKey      POS Indiv   gt_GT gt_GT_alleles
##       &lt;int&gt;    &lt;int&gt; &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;        
##  1        1 10005999 HG00096 0|0   C|C          
##  2        1 10325486 HG00096 0|0   A|A          
##  3        1 10336823 HG00096 0|0   G|G          
##  4        1 10337236 HG00096 0|0   A|A          
##  5        1 10339129 HG00096 0|0   C|C          
##  6        1 10339141 HG00096 0|0   A|A          
##  7        1 10339175 HG00096 0|0   G|G          
##  8        1 10339803 HG00096 0|0   A|A          
##  9        1 10339869 HG00096 0|0   A|A          
## 10        1 10347083 HG00096 0|0   G|G          
## # … with 25,039,990 more rows
## 
## $meta
## # A tibble: 13 x 5
##    Tag    ID      Number Type   Description                                     
##    &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;                                           
##  1 INFO   AF      A      Float  Estimated allele frequency in the range (0,1)   
##  2 INFO   AC      A      Integ… Total number of alternate alleles in called gen…
##  3 INFO   NS      1      Integ… Number of samples with data                     
##  4 INFO   AN      1      Integ… Total number of alleles in called genotypes     
##  5 INFO   EAS_AF  A      Float  Allele frequency in the EAS populations calcula…
##  6 INFO   EUR_AF  A      Float  Allele frequency in the EUR populations calcula…
##  7 INFO   AFR_AF  A      Float  Allele frequency in the AFR populations calcula…
##  8 INFO   AMR_AF  A      Float  Allele frequency in the AMR populations calcula…
##  9 INFO   SAS_AF  A      Float  Allele frequency in the SAS populations calcula…
## 10 INFO   VT      .      String indicates what type of variant the line represe…
## 11 INFO   EX_TAR… 0      Flag   indicates whether a variant is within the exon …
## 12 INFO   DP      1      Integ… Approximate read depth; some reads may have bee…
## 13 FORMAT gt_GT   1      String Phased Genotype</code></pre>
<p><code>tidied</code> is a <em>list</em> (a type of R object) of three dataframes:</p>
<ul>
<li><code>fix</code>: Location, identity, allele frequency, etc. of <strong>variants</strong> in the VCF</li>
<li><code>gt</code>: <strong>Genotypes</strong> of the VCF samples</li>
<li><code>meta</code>: Information from the VCF <strong>header</strong></li>
</ul>
<p>We’ll be using the variants information from <code>tidied</code>, so let’s store it in a separate table:</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="tidying-vcf-data.html#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make variants dataframe</span></span>
<span id="cb81-2"><a href="tidying-vcf-data.html#cb81-2" aria-hidden="true" tabindex="-1"></a>variants <span class="ot">&lt;-</span> tidied<span class="sc">$</span>fix</span>
<span id="cb81-3"><a href="tidying-vcf-data.html#cb81-3" aria-hidden="true" tabindex="-1"></a>variants</span></code></pre></div>
<pre><code>## # A tibble: 10,000 x 20
##    ChromKey CHROM    POS ID    REF   ALT    QUAL FILTER    AF AC    NS    AN   
##       &lt;int&gt; &lt;chr&gt;  &lt;int&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
##  1        1 chr21 1.00e7 &lt;NA&gt;  C     A        NA PASS    0.02 89    2548  5008 
##  2        1 chr21 1.03e7 &lt;NA&gt;  A     G        NA PASS    0.02 84    2548  5008 
##  3        1 chr21 1.03e7 &lt;NA&gt;  G     A        NA PASS    0    3     2548  5008 
##  4        1 chr21 1.03e7 &lt;NA&gt;  A     C        NA PASS    0    1     2548  5008 
##  5        1 chr21 1.03e7 &lt;NA&gt;  C     A        NA PASS    0    3     2548  5008 
##  6        1 chr21 1.03e7 &lt;NA&gt;  A     T        NA PASS    0    4     2548  5008 
##  7        1 chr21 1.03e7 &lt;NA&gt;  G     A        NA PASS    0    4     2548  5008 
##  8        1 chr21 1.03e7 &lt;NA&gt;  A     G        NA PASS    0    7     2548  5008 
##  9        1 chr21 1.03e7 &lt;NA&gt;  A     C        NA PASS    0    1     2548  5008 
## 10        1 chr21 1.03e7 &lt;NA&gt;  G     C        NA PASS    0    1     2548  5008 
## # … with 9,990 more rows, and 8 more variables: EAS_AF &lt;dbl&gt;, EUR_AF &lt;dbl&gt;,
## #   AFR_AF &lt;dbl&gt;, AMR_AF &lt;dbl&gt;, SAS_AF &lt;dbl&gt;, VT &lt;chr&gt;, EX_TARGET &lt;chr&gt;,
## #   DP &lt;chr&gt;</code></pre>
</div>
<hr>
<center> 
  <div class="footer">
      All illustrations <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY. </a>
      <br>
      All other materials <a href= "https://creativecommons.org/licenses/by/4.0/"> CC-BY </a> unless noted otherwise.
  </div>
</center>
            </section>

          </div>
        </div>
      </div>
<a href="setup-2.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="the-allele-frequency-spectrum.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
