```{r, include = FALSE}
ottrpal::set_knitr_image_path()
knitr::opts_knit$set(root.dir = '10-gene_expression')
# install packages for the lab
# install.packages("MASS")
```

# Gene expression

In this lab, we'll use data from the Genotype-Tissue Expression (GTEx) Project to explore how genetic variation impacts gene expression.

#### Learning objectives

After completing this chapter, you'll be able to:

1. Define expression and splicing quantitative trait loci.
2. Explain the challenge of multiple testing in eQTL studies and how it is typically handled.
3. Navigate the GTEx Portal to find expression data for genes of interest.


## Gene expression

In this class, we've mostly discussed variation at the level of DNA sequence. However, variation in **gene expression** -- and how it interacts with genotype data -- is an equally important aspect of phenotype diversity.

***
<details> <summary> Measuring gene expression </summary>

Sequencing mRNA molecules (**RNA sequencing**), aligning them to a reference genome, and counting the reads per gene provides us with a proxy for measuring expression level.

<center>

![**Fig. 1 ([source](https://www.ebi.ac.uk/training/online/courses/functional-genomics-ii-common-technologies-and-data-analysis-methods/rna-sequencing/performing-a-rna-seq-experiment/data-analysis/read-mapping-or-alignment/)).** Counting transcript expression from RNA-seq reads.](10-gene_expression/images/rnaseq.jpg){width=85%}

</center>

</details>
***


## The Genotype-Tissue Expression project

In 2010, the NIH launched the first large-scale dataset of gene expression data, called the **[Genotype-Tissue Expression (GTEx)](https://gtexportal.org/home/tissueSummaryPage) project**. The final version of this dataset (v8) was released in 2020.

GTEx is currently the most comprehensive gene expression dataset in existence. It involved the sequencing of **(whole genomes (DNA-seq))** as well as **transcriptomes (RNA-seq)** from 948 recently-deceased individuals, with up to 54 tissues sampled throughout their bodies. 

One of the main motivations of GTEx was to better understand the genetic control of gene expression. How does genetic variation contribute to variation in amount, splicing, and tissue specificity of expressed RNA?

<center>

![**Fig. 2 ([source](https://gtexportal.org/home/tissueSummaryPage)).** Summary of individuals sequenced by GTEx.](10-gene_expression/images/gtex.png)

</center>


## GTEx portal

The [GTEx Portal](https://gtexportal.org/) provides an interface to summarize and visualize these data. In the search bar, we can look up a gene of interest and see:

* Violin plots depicting inter-individual variation in gene expression across all tissues
* Variants that impact expression or splicing of that gene

<center>

![**Fig. 3 ([source](https://gtexportal.org/home/gene/DNMT3A)).** Distributions of gene expression across tissues for the _DNMT3A_ gene.](10-gene_expression/images/dnmt3a.png)

</center>


## Genetic effects on gene expression

***
<details> <summary> Why does expression for the same gene vary across individuals? </summary>

There are many factors that can contribute to variability in gene expression in a population:

* Environment (ex: diet, climate, disease)
* Stochasticity/noise
* Epigenetics (methylation, repressive/activating histone marks)
* **Genetics** (genetic variants in promoter, enhancer, etc. sequences)

</details>
***

***
<details> <summary> How do we discover genetic variants that influence gene expression? </summary>

We can perform a **genome-wide association study**, similar to what we would do for an organismal phenotype like height. This time, our phenotype is the expression of a specific gene, and we can search for genetic variants that impact this expression.

</details>
***


## Expression QTLs

The variants we identify with this GWAS approach are called **expression quantitative trait loci (eQTLs)**.

Notably, because the number of phenotypes we have is equal to the number of genes we're testing, we would have to perform ~20,000 GWASs (the number of gemes) for every eQTL study in humans, and test ~5 million SNPs per GWAS.

***
<details> <summary> Multiple testing in eQTL studies </summary>

When we're performing this many repeated tests, you run the risk that any hits you find are just due to chance.

In a GWAS, we would use **multiple testing correction** to account for this. However, because of the overwhelming number of tests we need to perform to identify eQTLs, our significance threshold would be so low that we likely wouldn't discover anything.

</details>
***

***
<details> <summary> Reducing the number of tests </summary>

In order to reduce the number of tests, we typically focus on eQTL mapping just in in the region around the gene body, which dramatically decreases the number of SNPs we're testing.

The rationale for this approach is that nearby regulatory regions (promoters, enhancers, etc.) are more likely to influence expression.

<center>

![**Fig. 4.** eQTL studies typically restrict their QTL search space to variants within 1 Mb of the gene.](10-gene_expression/images/cis_eqtl.jpg){width=85%}

</center>

</details>
***


## eQTLs in the GTEx Portal

The GTEx Portal includes an option to view all significant eQTLs for a gene, below the plots of tissue expression. Note that these eQTLs are labeled by **tissue** -- they may not impact transcription outside of that tissue.

<center>

![**Fig. 5.** Significant eQTLs for _DNMT3A_ in the GTEx Portal.](10-gene_expression/images/gtex_eqtl.png)

</center>

***
<details> <summary> eQTL violin plots </summary>

The `Actions` column provides quick access to several visualizations of each eQTL, including a **violin plot** that shows how it impacts expression:

<center>

![**Fig. 6.** The `T` allele at the `chr2_25264228_C_T_b38` SNP decreases _DNMT3A_ expression.](10-gene_expression/images/dnmt3a_eqtl.png){width=50%}

</center>

</details>
***


## Splicing QTLs

In addition to variants that impact gene expression, there are **splicing quantitative trait loci (sQTLs)** that alter how a gene is spliced. Some sQTL mechanisms include:

* Create or destroying a splicing donor/acceptor site
* Changing the binding site for a protein that regulates splicing
* Altering a splicing factor protein itself

<center>

![**Fig. 7.** sQTLs may alter splice donor/acceptor sites, or binding sites for transcription factors that regulate splicing.](10-gene_expression/images/sqtl.jpg){width=85%}

</center>

***
<details> <summary> sQTLs in GTEx </summary>

The GTEx Portal also provides a list of sQTLs for each gene. Note that the violin plots for these QTLs refer to a specific intron of _DNMT3A_ whose inclusion rate is affected by each SNP.

<center>

![**Fig. 8.** The `A` allele at the `chr2_25530386_G_A_b38` SNP reduces excision of the `chr2:25246776:25247051:clu_42604` intron of  _DNMT3A_.](10-gene_expression/images/dnmt3a_sqtl.png){width=50%}

</center>

</details>
***


## Setup

Now we'll look at the GTEx data ourselves to perform our own gene expression analyses.

### R packages

In addition to `tidyverse`, we'll also load the `MASS` package, which is useful for various statistical functions.

```{r, warning = FALSE}
library(tidyverse)
library(MASS)
```


## Data

The GTEx Portal provides [links for downloading](https://gtexportal.org/home/datasets) curated and summarized forms of its data, including giant matrices that encode the expression of every gene across all samples and tissues.

For ease of manipulation in R, we've subset this data to 150 samples, highly expressed genes, and only data from **liver** and **lung** tissue.

```{r, eval = FALSE}
gtex <- read_table("gtex_subset.txt.gz")

head(gtex)
```
```{r, message = FALSE, echo = FALSE}
# code block to preview gtex file since loading it into bookdown won't work
# write.table(head(gtex), "gtex_head.txt", sep = "\t", quote = FALSE, row.names = FALSE)
read_table("gtex_head.txt") %>%
  head()
```


The columns of this dataframe are:

* `Sample`: Individual sequenced
* `Age`: Individual's age range
* `Sex`: Individual's sex
* `Death_Hardy`: Individual's cause of death, measured on the [Hardy Scale](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/variable.cgi?study_id=phs000424.v4.p1&phv=169092)
* `Tissue`: Tissue measured
* `Gene_ID`: Ensembl gene ID
* `Gene_Name`: The common gene name
* `Counts`: Expression level for the gene
  * Ex: `GTEX-111YS` has 59 sequencing reads that mapped to the `SAMD11` gene

***
<details> <summary> Data normalization </summary>

The expression levels in this table have been **normalized** to account for factors such as sequencing variation between samples -- i.e., if we collected more sequencing data from one individual than another.

</details>
***


## Differential gene expression

Using this dataset, we can fit **regression models** to ask about differences in gene expression between conditions. For example, let's say we're interested in whether [the _ACE2_ gene](https://www.nytimes.com/interactive/2020/03/11/science/how-coronavirus-hijacks-your-cells.html) exhibits differences in expression between males and females.

First we subset the data to the relevant gene and tissue (_ACE2_ and lung) for our test:

```{r, eval = FALSE}
subset <- gtex %>%
  # filter for tissue and gene of interest
  filter(Tissue == "Lung" & Gene_Name == "ACE2")

head(subset)
```
```{r, echo = FALSE}
# code block to preview gtex file since loading it into bookdown won't work
# write.table(subset, "gtex_ace2.txt", sep = "\t", quote = FALSE, row.names = FALSE)
subset <- read_table("gtex_ace2.txt")
head(subset)
```

Then we can fit a model to our data. Note that we're using the `glm.nb` function, which uses a **negative binomial distribution** -- good for modeling discrete (i.e., non-continuous) data, such as sequencing read counts.

```{r, eval = FALSE}
# fit model and print summary
glm.nb(formula = Counts ~ factor(Sex),
       data = subset) %>%
  summary()
```
```
Call:
glm.nb(formula = Counts ~ factor(Sex), data = subset, init.theta = 0.9820746913, 
    link = log)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.0205  -0.9991  -0.5441  -0.0219   4.9984  

Coefficients:
             Estimate Std. Error z value Pr(>|z|)    
(Intercept)    5.5530     0.1709  32.494  < 2e-16 ***
factor(Sex)M  -0.7907     0.2121  -3.727 0.000194 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for Negative Binomial(0.9821) family taken to be 1)

    Null deviance: 129.67  on 99  degrees of freedom
Residual deviance: 115.01  on 98  degrees of freedom
AIC: 1214.5

Number of Fisher Scoring iterations: 1


              Theta:  0.982 
          Std. Err.:  0.123 

 2 x log-likelihood:  -1208.475 
```

***
<details> <summary> Does sex impact _ACE2_ expression? </summary>

The p-value is significant (`p = 0.000194`), suggesting that sex does affect _ACE2_ expression. Based on the coefficient of `-0.7907`, it looks like males tend to have ~0.8 fewer _ACE2_ transcripts in lung tissue.

</details>
***


## Conclusion

In this lab, we explored the **GTEx** dataset to understand how genetic variation impacts gene expression.

* We can treat gene expression as a phenotype to search for **expression and splicing QTLs**, the same way we would look for variants associated with an organismal trait.
<br></br>
* Due to the **multiple testing burden** of testing millions of variants across thousands of genes, eQTL studies typically limit themselves to variants close to the gene body.
<br></br>
* The **GTEx Portal** provides a convenient interface for browsing QTL and gene expression information across tissues.


## Homework

#### Learning Objectives

* Become familiar with the data available in the GTEx dataset
* Interpret linear models in the context of gene expression

#### Assignment

Fit a regression model (or multiple!) to the GTEx data we downloaded in class and test for differential expression between two (or more) conditions. For example, you could look at another gene, compare between tissues, age groups, or Hardy classifications, etc.

Based on the results, does the condition affect expression of your gene?