# Gene expression

In this lab, we'll use data from the Genotype-Tissue Expression (GTEx) Project to explore how genetic variation impacts gene expression.

#### Learning objectives

After completing this chapter, you'll be able to:

1. Define expression and splicing quantitative trait loci.
2. Explain the challenge of multiple testing in eQTL studies and how it is typically handled.
3. Navigate the GTEx Portal to find expression data for genes of interest.


***


## The GTEx portal

https://gtexportal.org/home/tissueSummaryPage


***


## Setup

### R packages

```{r}
library(tidyverse)
library(MASS)
```

### Data

```{r}
gtex <- read.table("gtex_subset.txt.gz", header = TRUE)

head(gtex)
```


***


## Differential gene expression

Run the code block below to subset the data to the relevant gene and tissue:

```{r}
subset <- gtex %>%
  # filter for tissue and gene of interest
  filter(Tissue == "Lung" & Gene_Name == "ACE2")

head(subset)
```

Fit a regression model with a negative binomial distribution:

```{r}
# fit model and print summary
glm.nb(formula = Counts ~ factor(Sex),
       data = subset) %>%
  summary()
```


***


## Homework

#### Learning Objectives

* Become familiar with the data available in the GTEx dataset
* Interpret linear models in the context of gene expression

#### Assignment

Fit a regression model (or multiple!) to the GTEx data we downloaded in class and test for differential expression between two (or more) conditions. For example, you could look at another gene, compare between tissues, age groups, or Hardy classifications, etc.

Based on the results, does the condition affect expression of your gene?

```{r}

```

