# Scans for selection

In this lab, we'll explore three methods for identifying signatures of selection: F~ST~, population branch statistic (PBS), and extended haplotype homozygosity (EHH).

#### Learning objectives

After completing this chapter, you'll be able to:

1. Describe the genetic signatures of selection and what they reveal about the strength and timing of the selective sweep.
2. Calculate and interpret population differentiation with F~ST~.
3. Understand why PBS provides additional information over two-population tests of selection.
4. Explain how EHH and iHS leverage haplotype information to identify selection.


***


## Setup

### R packages

```{r}
library(tidyverse)
library(vcfR)
library(ggtree)
```


***


## The F~ST~ statistic

F~ST~ quantifies differentiation between populations, based on the frequencies of alleles at one variable site.

$$
\textrm{F}_{ST} = \frac{H_T - \textrm{mean}(H_S)}{H_T}
$$

We'll calculate F~ST~ with the `genetic_diff` function from `vcfR`.

```{r}
?genetic_diff
```


***


## Data (for F~ST~)

```{r}
# read genotype data with vcfR
vcf <- read.vcfR(file = "random_variable_sites.vcf.gz")
```

```{r}
# read metadata
metadata <- read.table(file = "integrated_call_samples.txt",
                       header = TRUE)
head(metadata)
```


***


## Calculating F~ST~

Use the `metadata` table to generate a vector of superpopulation labels for `genetic_diff`, converting to factors:

```{r}
pop_labels <- as.factor(metadata$superpop)
pop_labels
```

Run `genetic_diff` on the VCF:

```{r }
# calculate gst
gst_results <- genetic_diff(vcf, pop_labels) %>%
  # order dataframe by descending gst value
  arrange(-Gst)

head(gst_results)
```

Plot the distribution of G~ST~ across all variants.

```{r}

```

Look up top G~ST~ hit: https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38

Examine top variant in the GGV browser: https://popgen.uchicago.edu/ggv/?data=%221000genomes


***


## Population branch statistic (PBS)

PBS is based on calculating $\textrm{F}_{ST}$ for each pairwise comparison of the three populations. We then calculate the branch length ($\mathbf{T}$) that separates each population pair:

$$
T = -\mathrm{log}(1 - \textrm{F}_{ST})
$$

**PBS** is then computed as:

$$
\textrm{PBS} = \frac{\textrm{T}_{AB} + \textrm{T}_{AC} - \textrm{T}_{BC}}{2}
$$


***


## Data (for PBS)

We'll calculate PBS using data from the paper: [Tucci, S. et al. (2018). Evolutionary history and adaptation of a human pygmy population of Flores Island, Indonesia.](https://science.sciencemag.org/content/361/6401/511.abstract).

Flores Island was home to fossils of the archaic hominin species _**Homo floresiensis**_, also called "hobbits" because their skeletons are 3 ft. tall. This study analyzed data from three modern human populations:

* The Rampasasa (RPS) population from Flores
* A Han Chinese (CHB) population
* A Melanesian population from Papua New Guinea (PNG)

Load pre-calculated pairwise F~ST~ results for chromosome 11 into R:

```{r}
fst_results <- read.table("fst_results.txt.gz", header = TRUE)

head(fst_results)
```

Fill in the gaps in the code block below to calculate T and PBS from the FST values (using `mutate`):

```{r, eval = FALSE}
pbs <- fst_results %>%
  # calculate branch lengths between populations
  mutate(T_rps_chb = _________,
         T_rps_png = _________,
         T_png_chb = _________,) %>%
  # calculate pbs
  mutate(pbs = _________) %>%
  # sort by descending pbs value
  arrange(-pbs)
```

Generate a Manhattan plot with `geom_point`, where we plot PBS on the y-axis and position on the x-axis.

```{r}

```

View top PBS hits: https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19


***


## Plotting PBS trees

Another useful way to visualize PBS is by comparing trees for the top PBS outliers to the genome-wide average tree. Run the code blocks below to plot these trees:

```{r}
# create average tree
tr_mean <- rtree(n = 3,
                 rooted = FALSE,
                 br = c(mean(pbs$T_rps_png),
                        mean(pbs$T_rps_chb),
                        mean(pbs$T_png_chb)))

# plot average tree
ggtree(tr_mean, layout = "daylight") +
  geom_treescale(width = 0.1) +
  geom_tip(label = c("RPS", "PNG", "CHB"))
```

```{r}
# create tree for top snp
tr_top <- rtree(n = 3,
                rooted = FALSE,
                br = c(pbs[1,]$T_rps_png,
                       pbs[1,]$T_rps_chb,
                       pbs[1,]$T_png_chb))

# plot top snp tree
ggtree(tr_top, layout = "daylight") +
  geom_treescale(width = 0.1) +
  geom_tip(label = c("RPS", "PNG", "CHB"))
```


***


## Extended haplotype homozygosity (EHH)

https://pophuman.uab.cat/


***


## Homework

#### Goals & Learning Objectives

* Interpret multiple statistics for measuring selection
* Explain how specific statistics can give different results because they measure different genetic signatures

### Required homework

Read [**this review paper**](https://www.science.org/doi/10.1126/science.aaf5098) on genomic evidence of human local adaptation by Fan et al. Find examples of local adaptation (genes and populations) in the paper, and look up the relevant populations (or related populations) and tests of selection in the [PopHuman browser](pophuman.uab.cat).

- Are the signatures of selection apparent?
- Are the signature apparent based on all statistics?
- Why do certain statistics capture evidence of selection at certain loci but not others?