Generate VCF without AF = 0 variants for GST:

```{r}
# read random_snippet.vcf.gz from pop_structure module
vcf <- read.vcfR(file = "../04-pop_structure/random_snippet.vcf.gz")

# calculate af
afs <- maf(vcf)
# get AF > 0 rows
nonzero <- afs[, 4] > 0

# subset vcf
vcf_nonzero <- vcf[nonzero, ]
# write new vcf
write.vcf(vcf_nonzero, file = "random_variable_sites.vcf.gz")
```


Generate pre-processed data for calculating PBS:

```{r}
# FST for RPS-CHB comparison
rps_chb <- read_tsv("chr11.rps.chb.fst", 
                    col_names = c("chr", "pos", "rps.af", "chb.af", "fst.rps.chb"))

head(rps_chb)

# FST for RPS-PNG comparison
rps_png <- read_tsv("chr11.rps.png.fst", 
                    col_names = c("chr", "pos", "rps.af", "png.af", "fst.rps.png"))

head(rps_png)

# FST for PNG-CHB comparison
png_chb <- read_tsv("chr11.png.chb.fst", 
                    col_names = c("chr", "pos", "png.af", "chb.af", "fst.png.chb"))

head(png_chb)
```


```{r}
# merge fst results
fst_results <- merge(rps_chb,
                     rps_png[, c("chr", "pos", "png.af", "fst.rps.png")],
             by = c("chr", "pos")) %>%
  merge(.,
        png_chb[, c("chr", "pos", "fst.png.chb")],
        by = c("chr", "pos")) %>%
  # drop nas
  drop_na() %>%
  # drop variants that are fixed in at least one population
  filter(rps.af > 0 & rps.af < 1 &
         png.af > 0 & png.af < 1 &
         chb.af > 0 & chb.af < 1) %>%
  # reorder columns
  .[, c("chr", "pos", "rps.af", "chb.af", "png.af",
        "fst.rps.chb", "fst.rps.png", "fst.png.chb")]

write.table(fst_results, file = "fst_results.txt",
            sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)
```


## Haplotype-based statistics

it removes the variation that existed in this region before

During the last portion of class, I described two related haplotype-based statistics for quantifying evidence of positive selection (see PowerPoint). The statistics are termed "extended haplotype homozygosity" (EHH) - how identical are the haps in this region? The probability that any two chromosomes are homozygous at SNPs between point A and B and "integrated haplotype statistic" (iHS). 

While some R packages exist to compute these statistics (e.g., the `rehh` package), they require some tricky preprocessing of VCF files, which did not seem like a good use of class time. Moreover, these statistics have already been computed genome-wide for all of the populations in the 1000 Genomes dataset, available from the [PopHuman browser](pophuman.uab.cat). I showed the example of iHS for the CEU population (Utah Residents [CEPH] with Northern and Western European Ancestry) at the lactase (*LCT*) locus. Positive selection at this locus is the most famous example of local adaptation in humans, leaving dramatic signatures on patterns of variation in the surrounding region.

depending on the type of selection (existing set of haplotypes, new variant) the ehh signature is different

start with a core set of snps that are all in LD (or a new mutation), march out to flanking snps that are farther away, and you count how much recombination has happened

you can do this for the new haplotype and then the ancestral one. The signature of a selection sweep is the difference between the ehh for the derived and ancestral alleles (by comaparing area under the curves) - integrated haplotype statistic (iHS)

visualize with tree-like plots where the width of bar represents frequency of haplotype

look at example from lactase locus, classic example of positive selection in human populations. Selective sweep in ancestor of european populations that allowed digestion of milk into adulthood, EHH across megabases of sequence. Compare to the ancesetral haplotype and see that there's much more recombination and diversity in the ancestral regions

pophuman browser lets you pull up different popgen statistics like ihs, nucleotide diversity (pi - expected to be low in a region of selection). compare to mean across genome. This is one of the most extreme that we know in humans